// @dytesdk/ui-kit/dist/cjs/dyte-file-dropzone.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-f5ba2db8.js');
require('./store-7f623fdb.js');
require('./breakout-rooms-manager-87997e6a.js');
require('./livestream-3ff039ea.js');
const defaultIconPack = require('./default-icon-pack-ddb6b86f.js');
const index$1 = require('./index-d63f2c88.js');
require('./user-prefs-cefc791a.js');
require('./breakout-rooms-23529df9.js');
require('./flags-88945715.js');

const dyteFileDropzoneCss = "#dropzone{position:absolute;top:var(--dyte-space-0, 0px);right:var(--dyte-space-0, 0px);bottom:var(--dyte-space-0, 0px);left:var(--dyte-space-0, 0px);z-index:10;display:none;flex-direction:column;align-items:center;justify-content:center;--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-800, 30 30 30) / var(--tw-bg-opacity));color:rgb(var(--dyte-colors-text-700, 255 255 255 / 0.64))}#dropzone.active{display:flex;animation:0.2s slide-up ease-in}";

const DyteFileDropzone = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.onDropCallback = index.createEvent(this, "dropCallback", 7);
    this.hostEl = undefined;
    this.iconPack = defaultIconPack.defaultIconPack;
    this.t = index$1.useLanguage();
    this.dropzoneActivated = false;
  }
  connectedCallback() {
    if (!this.hostEl)
      throw new Error('hostEl prop is required');
    this.hostEl.addEventListener('dragover', (e) => {
      e.preventDefault();
      this.dropzoneActivated = true;
    });
    this.hostEl.addEventListener('dragleave', () => {
      this.dropzoneActivated = false;
    });
    this.hostEl.addEventListener('drop', (e) => {
      e.preventDefault();
      this.dropzoneActivated = false;
      this.onDropCallback.emit(e);
    });
  }
  render() {
    return (index.h(index.Host, null, index.h("div", { id: "dropzone", class: { active: this.dropzoneActivated }, part: "dropzone" }, index.h("dyte-icon", { icon: this.iconPack.attach, iconPack: this.iconPack, t: this.t }), index.h("p", null, this.t('chat.send_attachment')))));
  }
};
DyteFileDropzone.style = dyteFileDropzoneCss;

exports.dyte_file_dropzone = DyteFileDropzone;
