// @dytesdk/ui-kit/dist/cjs/dyte-header.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-f5ba2db8.js');
require('./store-7f623fdb.js');
require('./breakout-rooms-manager-87997e6a.js');
const defaultUiConfig = require('./default-ui-config-d5874568.js');
require('./livestream-3ff039ea.js');
const defaultIconPack = require('./default-icon-pack-ddb6b86f.js');
const index$1 = require('./index-d63f2c88.js');
require('./user-prefs-cefc791a.js');
const index$2 = require('./index-c0cb7142.js');
require('./breakout-rooms-23529df9.js');
require('./flags-88945715.js');

const dyteHeaderCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{display:flex;height:var(--dyte-space-12, 48px);align-items:center;--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-1000, 8 8 8) / var(--tw-bg-opacity));padding-left:var(--dyte-space-4, 16px);padding-right:var(--dyte-space-4, 16px)}@media only screen and (max-device-height: 480px) and (orientation: landscape){:host{display:none !important}}@media only screen and (max-height: 480px) and (orientation: landscape){:host{display:none !important}}";

const DyteHeader = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.variant = 'solid';
    this.disableRender = false;
    this.meeting = undefined;
    this.config = defaultUiConfig.defaultConfig;
    this.states = undefined;
    this.iconPack = defaultIconPack.defaultIconPack;
    this.t = index$1.useLanguage();
    this.size = undefined;
  }
  render() {
    const defaults = {
      meeting: this.meeting,
      config: this.config,
      states: this.states,
      t: this.t,
      iconPack: this.iconPack,
      size: this.size,
    };
    return (index.h(index.Host, null, !this.disableRender && index.h(index$2.Render, { element: "dyte-header", defaults: defaults, onlyChildren: true }), index.h("slot", null)));
  }
};
DyteHeader.style = dyteHeaderCss;

exports.dyte_header = DyteHeader;
