// @dytesdk/ui-kit/dist/cjs/dyte-leave-button.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-f5ba2db8.js');
const defaultIconPack = require('./default-icon-pack-ddb6b86f.js');
const index$1 = require('./index-d63f2c88.js');
const store = require('./store-7f623fdb.js');
require('./user-prefs-cefc791a.js');

const dyteLeaveButtonCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{display:block}";

const DyteLeaveButton = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.stateUpdate = index.createEvent(this, "dyteStateUpdate", 7);
    this.leave = () => {
      this.stateUpdate.emit({ activeLeaveConfirmation: true });
      store.state.activeLeaveConfirmation = true;
    };
    this.variant = 'button';
    this.size = undefined;
    this.iconPack = defaultIconPack.defaultIconPack;
    this.t = index$1.useLanguage();
  }
  render() {
    const text = this.t('leave');
    return (index.h(index.Host, { label: text }, index.h("dyte-controlbar-button", { size: this.size, iconPack: this.iconPack, t: this.t, class: "leave", onClick: this.leave, icon: this.iconPack.call_end, label: text, variant: this.variant, part: "controlbar-button" })));
  }
};
DyteLeaveButton.style = dyteLeaveButtonCss;

exports.dyte_leave_button = DyteLeaveButton;
