// @dytesdk/ui-kit/dist/cjs/dyte-simple-grid.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-f5ba2db8.js');
const defaultUiConfig = require('./default-ui-config-d5874568.js');
const defaultIconPack = require('./default-icon-pack-ddb6b86f.js');
const index$1 = require('./index-d63f2c88.js');
const index$2 = require('./index-c0cb7142.js');
const grid = require('./grid-5774c6e4.js');
const ResizeObserver_es = require('./ResizeObserver.es-ba961f16.js');
const string = require('./string-29af7ef8.js');

const dyteSimpleGridCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{position:relative;display:block;height:100%;width:100%}";

const DyteSimpleGrid = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.setHostDimensions = () => {
      const { clientWidth: width, clientHeight: height } = this.host;
      this.dimensions = { width, height };
    };
    this.onParticipantTileLoad = (event) => {
      var _a, _b, _c;
      if (!this.meeting)
        return;
      const { participant, videoElement } = event.detail;
      if (!participant || !videoElement)
        return;
      (_a = this.meeting.participants.pip) === null || _a === void 0 ? void 0 : _a.addSource(participant.id, videoElement, participant.videoEnabled, false, (_b = participant.name) !== null && _b !== void 0 ? _b : string.getInitials(participant.name));
      if (participant.videoEnabled) {
        (_c = this.meeting.participants.pip) === null || _c === void 0 ? void 0 : _c.enableSource(participant.id);
      }
    };
    this.onParticipantTileUnload = (event) => {
      var _a;
      if (!this.meeting)
        return;
      const participant = event.detail;
      (_a = this.meeting.participants.pip) === null || _a === void 0 ? void 0 : _a.removeSource(participant.id);
    };
    this.participants = [];
    this.aspectRatio = '16:9';
    this.gap = 8;
    this.size = undefined;
    this.meeting = undefined;
    this.states = undefined;
    this.config = defaultUiConfig.defaultConfig;
    this.iconPack = defaultIconPack.defaultIconPack;
    this.t = index$1.useLanguage();
    this.dimensions = { width: 0, height: 0 };
  }
  connectedCallback() {
    this.resizeObserver = new ResizeObserver_es.index(this.setHostDimensions);
    this.resizeObserver.observe(this.host);
  }
  disconnectedCallback() {
    var _a;
    (_a = this.resizeObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
  }
  render() {
    const defaults = {
      meeting: this.meeting,
      config: this.config,
      states: this.states,
      size: this.size,
      iconPack: this.iconPack,
      t: this.t,
    };
    const { width, height, getPosition } = grid.useGrid({
      dimensions: this.dimensions,
      count: this.participants.length,
      aspectRatio: this.aspectRatio,
      gap: this.gap,
    });
    return (index.h(index.Host, null, this.participants.map((participant, index$1) => {
      const { top, left } = getPosition(index$1);
      return (index.h(index$2.Render, { element: "dyte-participant-tile", defaults: defaults, props: {
          participant,
          style: {
            position: 'absolute',
            top: `${top}px`,
            left: `${left}px`,
            width: `${width}px`,
            height: `${height}px`,
          },
          key: participant.id,
          'data-participant': participant.id,
          onTileLoad: this.onParticipantTileLoad,
          onTileUnload: this.onParticipantTileUnload,
        }, childProps: { participant }, deepProps: true }));
    }), index.h("slot", null)));
  }
  get host() { return index.getElement(this); }
};
DyteSimpleGrid.style = dyteSimpleGridCss;

exports.dyte_simple_grid = DyteSimpleGrid;
