// @dytesdk/ui-kit/dist/cjs/dyte-meeting.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-f5ba2db8.js');
const config = require('./config-ba1005b1.js');
const size = require('./size-d5e6fe50.js');
const index$1 = require('./index-d63f2c88.js');
const defaultUiConfig = require('./default-ui-config-d5874568.js');
const index$2 = require('./index-c0cb7142.js');
const store = require('./store-7f623fdb.js');
require('./breakout-rooms-manager-87997e6a.js');
require('./livestream-3ff039ea.js');
const userPrefs = require('./user-prefs-cefc791a.js');
const ResizeObserver_es = require('./ResizeObserver.es-ba961f16.js');
const defaultIconPack = require('./default-icon-pack-ddb6b86f.js');
require('./keysIn-c1d310a7.js');
require('./isObjectLike-890299f1.js');
require('./flags-88945715.js');
require('./breakout-rooms-23529df9.js');

const dyteMeetingCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{box-sizing:border-box;display:flex;flex-direction:column;--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-1000, 8 8 8) / var(--tw-bg-opacity));color:rgb(var(--dyte-colors-text-900, 255 255 255 / 0.88));overflow:hidden;position:fixed;top:var(--dyte-space-0, 0px);right:var(--dyte-space-0, 0px);bottom:var(--dyte-space-0, 0px);left:var(--dyte-space-0, 0px);height:100%;width:100%}:host([mode='fill']){position:relative}";

const DyteMeeting = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.stateUpdate = index.createEvent(this, "dyteStateUpdate", 7);
    this.roomJoinedListener = () => {
      this.setStates({ meeting: 'joined' });
      store.state.meeting = 'joined';
    };
    this.socketServiceRoomJoinedListener = () => {
      if (this.meeting.stage.status === 'ON_STAGE' || this.meeting.stage.status === undefined)
        return;
      this.setStates({ meeting: 'joined' });
      store.state.meeting = 'joined';
    };
    this.waitlistedListener = () => {
      this.setStates({ meeting: 'waiting' });
      store.state.meeting = 'waiting';
    };
    this.roomLeftListener = ({ state }) => {
      const states = this.states || store.state;
      if ((states === null || states === void 0 ? void 0 : states.roomLeftState) === 'disconnected') {
        this.setStates({ meeting: 'ended', roomLeftState: state });
        store.state.meeting = 'ended';
        return;
      }
      this.setStates({ meeting: 'ended', roomLeftState: state });
      store.state.meeting = 'ended';
      store.state.roomLeftState = state;
    };
    this.mediaPermissionUpdateListener = ({ kind, message }) => {
      if (['audio', 'video'].includes(kind)) {
        if (message === 'ACCEPTED' || message === 'NOT_REQUESTED' || store.state.activeDebugger)
          return;
        const permissionModalSettings = {
          enabled: true,
          kind,
        };
        this.setStates({ activePermissionsMessage: permissionModalSettings });
        store.state.activePermissionsMessage = permissionModalSettings;
      }
    };
    this.joinStateAcceptedListener = () => {
      this.setStates({ activeJoinStage: true });
      this.stateUpdate.emit({ activeJoinStage: true });
      store.state.activeJoinStage = true;
    };
    this.loadConfigFromPreset = true;
    this.applyDesignSystem = true;
    this.mode = 'fixed';
    this.leaveOnUnmount = false;
    this.meeting = undefined;
    this.showSetupScreen = undefined;
    this.t = index$1.useLanguage();
    this.config = defaultUiConfig.defaultConfig;
    this.iconPackUrl = undefined;
    this.size = undefined;
    this.gridLayout = 'row';
    this.states = {
      meeting: 'idle',
      prefs: userPrefs.getUserPreferences(),
    };
    this.iconPack = defaultIconPack.defaultIconPack;
  }
  connectedCallback() {
    var _a;
    this.resizeObserver = new ResizeObserver_es.index(() => this.handleResize());
    this.resizeObserver.observe(this.host);
    if (this.applyDesignSystem &&
      ((_a = this.config) === null || _a === void 0 ? void 0 : _a.designTokens) != null &&
      typeof document !== 'undefined') {
      config.provideDyteDesignSystem(document.documentElement, this.config.designTokens);
    }
    this.meetingChanged(this.meeting);
    this.iconPackUrlChanged(this.iconPackUrl);
  }
  clearListeners(meeting) {
    if (meeting == undefined)
      return;
    meeting.self.removeListener('roomJoined', this.roomJoinedListener);
    meeting.self.removeListener('socketServiceRoomJoined', this.roomJoinedListener);
    meeting.meta.removeListener('socketReconnected', this.roomJoinedListener);
    meeting.self.removeListener('roomLeft', this.roomLeftListener);
    meeting.self.removeListener('mediaPermissionUpdate', this.mediaPermissionUpdateListener);
    meeting.self.removeListener('waitlisted', this.waitlistedListener);
    meeting.self.removeListener('joinStageRequestAccepted', this.joinStateAcceptedListener);
  }
  disconnectedCallback() {
    var _a;
    if (this.leaveOnUnmount) {
      (_a = this.meeting) === null || _a === void 0 ? void 0 : _a.leaveRoom();
    }
    this.resizeObserver.disconnect();
    this.clearListeners(this.meeting);
  }
  meetingChanged(meeting) {
    var _a, _b;
    if (meeting == null)
      return;
    this.setStates({ viewType: meeting.meta.viewType });
    if (this.loadConfigFromPreset && meeting.self.config != null) {
      const theme = meeting.self.config;
      const { config: config$1, data } = config.generateConfig(theme, meeting);
      if (this.config === defaultUiConfig.defaultConfig) {
        // only override the config if the object is same as defaultConfig
        // which means it's a different object passed via prop
        this.config = config$1;
      }
      if (this.showSetupScreen == null) {
        // only override this value if the prop isn't passed
        this.showSetupScreen = data.showSetupScreen;
      }
      if (meeting.connectedMeetings.supportsConnectedMeetings &&
        ((_a = store.state.activeBreakoutRoomsManager) === null || _a === void 0 ? void 0 : _a.destinationMeetingId)) {
        this.showSetupScreen = false;
      }
    }
    if (this.applyDesignSystem &&
      ((_b = this.config) === null || _b === void 0 ? void 0 : _b.designTokens) != null &&
      typeof document !== 'undefined') {
      config.provideDyteDesignSystem(document.documentElement, this.config.designTokens);
    }
    if (meeting.meta.viewType === 'LIVESTREAM') {
      meeting.self.addListener('socketServiceRoomJoined', this.socketServiceRoomJoinedListener);
    }
    meeting.self.addListener('roomJoined', this.roomJoinedListener);
    meeting.self.addListener('waitlisted', this.waitlistedListener);
    meeting.self.addListener('roomLeft', this.roomLeftListener);
    meeting.self.addListener('mediaPermissionUpdate', this.mediaPermissionUpdateListener);
    meeting.self.addListener('joinStageRequestAccepted', this.joinStateAcceptedListener);
    if (meeting.connectedMeetings.supportsConnectedMeetings) {
      meeting.connectedMeetings.once('changingMeeting', this.handleChangingMeeting);
    }
    if (meeting.self.roomJoined) {
      this.states = Object.assign(Object.assign({}, this.states), { meeting: 'joined' });
      store.state.meeting = 'joined';
    }
    else {
      if (this.showSetupScreen) {
        this.states = Object.assign(Object.assign({}, this.states), { meeting: 'setup' });
        store.state.meeting = 'setup';
      }
      else {
        // join directly to the meeting
        meeting.joinRoom();
      }
    }
  }
  async iconPackUrlChanged(url) {
    this.iconPack = await size.getIconPack(url);
  }
  listenState(e) {
    e.stopPropagation();
    this.setStates(e.detail);
  }
  handleChangingMeeting(destinationMeetingId) {
    store.state.activeBreakoutRoomsManager = Object.assign(Object.assign({}, store.state.activeBreakoutRoomsManager), { destinationMeetingId });
  }
  handleResize() {
    this.size = size.getSize(this.host.clientWidth);
  }
  setStates(states) {
    const newStates = Object.assign({}, this.states);
    config.merge(newStates, states);
    this.states = newStates;
  }
  render() {
    var _a, _b;
    const defaults = {
      meeting: this.meeting,
      size: this.size,
      states: this.states || store.state,
      config: this.config,
      iconPack: this.iconPack,
      t: this.t,
    };
    const elementProps = {
      'dyte-grid': {
        layout: this.gridLayout,
      },
    };
    if (((_b = (_a = this.meeting) === null || _a === void 0 ? void 0 : _a.meta) === null || _b === void 0 ? void 0 : _b.viewType) === 'CHAT')
      return index.h(index$2.Render, { element: "dyte-chat", defaults: defaults });
    return index.h(index$2.Render, { element: "dyte-meeting", defaults: defaults, asHost: true, elementProps: elementProps });
  }
  get host() { return index.getElement(this); }
  static get watchers() { return {
    "meeting": ["meetingChanged"],
    "iconPackUrl": ["iconPackUrlChanged"]
  }; }
};
DyteMeeting.style = dyteMeetingCss;

exports.dyte_meeting = DyteMeeting;
