// @dytesdk/ui-kit/dist/cjs/dyte-debugger-toggle.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-f5ba2db8.js');
const defaultIconPack = require('./default-icon-pack-ddb6b86f.js');
const index$1 = require('./index-d63f2c88.js');
const store = require('./store-7f623fdb.js');
require('./user-prefs-cefc791a.js');

const dyteDebuggerToggleCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{display:block}";

const DyteDebuggerToggle = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.stateUpdate = index.createEvent(this, "dyteStateUpdate", 7);
    this.variant = 'button';
    this.meeting = undefined;
    this.states = undefined;
    this.t = index$1.useLanguage();
    this.iconPack = defaultIconPack.defaultIconPack;
    this.size = undefined;
  }
  toggleDebugger() {
    var _a;
    this.stateUpdate.emit({
      activeDebugger: !((_a = this.states) === null || _a === void 0 ? void 0 : _a.activeDebugger),
      activeMoreMenu: false,
    });
    store.state.activeDebugger = !store.state.activeDebugger;
    store.state.activeMoreMenu = false;
  }
  render() {
    var _a;
    if (!((_a = this.meeting) === null || _a === void 0 ? void 0 : _a.troubleshoot))
      return;
    return (index.h(index.Host, { title: this.t('Network & Trouble Shooting') }, index.h("dyte-controlbar-button", { size: this.size, iconPack: this.iconPack, t: this.t, onClick: () => this.toggleDebugger(), icon: this.iconPack.debug, label: this.t('Troubleshooting and Help'), variant: this.variant })));
  }
};
DyteDebuggerToggle.style = dyteDebuggerToggleCss;

exports.dyte_debugger_toggle = DyteDebuggerToggle;
