// @dytesdk/ui-kit/dist/cjs/dyte-caption-toggle.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-f5ba2db8.js');
const store = require('./store-7f623fdb.js');
require('./breakout-rooms-manager-87997e6a.js');
const defaultUiConfig = require('./default-ui-config-d5874568.js');
require('./livestream-3ff039ea.js');
const defaultIconPack = require('./default-icon-pack-ddb6b86f.js');
const index$1 = require('./index-d63f2c88.js');
require('./user-prefs-cefc791a.js');
require('./breakout-rooms-23529df9.js');
require('./flags-88945715.js');

const dyteCaptionToggleCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{display:block}";

const DyteCaptionToggle = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.stateUpdate = index.createEvent(this, "dyteStateUpdate", 7);
    this.variant = 'button';
    this.meeting = undefined;
    this.states = undefined;
    this.config = defaultUiConfig.defaultConfig;
    this.iconPack = defaultIconPack.defaultIconPack;
    this.size = undefined;
    this.t = index$1.useLanguage();
    this.captionEnabled = false;
  }
  connectedCallback() {
    this.meetingChanged(this.meeting);
  }
  meetingChanged(meeting) {
    var _a;
    if (meeting === null)
      return;
    this.captionEnabled =
      (_a = this.meeting.self.permissions.transcriptionEnabled) !== null && _a !== void 0 ? _a : false;
  }
  toggleCaptions() {
    this.stateUpdate.emit({ activeCaptions: !store.state.activeCaptions });
    store.state.activeCaptions = !store.state.activeCaptions;
    this.stateUpdate.emit({ activeMoreMenu: false });
    store.state.activeMoreMenu = false;
  }
  render() {
    if (!this.captionEnabled)
      return null;
    const captionsEnabled = this.states.activeCaptions;
    return (index.h(index.Host, { tabIndex: 0, role: "log", "aria-label": `Picture-in-Picture mode` }, index.h("dyte-controlbar-button", { part: "controlbar-button", size: this.size, iconPack: this.iconPack, t: this.t, onClick: () => this.toggleCaptions(), icon: captionsEnabled ? this.iconPack.captionsOff : this.iconPack.captionsOn, label: captionsEnabled ? this.t('transcript.off') : this.t('transcript.on'), variant: this.variant })));
  }
  static get watchers() { return {
    "meeting": ["meetingChanged"]
  }; }
};
DyteCaptionToggle.style = dyteCaptionToggleCss;

exports.dyte_caption_toggle = DyteCaptionToggle;
