// @dyteinternals/utils/dist/types/HybridPreset.js
export var ViewType;
(function (ViewType) {
    ViewType["GroupCall"] = "GROUP_CALL";
    ViewType["Webinar"] = "WEBINAR";
    ViewType["AudioRoom"] = "AUDIO_ROOM";
    ViewType["Livestream"] = "LIVESTREAM";
    ViewType["Chat"] = "CHAT";
})(ViewType || (ViewType = {}));
export var WaitingRoomTypes;
(function (WaitingRoomTypes) {
    WaitingRoomTypes["Skip"] = "SKIP";
    WaitingRoomTypes["OnPrivilegedUserEntry"] = "ON_PRIVILEGED_USER_ENTRY";
    WaitingRoomTypes["SkipOnAccept"] = "SKIP_ON_ACCEPT";
})(WaitingRoomTypes || (WaitingRoomTypes = {}));
export var MediaProductionPermissionType;
(function (MediaProductionPermissionType) {
    MediaProductionPermissionType["Allowed"] = "ALLOWED";
    MediaProductionPermissionType["NotAllowed"] = "NOT_ALLOWED";
    MediaProductionPermissionType["CanRequest"] = "CAN_REQUEST";
})(MediaProductionPermissionType || (MediaProductionPermissionType = {}));
export var MediaConsumptionPermissionType;
(function (MediaConsumptionPermissionType) {
    MediaConsumptionPermissionType["Allowed"] = "ALLOWED";
    MediaConsumptionPermissionType["NotAllowed"] = "NOT_ALLOWED";
})(MediaConsumptionPermissionType || (MediaConsumptionPermissionType = {}));
export var PluginAccessControls;
(function (PluginAccessControls) {
    PluginAccessControls["FULL_ACCESS"] = "FULL_ACCESS";
    PluginAccessControls["VIEW_ONLY"] = "VIEW_ONLY";
})(PluginAccessControls || (PluginAccessControls = {}));
