// @dyteinternals/utils/dist/types/index.js
import { WaitingRoomTypes } from './HybridPreset';
export var WaitingRoomTypesV1;
(function (WaitingRoomTypesV1) {
    WaitingRoomTypesV1["skip"] = "SKIP";
    WaitingRoomTypesV1["onAccept"] = "ON_ACCEPT";
    WaitingRoomTypesV1["skipOnPrivilegedUserEntry"] = "SKIP_ON_PRIVILEGED_USER_ENTRY";
    WaitingRoomTypesV1["skipOnAccept"] = "SKIP_ON_ACCEPT"; // should be removed later
})(WaitingRoomTypesV1 || (WaitingRoomTypesV1 = {}));
export var RecorderType;
(function (RecorderType) {
    RecorderType["none"] = "NONE";
    RecorderType["recorder"] = "RECORDER";
    RecorderType["livestreamer"] = "LIVESTREAMER";
})(RecorderType || (RecorderType = {}));
export var DyteViewTypes;
(function (DyteViewTypes) {
    DyteViewTypes["groupCall"] = "GROUP_CALL";
    DyteViewTypes["webinar"] = "WEBINAR";
    DyteViewTypes["audioRoom"] = "AUDIO_ROOM";
    DyteViewTypes["livestream"] = "LIVESTREAM";
    DyteViewTypes["chat"] = "CHAT";
})(DyteViewTypes || (DyteViewTypes = {}));
export var WaitingRoomConversionToV1Type;
(function (WaitingRoomConversionToV1Type) {
    WaitingRoomConversionToV1Type["SKIP"] = "skip";
    WaitingRoomConversionToV1Type["SKIP_ON_PRIVILEGED_USER_ENTRY"] = "skipOnPrivilegedUserEntry";
    WaitingRoomConversionToV1Type["ON_ACCEPT"] = "skipOnAccept";
    WaitingRoomConversionToV1Type["SKIP_ON_ACCEPT"] = "skipOnAccept"; // TODO: should be removed later
})(WaitingRoomConversionToV1Type || (WaitingRoomConversionToV1Type = {}));
export { WaitingRoomTypes };
