// @dytesdk/ui-kit/dist/components/dyte-chat-search-results2.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import './store.js';
import './breakout-rooms-manager.js';
import './livestream.js';
import { d as defaultIconPack } from './default-icon-pack.js';
import { u as useLanguage } from './index2.js';
import './user-prefs.js';
import { d as defineCustomElement$c } from './dyte-avatar2.js';
import { d as defineCustomElement$b } from './dyte-button2.js';
import { d as defineCustomElement$a } from './dyte-chat-message2.js';
import { d as defineCustomElement$9 } from './dyte-file-message2.js';
import { d as defineCustomElement$8 } from './dyte-icon2.js';
import { d as defineCustomElement$7 } from './dyte-image-message2.js';
import { d as defineCustomElement$6 } from './dyte-menu2.js';
import { d as defineCustomElement$5 } from './dyte-menu-item2.js';
import { d as defineCustomElement$4 } from './dyte-menu-list2.js';
import { d as defineCustomElement$3 } from './dyte-paginated-list2.js';
import { d as defineCustomElement$2 } from './dyte-spinner2.js';
import { d as defineCustomElement$1 } from './dyte-text-message2.js';

const dyteChatSearchResultsCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{display:flex;height:100%;flex-direction:column;position:relative;--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-900, 26 26 26) / var(--tw-bg-opacity))}";

const DyteChatSearchResults = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.pageSize = 50;
    this.searchMessages = async (timestamp, size, reversed) => {
      return this.meeting.chat.searchMessages(this.query, {
        channelId: this.channelId,
        timestamp,
        size,
        reversed,
      });
    };
    this.nodeRenderer = (messages) => {
      return messages.map((message) => (h("dyte-chat-message", { key: message.id, message: message, disableControls: true })));
    };
    this.meeting = undefined;
    this.query = undefined;
    this.channelId = undefined;
    this.iconPack = defaultIconPack;
    this.t = useLanguage();
  }
  render() {
    return (h(Host, null, h("dyte-paginated-list", { pageSize: this.pageSize, pagesAllowed: 3, fetchData: this.searchMessages, createNodes: this.nodeRenderer, selectedItemId: this.query })));
  }
  static get style() { return dyteChatSearchResultsCss; }
}, [1, "dyte-chat-search-results", {
    "meeting": [16],
    "query": [1],
    "channelId": [1, "channel-id"],
    "iconPack": [16],
    "t": [16]
  }]);
function defineCustomElement() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["dyte-chat-search-results", "dyte-avatar", "dyte-button", "dyte-chat-message", "dyte-file-message", "dyte-icon", "dyte-image-message", "dyte-menu", "dyte-menu-item", "dyte-menu-list", "dyte-paginated-list", "dyte-spinner", "dyte-text-message"];
  components.forEach(tagName => { switch (tagName) {
    case "dyte-chat-search-results":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DyteChatSearchResults);
      }
      break;
    case "dyte-avatar":
      if (!customElements.get(tagName)) {
        defineCustomElement$c();
      }
      break;
    case "dyte-button":
      if (!customElements.get(tagName)) {
        defineCustomElement$b();
      }
      break;
    case "dyte-chat-message":
      if (!customElements.get(tagName)) {
        defineCustomElement$a();
      }
      break;
    case "dyte-file-message":
      if (!customElements.get(tagName)) {
        defineCustomElement$9();
      }
      break;
    case "dyte-icon":
      if (!customElements.get(tagName)) {
        defineCustomElement$8();
      }
      break;
    case "dyte-image-message":
      if (!customElements.get(tagName)) {
        defineCustomElement$7();
      }
      break;
    case "dyte-menu":
      if (!customElements.get(tagName)) {
        defineCustomElement$6();
      }
      break;
    case "dyte-menu-item":
      if (!customElements.get(tagName)) {
        defineCustomElement$5();
      }
      break;
    case "dyte-menu-list":
      if (!customElements.get(tagName)) {
        defineCustomElement$4();
      }
      break;
    case "dyte-paginated-list":
      if (!customElements.get(tagName)) {
        defineCustomElement$3();
      }
      break;
    case "dyte-spinner":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
    case "dyte-text-message":
      if (!customElements.get(tagName)) {
        defineCustomElement$1();
      }
      break;
  } });
}

export { DyteChatSearchResults as D, defineCustomElement as d };
