// @dytesdk/ui-kit/dist/components/dyte-waiting-screen.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { u as useLanguage } from './index2.js';
import { d as defaultIconPack } from './default-icon-pack.js';
import './store.js';
import './breakout-rooms-manager.js';
import { d as defaultConfig } from './default-ui-config.js';
import './livestream.js';
import './user-prefs.js';
import { d as defineCustomElement$2 } from './dyte-logo2.js';

const dyteWaitingScreenCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{height:100%;width:100%;display:flex;flex-direction:column;align-items:center;justify-content:center}.centered{display:flex;flex-direction:column;align-items:center}dyte-logo{margin-bottom:var(--dyte-space-8, 32px);height:var(--dyte-space-12, 48px)}p{font-size:16px;border-radius:var(--dyte-border-radius-lg, 12px);--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-800, 30 30 30) / var(--tw-bg-opacity));padding-left:var(--dyte-space-8, 32px);padding-right:var(--dyte-space-8, 32px);padding-top:var(--dyte-space-4, 16px);padding-bottom:var(--dyte-space-4, 16px);color:rgb(var(--dyte-colors-text-1000, 255 255 255))}";

const DyteWaitingScreen$1 = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.meeting = undefined;
    this.config = defaultConfig;
    this.iconPack = defaultIconPack;
    this.t = useLanguage();
  }
  render() {
    return (h(Host, null, h("slot", null, h("div", { class: "centered", part: "content" }, h("dyte-logo", { meeting: this.meeting, config: this.config, part: "logo", t: this.t }), h("p", null, this.t('waitlist.body_text'))))));
  }
  static get style() { return dyteWaitingScreenCss; }
}, [1, "dyte-waiting-screen", {
    "meeting": [16],
    "config": [16],
    "iconPack": [16],
    "t": [16]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["dyte-waiting-screen", "dyte-logo"];
  components.forEach(tagName => { switch (tagName) {
    case "dyte-waiting-screen":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DyteWaitingScreen$1);
      }
      break;
    case "dyte-logo":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const DyteWaitingScreen = DyteWaitingScreen$1;
const defineCustomElement = defineCustomElement$1;

export { DyteWaitingScreen, defineCustomElement };
