// @dytesdk/ui-kit/dist/components/dyte-tooltip2.js
import { proxyCustomElement, HTMLElement, createEvent, writeTask, h, Host } from '@stencil/core/internal/client';
import { c as computePosition, o as offset, f as flip, s as shift, a as arrow } from './floating-ui.dom.esm.js';
import { d as defaultIconPack } from './default-icon-pack.js';
import { u as useLanguage } from './index2.js';

const dyteTooltipCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{--background-color:var(--dyte-tooltip-background-color, rgb(var(--dyte-colors-background-600, 60 60 60)));--color:var(--dyte-tooltip-color, rgb(var(--dyte-colors-text-1000, 255 255 255)));display:inline-flex}#trigger{display:block;width:100%;flex:1 1 0%}.tooltip{max-width:var(--dyte-space-64, 256px);position:fixed;z-index:20;display:none;width:-moz-max-content;width:max-content;padding-left:var(--dyte-space-2, 8px);padding-right:var(--dyte-space-2, 8px);padding-top:var(--dyte-space-1, 4px);padding-bottom:var(--dyte-space-1, 4px);border-radius:var(--dyte-border-radius-sm, 4px);--tw-shadow:0 1px 2px 0 rgb(0 0 0 / 0.05);--tw-shadow-colored:0 1px 2px 0 var(--tw-shadow-color);box-shadow:var(--tw-ring-offset-shadow, 0 0 #0000), var(--tw-ring-shadow, 0 0 #0000), var(--tw-shadow);font-size:12px}.tooltip,#arrow{position:absolute;background-color:var(--background-color);color:var(--color)}#arrow{position:absolute;height:var(--dyte-space-2, 8px);width:var(--dyte-space-2, 8px);transform:rotate(45deg)}:host([variant='primary']){--background-color:rgb(var(--dyte-colors-brand-500, 33 96 253));--color:rgb(var(--dyte-colors-text-on-brand-1000, var(--dyte-colors-text-1000, 255 255 255)))}:host([kind='block']){display:block}";

const DyteMenu = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.dyteOpenChange = createEvent(this, "dyteOpenChange", 7);
    this.showMenu = () => {
      if (this.disabled)
        return;
      this.isInFocus = true;
      setTimeout(() => {
        if (this.isInFocus) {
          this.tooltipEl.style.display = 'block';
          this.update();
          this.dyteOpenChange.emit(true);
          if (this.size === 'sm') {
            setTimeout(() => {
              if (this.isInFocus) {
                this.hideMenu();
              }
            }, 1000);
          }
        }
      }, this.delay);
    };
    this.hideMenu = () => {
      if (this.open || this.disabled)
        return;
      this.isInFocus = false;
      this.tooltipEl.style.display = 'none';
      this.dyteOpenChange.emit(false);
    };
    this.label = '';
    this.variant = 'secondary';
    this.disabled = false;
    this.open = false;
    this.kind = 'inline';
    this.size = undefined;
    this.placement = 'top';
    this.delay = 0;
    this.iconPack = defaultIconPack;
    this.t = useLanguage();
    this.isInFocus = false;
  }
  componentDidLoad() {
    this.triggerEl.addEventListener('focusin', this.showMenu);
    this.triggerEl.addEventListener('mouseenter', this.showMenu);
    this.triggerEl.addEventListener('focusout', this.hideMenu);
    this.triggerEl.addEventListener('mouseleave', this.hideMenu);
    writeTask(() => {
      this.openChanged(this.open);
    });
  }
  disconnectedCallback() {
    if (!this.triggerEl)
      return;
    this.triggerEl.removeEventListener('focusin', this.showMenu);
    this.triggerEl.removeEventListener('mouseenter', this.showMenu);
    this.triggerEl.removeEventListener('focusout', this.hideMenu);
    this.triggerEl.removeEventListener('mouseleave', this.hideMenu);
    this.triggerEl = undefined;
  }
  openChanged(open) {
    if (open) {
      this.showMenu();
    }
    else {
      this.hideMenu();
    }
  }
  update() {
    computePosition(this.triggerEl, this.tooltipEl, {
      placement: this.placement,
      middleware: [offset(8), flip(), shift({ padding: 5 }), arrow({ element: this.arrowEl })],
    }).then(({ x, y, placement, middlewareData }) => {
      Object.assign(this.tooltipEl.style, {
        left: `${x}px`,
        top: `${y}px`,
      });
      const { x: arrowX, y: arrowY } = middlewareData.arrow;
      const staticSide = {
        top: 'bottom',
        right: 'left',
        bottom: 'top',
        left: 'right',
      }[placement.split('-')[0]];
      Object.assign(this.arrowEl.style, {
        left: arrowX != null ? `${arrowX}px` : '',
        top: arrowY != null ? `${arrowY}px` : '',
        right: '',
        bottom: '',
        [staticSide]: '-4px',
      });
    });
  }
  render() {
    return (h(Host, null, h("span", { part: "trigger", id: "trigger", ref: (el) => (this.triggerEl = el) }, h("slot", null)), h("div", { part: "tooltip", class: "tooltip", id: "tooltip", role: "tooltip", ref: (el) => (this.tooltipEl = el) }, h("div", { id: "arrow", ref: (el) => (this.arrowEl = el), part: "arrow" }), this.label, h("slot", { name: "tooltip" }))));
  }
  static get watchers() { return {
    "open": ["openChanged"]
  }; }
  static get style() { return dyteTooltipCss; }
}, [1, "dyte-tooltip", {
    "label": [1],
    "variant": [513],
    "disabled": [516],
    "open": [516],
    "kind": [513],
    "size": [513],
    "placement": [1],
    "delay": [2],
    "iconPack": [16],
    "t": [16],
    "isInFocus": [32]
  }]);
function defineCustomElement() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["dyte-tooltip"];
  components.forEach(tagName => { switch (tagName) {
    case "dyte-tooltip":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DyteMenu);
      }
      break;
  } });
}

export { DyteMenu as D, defineCustomElement as d };
