// @dytesdk/ui-kit/dist/components/dyte-name-tag2.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { s as shorten, f as formatName } from './string.js';
import { u as useLanguage } from './index2.js';
import { d as defaultIconPack } from './default-icon-pack.js';

const dyteNameTagCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{display:inline-flex;-webkit-user-select:none;-moz-user-select:none;user-select:none;align-items:center;padding-top:var(--dyte-space-1, 4px);padding-bottom:var(--dyte-space-1, 4px);padding-left:var(--dyte-space-1\\.5, 6px);padding-right:var(--dyte-space-1\\.5, 6px);font-size:14px;border-radius:var(--dyte-border-radius-sm, 4px);--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-700, 44 44 44) / var(--tw-bg-opacity));color:rgb(var(--dyte-colors-text-900, 255 255 255 / 0.88))}span.name{overflow:hidden;display:-webkit-box;-webkit-box-orient:vertical;-webkit-line-clamp:1}::slotted(dyte-audio-visualizer[slot='start']){margin-right:var(--dyte-space-1\\.5, 6px)}::slotted(dyte-audio-visualizer[slot='end']){margin-left:var(--dyte-space-1\\.5, 6px)}:host([size='sm']){font-size:12px;--tw-bg-opacity:0.6}:host([variant='text']){background-color:transparent;padding:var(--dyte-space-0, 0px)}";

const DyteNameTag = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.participant = undefined;
    this.meeting = undefined;
    this.size = undefined;
    this.maxLength = 20;
    this.isScreenShare = false;
    this.variant = 'default';
    this.iconPack = defaultIconPack;
    this.t = useLanguage();
    this.length = 13;
  }
  sizeChanged(size) {
    this.length = size === 'sm' ? 8 : 13;
  }
  formatNameTag(name, isSelf) {
    return !this.isScreenShare
      ? isSelf
        ? `${shorten(name, this.length - 3)} (${this.t('you')})`
        : shorten(name, this.length)
      : isSelf
        ? `${this.t('screen')} - ${shorten(name, this.length - 3)} (${this.t('you')})`
        : `${this.t('screen')} - ${shorten(name, this.length)}`;
  }
  render() {
    var _a, _b, _c;
    const name = formatName(((_a = this.participant) === null || _a === void 0 ? void 0 : _a.name) || '');
    const isSelf = ((_b = this.participant) === null || _b === void 0 ? void 0 : _b.id) === ((_c = this.meeting) === null || _c === void 0 ? void 0 : _c.self.id);
    return (h(Host, { title: name }, h("slot", { name: "start" }), h("span", { class: "name" }, this.formatNameTag(name, isSelf)), h("slot", { name: "end" })));
  }
  static get watchers() { return {
    "size": ["sizeChanged"]
  }; }
  static get style() { return dyteNameTagCss; }
}, [1, "dyte-name-tag", {
    "participant": [16],
    "meeting": [16],
    "size": [513],
    "maxLength": [2, "max-length"],
    "isScreenShare": [4, "is-screen-share"],
    "variant": [513],
    "iconPack": [16],
    "t": [16],
    "length": [32]
  }]);
function defineCustomElement() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["dyte-name-tag"];
  components.forEach(tagName => { switch (tagName) {
    case "dyte-name-tag":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DyteNameTag);
      }
      break;
  } });
}

export { DyteNameTag as D, defineCustomElement as d };
