// @dytesdk/ui-kit/dist/components/dyte-mute-all-confirmation.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client';
import { u as useLanguage } from './index2.js';
import { d as defaultIconPack } from './default-icon-pack.js';
import { s as state } from './store.js';
import { d as defineCustomElement$2 } from './dyte-button2.js';

const dyteMuteAllConfirmationCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{display:block;-webkit-user-select:none;-moz-user-select:none;user-select:none}.leave-modal{position:relative;display:flex;flex-direction:column;border-radius:var(--dyte-border-radius-md, 8px);--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-900, 26 26 26) / var(--tw-bg-opacity));padding:var(--dyte-space-4, 16px);color:rgb(var(--dyte-colors-text-1000, 255 255 255))}.leave-modal .header h2{margin:var(--dyte-space-0, 0px);margin-bottom:var(--dyte-space-3, 12px)}.leave-modal .content{font-size:14px}.leave-message p{margin-top:var(--dyte-space-3, 12px);margin-bottom:var(--dyte-space-3, 12px)}.leave-meeting{display:flex;flex-direction:row;justify-content:space-between;gap:var(--dyte-space-2, 8px)}.leave-meeting dyte-button{flex:1 1 0%}.end-meeting{margin-top:var(--dyte-space-2, 8px)}.message{margin-bottom:var(--dyte-space-4, 16px)}label{display:flex;align-items:center}input[type='checkbox']{margin-right:var(--dyte-space-2, 8px);accent-color:rgb(var(--dyte-colors-brand-500, 33 96 253))}";

const DyteMuteAllConfirmation$1 = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.stateUpdate = createEvent(this, "dyteStateUpdate", 7);
    this.onClose = () => {
      this.stateUpdate.emit({ activeMuteAllConfirmation: false });
      state.activeMuteAllConfirmation = false;
    };
    this.onMuteAll = () => {
      var _a;
      (_a = this.meeting) === null || _a === void 0 ? void 0 : _a.participants.disableAllAudio(this.allowUnmute);
      this.onClose();
    };
    this.meeting = undefined;
    this.states = undefined;
    this.iconPack = defaultIconPack;
    this.t = useLanguage();
    this.allowUnmute = true;
  }
  render() {
    return (h(Host, null, h("div", { class: "leave-modal" }, h("div", { class: "header" }, h("h2", { class: "title" }, this.t('mute_all.header'))), h("p", { class: "message" }, this.t('mute_all.description')), h("div", { class: "content" }, h("div", { class: "leave-meeting" }, h("dyte-button", { variant: "secondary", title: this.t('close'), onClick: this.onClose, iconPack: this.iconPack, t: this.t }, this.t('cancel')), h("dyte-button", { variant: "danger", title: this.t('mute_all'), onClick: this.onMuteAll, iconPack: this.iconPack, t: this.t }, this.t('mute_all')))))));
  }
  static get style() { return dyteMuteAllConfirmationCss; }
}, [1, "dyte-mute-all-confirmation", {
    "meeting": [16],
    "states": [16],
    "iconPack": [16],
    "t": [16],
    "allowUnmute": [32]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["dyte-mute-all-confirmation", "dyte-button"];
  components.forEach(tagName => { switch (tagName) {
    case "dyte-mute-all-confirmation":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DyteMuteAllConfirmation$1);
      }
      break;
    case "dyte-button":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const DyteMuteAllConfirmation = DyteMuteAllConfirmation$1;
const defineCustomElement = defineCustomElement$1;

export { DyteMuteAllConfirmation, defineCustomElement };
