// @dytesdk/ui-kit/dist/components/dyte-debugger-audio2.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client';
import { s as state } from './store.js';
import './breakout-rooms-manager.js';
import './livestream.js';
import { d as defaultIconPack } from './default-icon-pack.js';
import { u as useLanguage } from './index2.js';
import './user-prefs.js';
import { d as defineCustomElement$5 } from './dyte-button2.js';
import { d as defineCustomElement$4 } from './dyte-debugger-issues-ui2.js';
import { d as defineCustomElement$3 } from './dyte-debugger-permissions-ui2.js';
import { d as defineCustomElement$2 } from './dyte-icon2.js';
import { d as defineCustomElement$1 } from './dyte-information-tooltip2.js';

const dyteDebuggerAudioCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{position:relative;display:flex;flex-direction:column;overflow:hidden;height:100%;width:100%;color:rgb(var(--dyte-colors-text-1000, 255 255 255))}h3{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px);font-size:16px;font-weight:600}.status{margin-left:var(--dyte-space-2, 8px);display:flex;flex-direction:row;align-items:center;justify-content:center;font-size:12px}.status dyte-icon{margin-right:var(--dyte-space-2, 8px)}.scroll{overflow:auto;border-radius:var(--dyte-border-radius-none, 0)}.warning,.NOT_REQUESTED{--tw-text-opacity:1 !important;color:rgba(var(--dyte-colors-warning, 255 205 7) / var(--tw-text-opacity)) !important;opacity:0.8}.success,.ACCEPTED{--tw-text-opacity:1 !important;color:rgba(var(--dyte-colors-success, 98 165 4) / var(--tw-text-opacity)) !important;opacity:0.8}.COULD_NOT_START,.error,.failed,.DENIED,.SYSTEM_DENIED{--tw-text-opacity:1 !important;color:rgba(var(--dyte-colors-danger, 255 45 45) / var(--tw-text-opacity)) !important;opacity:0.8}.row{display:flex;flex-direction:row;align-items:center;justify-content:flex-start}.section{display:flex;flex-direction:column;align-items:flex-start;justify-content:flex-start}h3{margin:var(--dyte-space-0, 0px);font-size:14px;font-weight:600}.stage-indicator{margin-bottom:var(--dyte-space-1, 4px);display:flex;flex-direction:row;align-items:center;font-size:12px;color:rgb(var(--dyte-colors-text-600, 255 255 255 / 0.52))}.status-text{margin-top:var(--dyte-space-3, 12px);font-size:14px;font-weight:100 !important}.status-text dyte-icon{margin-right:var(--dyte-space-2, 8px)}.tooltip-text{display:flex;flex-direction:column;font-size:12px}.tooltip-text span{margin-left:var(--dyte-space-1, 4px);color:rgb(var(--dyte-colors-text-900, 255 255 255 / 0.88)) !important}.tooltip-text div{margin-top:var(--dyte-space-1, 4px)}.tooltip-text div:nth-child(1){margin:var(--dyte-space-0, 0px)}.row{display:flex;flex-direction:row;align-items:center;justify-content:center}.row dyte-switch{margin-left:var(--dyte-space-8, 32px)}.row dyte-button{margin-top:var(--dyte-space-1, 4px);margin-bottom:var(--dyte-space-1, 4px);margin-left:var(--dyte-space-2, 8px)}.row .label{width:var(--dyte-space-40, 160px)}.row .text{margin-top:var(--dyte-space-1, 4px);margin-left:var(--dyte-space-2, 8px);margin-bottom:var(--dyte-space-0, 0px)}.m3{margin-top:var(--dyte-space-3, 12px)}.m1{margin-top:var(--dyte-space-1, 4px)}.col{display:flex;width:100%;flex-direction:column;align-items:flex-start;justify-content:flex-start}.label{margin-top:var(--dyte-space-1, 4px);display:flex;flex-direction:row;justify-content:space-between;font-size:14px;font-weight:100 !important;color:rgb(var(--dyte-colors-text-700, 255 255 255 / 0.64))}.text{margin-bottom:var(--dyte-space-2, 8px);font-size:14px;font-weight:100 !important;color:rgb(var(--dyte-colors-text-700, 255 255 255 / 0.64));opacity:0.7}.title{margin-bottom:var(--dyte-space-2, 8px);display:flex;flex-direction:row;align-items:center;font-size:14px;font-weight:600;color:rgb(var(--dyte-colors-text-900, 255 255 255 / 0.88))}.sub-title{margin-top:var(--dyte-space-4, 16px);margin-bottom:var(--dyte-space-2, 8px);display:flex;flex-direction:row;align-items:center;font-size:14px;font-weight:600;color:rgb(var(--dyte-colors-text-900, 255 255 255 / 0.88))}.error-text{margin-top:var(--dyte-space-1, 4px);display:flex;flex-direction:row;font-size:14px;font-weight:100 !important;--tw-text-opacity:1;color:rgba(var(--dyte-colors-danger, 255 45 45) / var(--tw-text-opacity))}.error-text dyte-icon{margin-right:var(--dyte-space-2, 8px)}.volume-indicator{margin-left:var(--dyte-space-4, 16px);height:var(--dyte-space-3, 12px);width:var(--dyte-space-56, 224px);overflow:hidden;border-radius:var(--dyte-border-radius-lg, 12px);--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-1000, 8 8 8) / var(--tw-bg-opacity))}.volume-level{height:var(--dyte-space-3, 12px);width:var(--dyte-space-2, 8px);--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-brand-400, 53 110 253) / var(--tw-bg-opacity));transition:width 0.5s}.stage-manager{position:sticky;bottom:var(--dyte-space-0, 0px);margin-top:var(--dyte-space-4, 16px);display:flex;flex-direction:row-reverse;align-items:center;justify-content:space-between;font-size:14px;--tw-text-opacity:1;color:rgba(var(--dyte-colors-brand-600, 13 81 253) / var(--tw-text-opacity))}.stage-button{display:flex;cursor:pointer;flex-direction:row;align-items:center;border-radius:var(--dyte-border-radius-sm, 4px);--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-700, 44 44 44) / var(--tw-bg-opacity));padding-top:var(--dyte-space-1, 4px);padding-bottom:var(--dyte-space-1, 4px);padding-left:var(--dyte-space-2, 8px);padding-right:var(--dyte-space-2, 8px)}.stage-button dyte-icon{padding:var(--dyte-space-0, 0px)}.disabled{cursor:not-allowed;color:rgb(var(--dyte-colors-text-600, 255 255 255 / 0.52)) !important}#header{margin-top:var(--dyte-space-2, 8px);display:flex;align-items:center;justify-content:space-evenly;border-bottom:var(--dyte-border-width-sm, 1px) solid rgb(var(--dyte-colors-background-600, 60 60 60))}#header dyte-button{border-bottom:var(--dyte-border-width-md, 2px) solid transparent;border-radius:var(--dyte-border-radius-none, 0)}#header dyte-button:hover{background-color:transparent}#header dyte-button.active{border-color:rgb(var(--dyte-colors-brand-400, 53 110 253))}.tab-body{display:flex;height:100%;flex-direction:column;justify-content:space-between;overflow-y:auto;padding:var(--dyte-space-4, 16px)}.speaker-test{margin-top:var(--dyte-space-4, 16px);display:flex;flex-direction:row}.speaker-test dyte-button{margin-right:var(--dyte-space-2, 8px)}";

const DyteDebuggerAudio = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.testProgress = createEvent(this, "testProgress", 7);
    this.lowVolCount = 0;
    this.silenceCount = 0;
    this.meeting = undefined;
    this.states = undefined;
    this.size = undefined;
    this.iconPack = defaultIconPack;
    this.t = useLanguage();
    this.micMetadata = {};
    this.micPreview = false;
    this.speakerPreview = false;
    this.audioTrackStats = {
      lowVolume: false,
      clip: false,
      silence: false,
      volume: 0,
      timestamp: 0,
    };
    this.activeTab = 'Report';
    this.stage = 0;
    this.mediaPermission = 'NOT_REQUESTED';
    this.speakerDevice = undefined;
    this.speakerTest = 'none';
  }
  connectedCallback() {
    this.meetingChanged(this.meeting);
    this.updateProgress();
  }
  disconnectedCallback() {
    var _a;
    const { self, troubleshoot } = this.meeting;
    this.micPreview && this.toggleMicTest();
    this.speakerPreview && this.toggleSpeakerTest();
    self === null || self === void 0 ? void 0 : self.removeListener('deviceUpdate', this.deviceUpdateListener);
    self === null || self === void 0 ? void 0 : self.removeListener('mediaPermissionUpdate', this.mediaPermissionUpdateListener);
    (_a = troubleshoot === null || troubleshoot === void 0 ? void 0 : troubleshoot.audio) === null || _a === void 0 ? void 0 : _a.removeListener('audioTrackStats', this.audioTrackUpdateListener);
  }
  async meetingChanged(meeting) {
    var _a, _b;
    if (!meeting)
      return;
    const { self, troubleshoot } = meeting;
    if (self === null || self === void 0 ? void 0 : self.audioTrack) {
      await this.getTrackInfo();
    }
    self === null || self === void 0 ? void 0 : self.on('deviceUpdate', this.deviceUpdateListener.bind(this));
    self.addListener('mediaPermissionUpdate', this.mediaPermissionUpdateListener.bind(this));
    const permission = (_a = self === null || self === void 0 ? void 0 : self.mediaPermissions) === null || _a === void 0 ? void 0 : _a.audio;
    this.mediaPermissionUpdateListener({ kind: 'audio', message: permission });
    (_b = troubleshoot === null || troubleshoot === void 0 ? void 0 : troubleshoot.audio) === null || _b === void 0 ? void 0 : _b.on('audioTrackStats', this.audioTrackUpdateListener.bind(this));
  }
  deviceUpdateListener({ device, preview }) {
    if (preview)
      return;
    if (device.kind === 'audioInput')
      this.speakerDevice = device;
    if (device.kind === 'audiooutput')
      this.getTrackInfo();
  }
  mediaPermissionUpdateListener({ kind, message }) {
    if (kind !== 'audio')
      return;
    this.mediaPermission = message;
    if (this.mediaPermission !== 'ACCEPTED') {
      this.stage = 0;
      this.updateProgress();
      if (this.micPreview)
        this.toggleMicTest();
      if (this.speakerPreview)
        this.toggleSpeakerTest();
    }
  }
  audioTrackUpdateListener(data) {
    this.audioTrackStats = Object.assign(Object.assign({}, data), { volume: this.normalizeVolume(data === null || data === void 0 ? void 0 : data.volume) });
    if (this.volumeEl)
      this.volumeEl.style.width = `${this.audioTrackStats.volume}%`;
  }
  setActiveTab(state) {
    this.activeTab = state;
    this.updateProgress();
  }
  async toggleMicTest() {
    var _a, _b, _c;
    const troubleshooter = (_a = this.meeting.troubleshoot) === null || _a === void 0 ? void 0 : _a.audio;
    this.micPreview = !this.micPreview;
    if (this.micPreview) {
      await this.getTrackInfo();
      troubleshooter.stopTrackAnalysis();
      await (troubleshooter === null || troubleshooter === void 0 ? void 0 : troubleshooter.startTrackAnalysis(true));
    }
    else {
      (_c = (_b = this.meeting.troubleshoot) === null || _b === void 0 ? void 0 : _b.audio) === null || _c === void 0 ? void 0 : _c.stopTrackAnalysis();
      this.audioTrackStats.volume = 0;
      if (this.volumeEl)
        this.volumeEl.style.width = '5%';
    }
  }
  async toggleSpeakerTest() {
    var _a, _b, _c, _d;
    const troubleshooter = (_a = this.meeting.troubleshoot) === null || _a === void 0 ? void 0 : _a.audio;
    if ((_b = this.testAudioEl) === null || _b === void 0 ? void 0 : _b.paused) {
      this.testAudioEl.currentTime = 0.2;
      (_c = this.testAudioEl) === null || _c === void 0 ? void 0 : _c.play();
      this.speakerPreview = true;
      troubleshooter === null || troubleshooter === void 0 ? void 0 : troubleshooter.startTrackAnalysis(true);
      this.testAudioEl.addEventListener('ended', () => {
        this.speakerPreview = false;
        troubleshooter === null || troubleshooter === void 0 ? void 0 : troubleshooter.stopTrackAnalysis();
      });
    }
    else {
      (_d = this.testAudioEl) === null || _d === void 0 ? void 0 : _d.pause();
      this.speakerPreview = false;
      troubleshooter === null || troubleshooter === void 0 ? void 0 : troubleshooter.stopTrackAnalysis();
    }
    if (this.speakerPreview) {
      const { self } = this.meeting;
      const speakerDevices = await (self === null || self === void 0 ? void 0 : self.getSpeakerDevices());
      if (!speakerDevices)
        return;
      this.speakerDevice = speakerDevices[0];
      self === null || self === void 0 ? void 0 : self.setDevice(speakerDevices[0]);
    }
  }
  async getTrackInfo() {
    var _a, _b;
    this.micMetadata = await ((_b = (_a = this.meeting.troubleshoot) === null || _a === void 0 ? void 0 : _a.audio) === null || _b === void 0 ? void 0 : _b.getTrackMetadata(true));
  }
  changeStage(stage) {
    if (this.mediaPermission !== 'ACCEPTED')
      return;
    stage = Math.max(0, stage);
    stage = Math.min(2, stage);
    this.stage = stage;
    this.updateProgress();
    if (this.micPreview)
      this.toggleMicTest();
    if (this.speakerPreview)
      this.toggleSpeakerTest();
  }
  updateProgress() {
    if (this.activeTab === 'Troubleshoot') {
      this.testProgress.emit(((this.stage + 1) * 100) / 3);
    }
    else
      this.testProgress.emit(0);
  }
  getAudioLevelStatus() {
    let message = 'All systems are functional.';
    let icon = 'checkmark';
    let style = 'success';
    if (this.detectLowVolume()) {
      message = 'Low volume detected, please move closer to the device.';
      icon = 'warning';
      style = 'warning';
    }
    if (this.audioTrackStats.clip) {
      message = 'Clipping detected. Your audio may be noisy.';
      icon = 'warning';
      style = 'error';
    }
    if (this.detectSilence()) {
      message = 'Silence detected. Your device is not picking up audio.';
      icon = 'warning';
      style = 'error';
    }
    return (h("div", { class: `row status-text ${style}` }, h("dyte-icon", { size: "sm", icon: this.iconPack[icon] }), this.t(message)));
  }
  getDeviceInformationStatus() {
    let message = 'All systems are functional.';
    let icon = 'checkmark';
    let style = 'success';
    const { channelCount: count, channelInterpretation: mode, channelType: type, outputs, } = this.micMetadata;
    if (mode === 'discrete' && outputs > count) {
      message = 'Audio quality might be degraded.';
      icon = 'warning';
      style = 'warning';
    }
    else if ((type === 'STEREO' && outputs === 1) ||
      (type === 'QUAD' && outputs === 1) ||
      (type === 'QUAD' && outputs === 2)) {
      message = 'Audio quality might be degraded.';
      icon = 'warning';
      style = 'warning';
    }
    if ((type === 'MONO' && count < 1) ||
      (type === 'STEREO' && count < 2) ||
      (type === 'QUAD' && count < 4)) {
      message = 'Looks like your device is not functioning properly.';
      icon = 'warning';
      style = 'error';
    }
    return (h("div", { class: `row status-text ${style}` }, h("dyte-icon", { size: "sm", icon: this.iconPack[icon] }), this.t(message)));
  }
  getSpeakerTestStatus() {
    if (this.speakerTest === 'success')
      return 'All systems are functional.';
    if (this.speakerTest === 'failed')
      return 'Test failed. Please ensure your system volume is not 0.';
  }
  setSpeakerTestResult(status) {
    if (this.speakerPreview)
      this.toggleSpeakerTest();
    this.speakerTest = status;
  }
  boolToStr(val) {
    if (val)
      return 'Yes';
    return 'No';
  }
  formatLabel(camelCaseLabel) {
    const result = camelCaseLabel.replace(/([A-Z])/g, ' $1');
    return result.charAt(0).toUpperCase() + result.slice(1);
  }
  normalizeVolume(val = 0) {
    let max = 0.3;
    const min = 0.001;
    if (val > max)
      max = val;
    return Math.max(5, Math.round((val * 100) / (max - min)));
  }
  detectLowVolume() {
    var _a;
    if ((_a = this.audioTrackStats) === null || _a === void 0 ? void 0 : _a.lowVolume) {
      this.lowVolCount++;
    }
    else {
      this.lowVolCount = 0;
    }
    return this.lowVolCount > 10;
  }
  detectSilence() {
    var _a;
    if ((_a = this.audioTrackStats) === null || _a === void 0 ? void 0 : _a.silence) {
      this.silenceCount++;
    }
    else {
      this.silenceCount = 0;
    }
    return this.silenceCount > 5;
  }
  render() {
    const defaults = {
      meeting: this.meeting,
      states: this.states || state,
      iconPack: this.iconPack,
      t: this.t,
      size: this.size,
    };
    const deviceInfoKeys = ['label', 'channelType'];
    return (h(Host, null, h("audio", { preload: "auto", src: "https://assets.dyte.io/ui-kit/speaker-test.mp3", ref: (el) => (this.testAudioEl = el) }), h("div", { id: "header" }, ['Report', 'Troubleshoot'].map((section) => (h("dyte-button", { key: section, variant: "ghost", class: { active: this.activeTab === section }, onClick: () => this.setActiveTab(section), iconPack: this.iconPack, t: this.t }, this.t(section))))), this.activeTab === 'Troubleshoot' && (h("div", { class: "tab-body" }, h("div", null, h("div", { class: "stage-indicator" }, "Step ", this.stage + 1, "/3"), this.stage === 0 && (h("dyte-debugger-permissions-ui", Object.assign({}, defaults, { mediaType: "audio" }))), this.stage === 1 && (h("div", null, h("div", { class: "title" }, "Microphone Test"), h("div", { class: "text" }, this.t('Please start the microphone test and start speaking in order to test your audio device.')), h("dyte-button", { size: "md", onClick: () => this.toggleMicTest() }, this.micPreview ? 'Stop' : 'Start', " Microphone Test"), this.micPreview && (h("div", { class: "col" }, h("div", { class: "sub-title" }, "Device Information", h("dyte-information-tooltip", { iconPack: this.iconPack }, h("div", { slot: "tootlip-text", class: "tooltip-text" }, Object.keys(this.micMetadata).map((key) => (h("div", null, this.formatLabel(key), ": ", h("span", null, this.micMetadata[key]))))))), deviceInfoKeys.map((key) => {
      const val = this.micMetadata[key];
      if (!val)
        return;
      return (h("div", { class: "row" }, h("div", { class: "label" }, h("span", null, this.t(this.formatLabel(key))), ":"), h("div", { class: "text" }, this.t(val))));
    }), this.getDeviceInformationStatus(), h("div", { class: "sub-title" }, "Audio Level Analysis", h("dyte-information-tooltip", { iconPack: this.iconPack }, h("div", { slot: "tootlip-text", class: "tooltip-text" }, h("div", null, "Volume Level:", h("span", null, this.audioTrackStats.volume, "/100")), h("div", null, "Low Volume:", h("span", null, this.boolToStr(this.audioTrackStats.lowVolume))), h("div", null, "Silence Detected:", h("span", null, this.boolToStr(this.audioTrackStats.silence))), h("div", null, "Noise Detected:", h("span", null, this.boolToStr(this.audioTrackStats.clip))), h("div", null, "Time:", h("span", null, new Date(this.audioTrackStats.timestamp).toLocaleTimeString()))))), h("div", { class: "row" }, h("dyte-icon", { icon: this.iconPack.speaker, size: "md" }), h("div", { class: "volume-indicator" }, h("div", { class: "volume-level", ref: (el) => (this.volumeEl = el) }))), this.getAudioLevelStatus())))), this.stage === 2 && (h("div", { class: "col" }, h("div", { class: "title" }, "Speaker Test"), h("div", { class: "text" }, this.t('Upon starting this test, an audio will be played. Please ensure your system volume is not set to zero.')), h("dyte-button", { onClick: () => this.toggleSpeakerTest(), size: "md" }, this.speakerPreview ? 'Stop' : 'Start', " Speaker Test"), this.speakerDevice && (h("div", null, h("div", { class: "sub-title" }, "Device Information"), h("div", { class: "row" }, h("div", { class: "label" }, h("span", null, "Label"), ":"), h("div", { class: "text" }, this.speakerDevice.label)))), this.speakerPreview && (h("div", null, h("p", { class: "sub-title" }, "Are you able to hear the audio?"), h("div", { class: "speaker-test" }, h("dyte-button", { onClick: () => this.setSpeakerTestResult('success'), size: "sm", variant: "primary" }, "Yes"), h("dyte-button", { onClick: () => this.setSpeakerTestResult('failed'), size: "sm", variant: "secondary" }, "No")))), !this.speakerPreview && this.speakerTest !== 'none' && (h("div", { class: `row status-text ${this.speakerTest}` }, h("dyte-icon", { size: "sm", icon: this.iconPack[this.speakerTest === 'failed' ? 'dismiss' : 'checkmark'] }), this.getSpeakerTestStatus()))))), h("div", { class: "stage-manager" }, this.stage < 2 && (h("div", { class: { 'stage-button': true, disabled: this.mediaPermission !== 'ACCEPTED' }, onClick: () => this.changeStage(this.stage + 1) }, "Next", h("dyte-icon", { size: "sm", icon: this.iconPack.chevron_right }))), this.stage > 0 && (h("div", { class: { 'stage-button': true, disabled: this.mediaPermission !== 'ACCEPTED' }, onClick: () => this.changeStage(this.stage - 1) }, h("dyte-icon", { size: "sm", icon: this.iconPack.chevron_left }), "Prev"))))), this.activeTab === 'Report' && (h("div", { class: "tab-body" }, h("dyte-debugger-issues-ui", Object.assign({}, defaults, { mediaType: "audio" }))))));
  }
  static get watchers() { return {
    "meeting": ["meetingChanged"]
  }; }
  static get style() { return dyteDebuggerAudioCss; }
}, [1, "dyte-debugger-audio", {
    "meeting": [16],
    "states": [16],
    "size": [513],
    "iconPack": [16],
    "t": [16],
    "micMetadata": [32],
    "micPreview": [32],
    "speakerPreview": [32],
    "audioTrackStats": [32],
    "activeTab": [32],
    "stage": [32],
    "mediaPermission": [32],
    "speakerDevice": [32],
    "speakerTest": [32]
  }]);
function defineCustomElement() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["dyte-debugger-audio", "dyte-button", "dyte-debugger-issues-ui", "dyte-debugger-permissions-ui", "dyte-icon", "dyte-information-tooltip"];
  components.forEach(tagName => { switch (tagName) {
    case "dyte-debugger-audio":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DyteDebuggerAudio);
      }
      break;
    case "dyte-button":
      if (!customElements.get(tagName)) {
        defineCustomElement$5();
      }
      break;
    case "dyte-debugger-issues-ui":
      if (!customElements.get(tagName)) {
        defineCustomElement$4();
      }
      break;
    case "dyte-debugger-permissions-ui":
      if (!customElements.get(tagName)) {
        defineCustomElement$3();
      }
      break;
    case "dyte-icon":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
    case "dyte-information-tooltip":
      if (!customElements.get(tagName)) {
        defineCustomElement$1();
      }
      break;
  } });
}

export { DyteDebuggerAudio as D, defineCustomElement as d };
