// @dytesdk/ui-kit/dist/components/dyte-debugger-toggle.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client';
import { d as defaultIconPack } from './default-icon-pack.js';
import { u as useLanguage } from './index2.js';
import { s as state } from './store.js';
import { d as defineCustomElement$4 } from './dyte-controlbar-button2.js';
import { d as defineCustomElement$3 } from './dyte-icon2.js';
import { d as defineCustomElement$2 } from './dyte-spinner2.js';

const dyteDebuggerToggleCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{display:block}";

const DyteDebuggerToggle$1 = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.stateUpdate = createEvent(this, "dyteStateUpdate", 7);
    this.variant = 'button';
    this.meeting = undefined;
    this.states = undefined;
    this.t = useLanguage();
    this.iconPack = defaultIconPack;
    this.size = undefined;
  }
  toggleDebugger() {
    var _a;
    this.stateUpdate.emit({
      activeDebugger: !((_a = this.states) === null || _a === void 0 ? void 0 : _a.activeDebugger),
      activeMoreMenu: false,
    });
    state.activeDebugger = !state.activeDebugger;
    state.activeMoreMenu = false;
  }
  render() {
    var _a;
    if (!((_a = this.meeting) === null || _a === void 0 ? void 0 : _a.troubleshoot))
      return;
    return (h(Host, { title: this.t('Network & Trouble Shooting') }, h("dyte-controlbar-button", { size: this.size, iconPack: this.iconPack, t: this.t, onClick: () => this.toggleDebugger(), icon: this.iconPack.debug, label: this.t('Troubleshooting and Help'), variant: this.variant })));
  }
  static get style() { return dyteDebuggerToggleCss; }
}, [1, "dyte-debugger-toggle", {
    "variant": [513],
    "meeting": [16],
    "states": [16],
    "t": [16],
    "iconPack": [16],
    "size": [513]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["dyte-debugger-toggle", "dyte-controlbar-button", "dyte-icon", "dyte-spinner"];
  components.forEach(tagName => { switch (tagName) {
    case "dyte-debugger-toggle":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DyteDebuggerToggle$1);
      }
      break;
    case "dyte-controlbar-button":
      if (!customElements.get(tagName)) {
        defineCustomElement$4();
      }
      break;
    case "dyte-icon":
      if (!customElements.get(tagName)) {
        defineCustomElement$3();
      }
      break;
    case "dyte-spinner":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const DyteDebuggerToggle = DyteDebuggerToggle$1;
const defineCustomElement = defineCustomElement$1;

export { DyteDebuggerToggle, defineCustomElement };
