// @dytesdk/ui-kit/dist/components/dyte-ui-provider.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';
import './store.js';
import './breakout-rooms-manager.js';
import { a as getIconPack, g as generateConfig, p as provideDyteDesignSystem, b as getSize } from './size.js';
import { d as defaultConfig } from './default-ui-config.js';
import { u as useLanguage } from './index2.js';
import './livestream.js';
import './user-prefs.js';
import { d as defaultIconPack } from './default-icon-pack.js';

const DyteUiProvider$1 = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.meeting = undefined;
    this.size = undefined;
    this.config = defaultConfig;
    this.t = useLanguage();
    this.iconPackUrl = undefined;
    this.loadConfigFromPreset = true;
    this.applyDesignSystem = true;
    this.joinRoom = true;
    this.isReady = false;
    this.iconPack = defaultIconPack;
  }
  async iconPackUrlChanged(url) {
    this.iconPack = await getIconPack(url);
  }
  async meetingChanged(meeting) {
    var _a;
    if (!meeting)
      return;
    if (!meeting.self.roomJoined && this.joinRoom) {
      this.isReady = false;
      await meeting.joinRoom();
      this.isReady = true;
    }
    if (this.loadConfigFromPreset && meeting.self.config != null) {
      const theme = meeting.self.config;
      const { config } = generateConfig(theme, meeting);
      if (this.config === defaultConfig) {
        // only override the config if the object is same as defaultConfig
        // which means it's a different object passed via prop
        this.config = Object.assign(Object.assign(Object.assign({}, defaultConfig), config), { designTokens: Object.assign(Object.assign({}, defaultConfig.designTokens), config.designTokens) });
      }
    }
    if (this.applyDesignSystem &&
      ((_a = this.config) === null || _a === void 0 ? void 0 : _a.designTokens) != null &&
      typeof document !== 'undefined') {
      provideDyteDesignSystem(document.documentElement, this.config.designTokens);
    }
    Array.from(this.hostEl.children)
      .filter((element) => element.tagName.startsWith('DYTE-'))
      .forEach((element) => {
      element['meeting'] = meeting;
      element['config'] = this.config;
      element['size'] = this.size;
      element['iconPack'] = this.iconPack;
      element['t'] = this.t;
    });
  }
  connectedCallback() {
    this.resizeObserver = new ResizeObserver(() => this.handleResize());
    this.resizeObserver.observe(this.hostEl);
    this.meetingChanged(this.meeting);
    this.iconPackUrlChanged(this.iconPackUrl);
  }
  handleResize() {
    this.size = getSize(this.hostEl.clientWidth);
  }
  render() {
    if (!this.isReady)
      return null;
    return h("slot", null);
  }
  get hostEl() { return this; }
  static get watchers() { return {
    "iconPackUrl": ["iconPackUrlChanged"],
    "meeting": ["meetingChanged"]
  }; }
}, [4, "dyte-ui-provider", {
    "meeting": [16],
    "size": [1537],
    "config": [1040],
    "t": [16],
    "iconPackUrl": [513, "icon-pack-url"],
    "loadConfigFromPreset": [4, "load-config-from-preset"],
    "applyDesignSystem": [4, "apply-design-system"],
    "joinRoom": [4, "join-room"],
    "isReady": [32],
    "iconPack": [32]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["dyte-ui-provider"];
  components.forEach(tagName => { switch (tagName) {
    case "dyte-ui-provider":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DyteUiProvider$1);
      }
      break;
  } });
}

const DyteUiProvider = DyteUiProvider$1;
const defineCustomElement = defineCustomElement$1;

export { DyteUiProvider, defineCustomElement };
