// @dytesdk/ui-kit/dist/esm/dyte-controlbar.entry.js
import { r as registerInstance, h, H as Host } from './index-926e26e8.js';
import './store-8f4d7acc.js';
import './breakout-rooms-manager-7dc0174a.js';
import { d as defaultConfig } from './default-ui-config-f87e0ba7.js';
import './livestream-1a40fc8d.js';
import { d as defaultIconPack } from './default-icon-pack-307ffa7c.js';
import { u as useLanguage } from './index-7b89c939.js';
import './user-prefs-47572e95.js';
import { R as Render } from './index-9bb85c23.js';
import './breakout-rooms-dea0806e.js';
import './flags-f14ed3df.js';

const dyteControlbarCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{box-sizing:border-box;display:flex;align-items:center;--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-1000, 8 8 8) / var(--tw-bg-opacity));padding-left:var(--dyte-space-4, 16px);padding-right:var(--dyte-space-4, 16px);position:relative;z-index:10}@media only screen and (max-device-height: 480px) and (orientation: landscape){:host{padding-top:var(--dyte-space-0, 0px) !important}}";

const DyteControlbar = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.variant = 'solid';
    this.disableRender = false;
    this.meeting = undefined;
    this.config = defaultConfig;
    this.states = undefined;
    this.iconPack = defaultIconPack;
    this.t = useLanguage();
    this.size = undefined;
  }
  render() {
    const defaults = {
      meeting: this.meeting,
      config: this.config,
      states: this.states,
      t: this.t,
      iconPack: this.iconPack,
      size: this.size,
    };
    return (h(Host, null, !this.disableRender && (h(Render, { element: "dyte-controlbar", defaults: defaults, onlyChildren: true })), h("slot", null)));
  }
};
DyteControlbar.style = dyteControlbarCss;

export { DyteControlbar as dyte_controlbar };
