// @dytesdk/ui-kit/dist/esm/dyte-channel-creator.entry.js
import { r as registerInstance, c as createEvent, w as writeTask, f as forceUpdate, h, H as Host, a as getElement } from './index-926e26e8.js';
import { s as state } from './store-8f4d7acc.js';
import './breakout-rooms-manager-7dc0174a.js';
import './livestream-1a40fc8d.js';
import { d as defaultIconPack } from './default-icon-pack-307ffa7c.js';
import { u as useLanguage } from './index-7b89c939.js';
import './user-prefs-47572e95.js';
import './breakout-rooms-dea0806e.js';
import './flags-f14ed3df.js';

const dyteChannelCreatorCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}.scrollbar{scrollbar-width:thin;scrollbar-color:var(--dyte-scrollbar-color, rgb(var(--dyte-colors-background-600, 60 60 60)))\n    var(--dyte-scrollbar-background, transparent)}.scrollbar::-webkit-scrollbar{height:var(--dyte-space-1\\.5, 6px);width:var(--dyte-space-1\\.5, 6px);border-radius:9999px;background-color:var(--dyte-scrollbar-background, transparent)}.scrollbar::-webkit-scrollbar-thumb{border-radius:9999px;background-color:var(--dyte-scrollbar-color, rgb(var(--dyte-colors-background-600, 60 60 60)))}:host{border-radius:var(--dyte-border-radius-md, 8px);--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-1000, 8 8 8) / var(--tw-bg-opacity));display:flex;flex-direction:column;width:var(--dyte-space-96, 384px);padding-left:var(--dyte-space-9, 36px);padding-right:var(--dyte-space-9, 36px);padding-top:var(--dyte-space-10, 40px);padding-bottom:var(--dyte-space-10, 40px)}header{margin-bottom:var(--dyte-space-8, 32px);display:flex;align-items:center;gap:var(--dyte-space-2, 8px);font-size:24px;font-weight:600}.channel-name-input{--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-900, 26 26 26) / var(--tw-bg-opacity));color:rgb(var(--dyte-colors-text-1000, 255 255 255));border-radius:var(--dyte-border-radius-sm, 4px);border-width:var(--dyte-border-width-none, 0);border-style:none;padding:var(--dyte-space-3, 12px);font-size:16px;-webkit-appearance:none;-moz-appearance:none;appearance:none;line-height:1.25;outline:2px solid transparent;outline-offset:2px}.channel-name-input:focus{outline-width:2px;outline-color:rgba(var(--dyte-colors-brand-300, 73 124 253) / 0.5)}footer{margin-top:var(--dyte-space-5, 20px);display:flex;justify-content:flex-end}.member{display:flex;align-items:center;gap:var(--dyte-space-1, 4px);padding:var(--dyte-space-2, 8px)}.member dyte-avatar{height:var(--dyte-space-5, 20px);width:var(--dyte-space-5, 20px);font-size:14px;color:rgb(var(--dyte-colors-text-1000, 255 255 255))}.input-container{position:relative;-webkit-appearance:none;-moz-appearance:none;appearance:none;line-height:1.25;outline:2px solid transparent;outline-offset:2px;outline-width:2px;outline-color:rgba(var(--dyte-colors-brand-300, 73 124 253) / 0.5);border-radius:var(--dyte-border-radius-sm, 4px)}.input-container .members{margin:var(--dyte-space-0, 0px);max-height:var(--dyte-space-28, 112px);border-radius:var(--dyte-border-radius-sm, 4px);padding:var(--dyte-space-2, 8px);list-style-type:none;display:flex;flex-wrap:wrap;gap:var(--dyte-space-1, 4px);cursor:text;font-size:16px;--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-900, 26 26 26) / var(--tw-bg-opacity));overflow-y:auto;min-height:var(--dyte-space-7, 28px)}.input-container .pill{display:flex;align-items:center;gap:var(--dyte-space-1, 4px);border-radius:var(--dyte-border-radius-sm, 4px);padding:var(--dyte-space-2, 8px);--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-800, 30 30 30) / var(--tw-bg-opacity))}.input-container .pill dyte-avatar{height:var(--dyte-space-5, 20px);width:var(--dyte-space-5, 20px);font-size:14px;color:rgb(var(--dyte-colors-text-on-brand-1000, var(--dyte-colors-text-1000, 255 255 255)))}.input-container .pill span{max-width:var(--dyte-space-32, 128px);overflow:hidden;text-overflow:ellipsis;white-space:nowrap}.input-container .pill dyte-icon{height:var(--dyte-space-4, 16px);width:var(--dyte-space-4, 16px);cursor:pointer}.input-container input{width:var(--dyte-space-24, 96px);border-radius:var(--dyte-border-radius-sm, 4px);border-width:var(--dyte-border-width-none, 0);border-style:none;padding:var(--dyte-space-1, 4px);--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-900, 26 26 26) / var(--tw-bg-opacity));color:rgb(var(--dyte-colors-text-1000, 255 255 255));outline:2px solid transparent;outline-offset:2px;font-size:16px;line-height:1.25rem}.input-container input.wide-input{width:100%}.search-results{margin:var(--dyte-space-0, 0px);margin-top:var(--dyte-space-1, 4px);max-height:var(--dyte-space-28, 112px);width:100%;padding:var(--dyte-space-0, 0px);position:absolute;list-style-type:none;overflow-y:auto;--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-900, 26 26 26) / var(--tw-bg-opacity));border-radius:var(--dyte-border-radius-md, 8px);--tw-border-spacing-x:var(--dyte-space-2, 8px);--tw-border-spacing-y:var(--dyte-space-2, 8px);border-spacing:var(--tw-border-spacing-x) var(--tw-border-spacing-y);border-style:solid;border-color:rgba(var(--dyte-colors-brand-300, 73 124 253) / 0.5)}.search-results .member{cursor:pointer}.search-results .member dyte-avatar{color:rgb(var(--dyte-colors-text-on-brand-1000, var(--dyte-colors-text-1000, 255 255 255)))}.search-results .member:hover,.search-results .member.selected{--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-brand-700, 2 70 253) / var(--tw-bg-opacity));color:rgb(var(--dyte-colors-text-on-brand-1000, var(--dyte-colors-text-1000, 255 255 255)))}";

const DyteChannelCreator = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.stateUpdate = createEvent(this, "dyteStateUpdate", 7);
    this.switchChannel = createEvent(this, "switchChannel", 7);
    this.allMembers = new Map();
    this.inputTextRef = null;
    this.searchInputTextRef = null;
    this.focusOnSearch = (selectText = false) => {
      this.focusedMemberIndex = -1;
      writeTask(() => {
        var _a, _b;
        (_a = this.searchInputTextRef) === null || _a === void 0 ? void 0 : _a.focus();
        if (selectText)
          (_b = this.searchInputTextRef) === null || _b === void 0 ? void 0 : _b.select();
      });
    };
    this.onClickHandler = async () => {
      if (this.channelName.length === 0)
        return;
      if (this.step === 1) {
        const members = this.meeting.participants.all.toArray();
        const selfId = this.meeting.self.userId;
        const nonSelfMembers = members.filter((member) => member.userId !== selfId);
        nonSelfMembers.forEach((member) => this.allMembers.set(member.userId, member));
        this.step = 2;
        this.focusOnSearch();
        return;
      }
      // step 2 - add members and create channel
      await this.createChannel();
    };
    this.createChannel = async () => {
      const members = Array.from(this.selectedMemberIds);
      const newChannel = await this.meeting.chat.createChannel(this.channelName, members, {
        displayPictureUrl: '',
        visibility: 'public',
        isDirectMessage: false,
      });
      this.switchChannel.emit(newChannel.id);
      this.stateUpdate.emit({ activeChannelCreator: false });
      state.activeChannelCreator = false;
    };
    this.onMemberAdd = (id) => {
      this.showAllMembersList = false;
      this.selectedMemberIds.add(id);
      this.searchQuery = '';
      this.focusOnSearch();
    };
    this.keyDownHandler = (e, filteredMembers) => {
      if (e.key === 'ArrowDown') {
        this.focusedMemberIndex = Math.min(this.focusedMemberIndex + 1, filteredMembers.length - 1);
      }
      else if (e.key === 'ArrowUp') {
        if (this.focusedMemberIndex === -1)
          return;
        if (this.focusedMemberIndex === 0) {
          this.focusOnSearch(true);
          return;
        }
        this.focusedMemberIndex = Math.max(this.focusedMemberIndex - 1, 0);
      }
      else if (e.key === 'Enter') {
        this.onMemberAdd(filteredMembers[this.focusedMemberIndex].userId);
      }
      else if (e.key === 'Backspace') {
        if (this.searchQuery.length !== 0)
          return;
        if (this.selectedMemberIds.size === 0)
          return;
        const lastMemberId = Array.from(this.selectedMemberIds.values()).at(-1);
        this.selectedMemberIds.delete(lastMemberId);
        forceUpdate(this.$el);
      }
    };
    this.renderMemberSelector = () => {
      const filteredMembers = Array.from(this.allMembers.values()).filter((member) => !this.selectedMemberIds.has(member.userId) &&
        member.name.toLowerCase().includes(this.searchQuery.toLowerCase()));
      const selectedMembers = Array.from(this.selectedMemberIds.values()).map((id) => this.allMembers.get(id));
      const disableInput = this.selectedMemberIds.size === this.allMembers.size;
      return (h("div", { class: "input-container" }, h("ul", { class: "members scrollbar", onClick: () => {
          var _a;
          (_a = this.searchInputTextRef) === null || _a === void 0 ? void 0 : _a.focus();
        } }, selectedMembers.map((member) => (h("li", { class: "pill" }, h("dyte-avatar", { participant: {
          name: member.name,
          picture: member.picture,
        }, size: "sm" }), h("span", null, member.name), h("dyte-icon", { icon: this.iconPack.dismiss, iconPack: this.iconPack, t: this.t, onClick: () => {
          this.selectedMemberIds.delete(member.userId);
          forceUpdate(this.$el);
          this.focusOnSearch();
        } })))), !disableInput && (h("input", { type: "text", ref: (el) => (this.searchInputTextRef = el), value: this.searchQuery, placeholder: this.selectedMemberIds.size === 0 ? this.t('chat.member_name') : '', class: {
          'wide-input': this.selectedMemberIds.size === 0,
        }, onInput: (e) => {
          this.searchQuery = e.target.value.trim();
        }, onClick: () => {
          this.showAllMembersList = !this.showAllMembersList;
        }, onKeyDown: (e) => this.keyDownHandler(e, filteredMembers) }))), (this.searchQuery.length !== 0 || this.showAllMembersList) && (h("ul", { class: "search-results" }, filteredMembers.map((member, index) => (h("li", { class: { member: true, selected: index === this.focusedMemberIndex }, onClick: () => this.onMemberAdd(member.userId), ref: ($li) => {
          if (index === this.focusedMemberIndex) {
            writeTask(() => {
              if ($li)
                $li.scrollIntoView({ behavior: 'smooth', block: 'end', inline: 'nearest' });
            });
          }
        } }, h("dyte-avatar", { participant: {
          name: member.name,
          picture: member.picture,
        }, size: "sm" }), h("span", null, member.name)))), filteredMembers.length === 0 && (h("li", { class: "member" }, h("span", null, this.t('chat.error.empty_results'))))))));
    };
    this.meeting = undefined;
    this.t = useLanguage();
    this.iconPack = defaultIconPack;
    this.channelName = '';
    this.searchQuery = '';
    this.step = 1;
    this.loading = false;
    this.selectedMemberIds = new Set();
    this.focusedMemberIndex = -1;
    this.showAllMembersList = false;
  }
  componentDidLoad() {
    var _a;
    (_a = this.inputTextRef) === null || _a === void 0 ? void 0 : _a.focus();
  }
  render() {
    return (h(Host, null, h("header", null, this.step === 1 ? this.t('chat.new_channel') : this.t('chat.add_members')), this.step === 1 && (h("input", { class: "channel-name-input", type: "text", placeholder: this.t('chat.channel_name'), ref: (el) => (this.inputTextRef = el), onInput: (e) => {
        this.channelName = e.target.value.trim();
      } })), this.step === 2 && this.renderMemberSelector(), h("footer", null, h("dyte-button", { kind: "button", iconPack: this.iconPack, t: this.t, size: "lg", disabled: this.channelName.length === 0, onClick: this.onClickHandler }, this.step === 1 ? this.t('chat.add_members') : this.t('create')))));
  }
  get $el() { return getElement(this); }
};
DyteChannelCreator.style = dyteChannelCreatorCss;

export { DyteChannelCreator as dyte_channel_creator };
