// @dytesdk/ui-kit/dist/esm/dyte-debugger-toggle.entry.js
import { r as registerInstance, c as createEvent, h, H as Host } from './index-926e26e8.js';
import { d as defaultIconPack } from './default-icon-pack-307ffa7c.js';
import { u as useLanguage } from './index-7b89c939.js';
import { s as state } from './store-8f4d7acc.js';
import './user-prefs-47572e95.js';

const dyteDebuggerToggleCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{display:block}";

const DyteDebuggerToggle = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.stateUpdate = createEvent(this, "dyteStateUpdate", 7);
    this.variant = 'button';
    this.meeting = undefined;
    this.states = undefined;
    this.t = useLanguage();
    this.iconPack = defaultIconPack;
    this.size = undefined;
  }
  toggleDebugger() {
    var _a;
    this.stateUpdate.emit({
      activeDebugger: !((_a = this.states) === null || _a === void 0 ? void 0 : _a.activeDebugger),
      activeMoreMenu: false,
    });
    state.activeDebugger = !state.activeDebugger;
    state.activeMoreMenu = false;
  }
  render() {
    var _a;
    if (!((_a = this.meeting) === null || _a === void 0 ? void 0 : _a.troubleshoot))
      return;
    return (h(Host, { title: this.t('Network & Trouble Shooting') }, h("dyte-controlbar-button", { size: this.size, iconPack: this.iconPack, t: this.t, onClick: () => this.toggleDebugger(), icon: this.iconPack.debug, label: this.t('Troubleshooting and Help'), variant: this.variant })));
  }
};
DyteDebuggerToggle.style = dyteDebuggerToggleCss;

export { DyteDebuggerToggle as dyte_debugger_toggle };
