// @dytesdk/ui-kit/dist/esm/dyte-more-toggle.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, a as getElement } from './index-926e26e8.js';
import { d as defaultIconPack } from './default-icon-pack-307ffa7c.js';
import { u as useLanguage } from './index-7b89c939.js';
import { s as state } from './store-8f4d7acc.js';
import './user-prefs-47572e95.js';

const dyteMoreToggleCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{position:relative;display:flex;flex-direction:column;overflow:visible}.more-menu{position:absolute;right:calc(var(--dyte-space-24, 96px) * -1);bottom:var(--dyte-space-16, 64px);z-index:50;margin-bottom:var(--dyte-space-3, 12px);box-sizing:border-box;max-height:60vh;width:var(--dyte-space-64, 256px);overflow:auto;border-radius:var(--dyte-border-radius-md, 8px);border-width:var(--dyte-border-width-sm, 1px);border-style:solid;--tw-border-opacity:1;border-color:rgba(var(--dyte-colors-background-600, 60 60 60) / var(--tw-border-opacity));--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-1000, 8 8 8) / var(--tw-bg-opacity));outline:2px solid transparent;outline-offset:2px}:host([size='sm']) .more-menu{bottom:var(--dyte-space-10, 40px)}.more-menu::-webkit-scrollbar{height:var(--dyte-space-0, 0px);width:var(--dyte-space-1\\.5, 6px);--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-900, 26 26 26) / var(--tw-bg-opacity))}.more-menu::-webkit-scrollbar-thumb{border-radius:var(--dyte-border-radius-sm, 4px);--tw-bg-opacity:1;background-color:rgba(var(--dyte-colors-background-600, 60 60 60) / var(--tw-bg-opacity))}";

const DyteMoreToggle = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.stateUpdate = createEvent(this, "dyteStateUpdate", 7);
    this.handleKeyDown = ({ key }) => {
      if (key === 'Escape' && this.states.activeMoreMenu) {
        this.stateUpdate.emit({ activeMoreMenu: false });
        state.activeMoreMenu = !state.activeMoreMenu;
      }
    };
    this.handleOnClick = (e) => {
      if (!e.composedPath().includes(this.host) && this.states.activeMoreMenu) {
        this.stateUpdate.emit({ activeMoreMenu: false });
        state.activeMoreMenu = !state.activeMoreMenu;
      }
    };
    this.toggleMoreMenu = () => {
      this.stateUpdate.emit({ activeMoreMenu: !state.activeMoreMenu });
      state.activeMoreMenu = !state.activeMoreMenu;
    };
    this.states = state;
    this.size = undefined;
    this.iconPack = defaultIconPack;
    this.t = useLanguage();
  }
  connectedCallback() {
    /** A11y */
    window.addEventListener('keydown', this.handleKeyDown);
    window.addEventListener('click', this.handleOnClick);
    // };
  }
  disconnectedCallback() {
    window.removeEventListener('keydown', this.handleKeyDown);
    window.removeEventListener('click', this.handleOnClick);
  }
  render() {
    const text = this.t('more');
    return (h(Host, { title: text }, this.states.activeMoreMenu && (h("div", { class: "more-menu" }, h("slot", { name: "more-elements" }))), h("dyte-controlbar-button", { size: this.size, iconPack: this.iconPack, t: this.t, onClick: (e) => {
        e.stopPropagation();
        this.toggleMoreMenu();
      }, icon: this.iconPack.horizontal_dots, label: text, part: "controlbar-button" }), h("slot", { name: "expanded" })));
  }
  get host() { return getElement(this); }
};
DyteMoreToggle.style = dyteMoreToggleCss;

export { DyteMoreToggle as dyte_more_toggle };
