// @dytesdk/ui-kit/dist/esm/dyte-transcripts.entry.js
import { r as registerInstance, w as writeTask, h, H as Host, a as getElement } from './index-926e26e8.js';
import { u as useLanguage } from './index-7b89c939.js';
import { s as state } from './store-8f4d7acc.js';
import './breakout-rooms-manager-7dc0174a.js';
import { d as defaultConfig } from './default-ui-config-f87e0ba7.js';
import './livestream-1a40fc8d.js';
import './user-prefs-47572e95.js';
import { c as clone } from './clone-e2a39cf9.js';
import './breakout-rooms-dea0806e.js';
import './flags-f14ed3df.js';
import './cloneDeep-6a74a147.js';
import './keysIn-8bf552ee.js';
import './isObjectLike-0d1a05d5.js';

const dyteTranscriptsCss = ":host{line-height:initial;font-family:var(--dyte-font-family, sans-serif);font-feature-settings:normal;font-variation-settings:normal}p{margin:var(--dyte-space-0, 0px);padding:var(--dyte-space-0, 0px)}:host{position:absolute;top:var(--dyte-space-4, 16px);right:var(--dyte-space-4, 16px);bottom:var(--dyte-space-4, 16px);left:var(--dyte-space-4, 16px);top:auto;display:flex;flex-direction:column;pointer-events:none;justify-content:center;width:100%;z-index:5}dyte-transcript{margin-top:var(--dyte-space-2, 8px)}";

const DyteTranscripts = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.onTranscript = (transcript) => {
      if (transcript.transcript) {
        this.add(transcript);
      }
    };
    this.meeting = undefined;
    this.states = state;
    this.config = defaultConfig;
    this.t = useLanguage();
    this.transcripts = [];
    this.listenerAttached = false;
  }
  connectedCallback() {
    this.meetingChanged(this.meeting);
  }
  addListener(meeting) {
    var _a;
    (_a = meeting === null || meeting === void 0 ? void 0 : meeting.ai) === null || _a === void 0 ? void 0 : _a.addListener('transcript', this.onTranscript);
    this.listenerAttached = true;
  }
  clearListeners(meeting) {
    var _a;
    this.onTranscript && ((_a = meeting === null || meeting === void 0 ? void 0 : meeting.ai) === null || _a === void 0 ? void 0 : _a.removeListener('transcript', this.onTranscript));
    this.listenerAttached = false;
    clearTimeout(this.disconnectTimeout);
    this.transcripts = [];
  }
  disconnectedCallback() {
    if (this.meeting == null)
      return;
    this.clearListeners(this.meeting);
  }
  meetingChanged(meeting, oldMeeting) {
    clearTimeout(this.disconnectTimeout);
    if (oldMeeting !== undefined)
      this.clearListeners(oldMeeting);
    if (meeting == null)
      return;
    if (this.states.activeCaptions) {
      this.addListener(meeting);
    }
  }
  statesChanged(s) {
    const states = s || state;
    if (states.activeCaptions && !this.listenerAttached) {
      this.addListener(this.meeting);
    }
    if (!states.activeCaptions && this.listenerAttached) {
      this.clearListeners(this.meeting);
    }
  }
  transcriptionsReducer(acc, t) {
    if (!acc.length) {
      return [t];
    }
    let lastElement = acc[acc.length - 1];
    if (lastElement.peerId !== t.peerId) {
      return acc.concat(t);
    }
    const maxTranscriptLength = 400;
    if (lastElement.transcript.length + t.transcript.length > maxTranscriptLength) {
      return acc.concat(t);
    }
    lastElement.transcript += ' ' + t.transcript;
    acc.pop();
    return acc.concat(clone(lastElement));
  }
  add(transcript) {
    // show transcripts only if tab is in focus and a maximum of 3 at a time
    this.transcripts.splice(0, this.transcripts.length - 2);
    this.transcripts = this.transcriptionsReducer(this.transcripts, transcript);
  }
  remove(id) {
    this.transcripts = this.transcripts.filter((transcript) => transcript.id !== id);
  }
  handleDismiss(e) {
    e.stopPropagation();
    const id = e.detail;
    const el = this.host.shadowRoot.querySelector(`[data-id="${id}"]`);
    // exit animation
    el === null || el === void 0 ? void 0 : el.classList.add('exit');
    setTimeout(() => {
      writeTask(() => {
        this.remove(id);
      });
    }, 400);
  }
  render() {
    var _a;
    if (!this.states.activeCaptions)
      return;
    return (h(Host, null, (_a = this.transcripts) === null || _a === void 0 ? void 0 : _a.map((transcript) => (h("dyte-transcript", { key: transcript.id, "data-id": transcript.id, transcript: transcript, onDyteTranscriptDismiss: (e) => this.handleDismiss(e), t: this.t })))));
  }
  get host() { return getElement(this); }
  static get watchers() { return {
    "meeting": ["meetingChanged"],
    "states": ["statesChanged"]
  }; }
};
DyteTranscripts.style = dyteTranscriptsCss;

export { DyteTranscripts as dyte_transcripts };
