// @dytesdk/web-core/types/src/roomNode/BaseRoomNodeClient.js
import DyteError from '../client/DyteError';
export default class RoomNodeClient {
    meetingTitle;
    roomName;
    roomUUID;
    isDisconnected;
    roomJoined;
    #sfuHandler;
    peerId;
    authToken;
    maxPreferredStreams;
    get sfuHandler() {
        return this.#sfuHandler;
    }
    reconnect() {
        throw new DyteError('Method not implemented.');
    }
    async joinRoom(displayName, audioMuted, roomUUID, roomTitle, currentUserSharedMediaStates = {}) {
        throw new DyteError('Method not implemented.');
    }
    async leaveRoom() {
        throw new DyteError('Method not implemented.');
    }
    async handleSocketCallbacks(message) {
        throw new DyteError('Method not implemented.');
    }
    async handleSockets(message, callBack) {
        throw new DyteError('Method not implemented.');
    }
    async shareMic(audioTrack) {
        throw new DyteError('Method not implemented.');
    }
    async shareWebcam(videoTrack) {
        throw new DyteError('Method not implemented.');
    }
    async shareScreen(screenShareTracks) {
        throw new DyteError('Method not implemented.');
    }
    async unmuteSelf() {
        throw new DyteError('Method not implemented.');
    }
    async muteSelf() {
        throw new DyteError('Method not implemented.');
    }
    async pauseWebcam() {
        throw new DyteError('Method not implemented.');
    }
    async disableScreenShare() {
        throw new DyteError('Method not implemented.');
    }
    async acceptWaitingRequest(id) {
        throw new DyteError('Method not implemented.');
    }
    async rejectWaitingRequest(id) {
        throw new DyteError('Method not implemented.');
    }
    async muteAll(allowUnMute) {
        throw new DyteError('Method not implemented.');
    }
    async muteAllVideo() {
        throw new DyteError('Method not implemented.');
    }
    async disableAudio(peerId) {
        throw new DyteError('Method not implemented.');
    }
    async disableVideo(peerId) {
        throw new DyteError('Method not implemented.');
    }
    async kickAll() {
        throw new DyteError('Method not implemented.');
    }
    async kick(peerId) {
        throw new DyteError('Method not implemented.');
    }
    async pinPeer(id) {
        throw new DyteError('Method not implemented.');
    }
    async getChatMessages() {
        throw new DyteError('Method not implemented.');
    }
    async requestRemoteControl(request) {
        throw new DyteError('Method not implemented.');
    }
    async acceptRemoteControl(request) {
        throw new DyteError('Method not implemented.');
    }
    async terminateRemoteControl(arg0, active) {
        throw new DyteError('Method not implemented.');
    }
    async sendEventRemoteControl(hostPeerId, id, data) {
        throw new DyteError('Method not implemented.');
    }
    async getPolls() {
        throw new DyteError('Method not implemented.');
    }
    async newPoll(pollCreateMessage) {
        throw new DyteError('Method not implemented.');
    }
    async votePoll(pollVoteMessage) {
        throw new DyteError('Method not implemented.');
    }
    async getPage(currentPage) {
        throw new DyteError('Method not implemented.');
    }
    getConsumers() {
        throw new DyteError('Method not implemented.');
    }
    async sendMessage(message) {
        throw new DyteError('Method not implemented.');
    }
    async getRoomState() {
        throw new DyteError('Method not implemented.');
    }
    async addRoomPlugin(arg0) {
        throw new DyteError('Method not implemented.');
    }
    async removeRoomPlugin(arg0) {
        throw new DyteError('Method not implemented.');
    }
    async unpinChatMessage(arg0) {
        throw new DyteError('Method not implemented.');
    }
    async pinChatMessage(arg0) {
        throw new DyteError('Method not implemented.');
    }
    async sendChatMessage(text) {
        throw new DyteError('Method not implemented.');
    }
    async broadcastMessage(type, payload) {
        throw new DyteError('Method not implemented.');
    }
    async acceptAllRequestToJoinStage(acceptJoinStageRequests) {
        throw new DyteError('Method not implemented.');
    }
    async stopPresenting() {
        throw new DyteError('Method not implemented.');
    }
    async startPresenting() {
        throw new DyteError('Method not implemented.');
    }
    async requestToJoinStage(PRESENT) {
        throw new DyteError('Method not implemented.');
    }
    async rejectRequestToJoinStage(id) {
        throw new DyteError('Method not implemented.');
    }
    async removePeerFromStage(id, REQUESTED_BY_MODERATOR) {
        throw new DyteError('Method not implemented.');
    }
    stopAllProducers() {
        throw new DyteError('Method not implemented.');
    }
}
