// @dytesdk/web-core/types/src/roomNode/RoomNodeConfig.js
export const VIDEO_CONSTRAINTS = {
    qvga: { width: { ideal: 320 }, height: { ideal: 240 } },
    vga: { width: { ideal: 640 }, height: { ideal: 480 } },
    hd: { width: { ideal: 1280 }, height: { ideal: 720 } },
    hd_cropped: { width: { ideal: 900 }, height: { ideal: 720 } },
};
export const VIDEO_CONSTRAINTS_SCREENSHARE = {
    vga: { width: { max: 640 }, height: { max: 480 } },
    hd: { width: { max: 1280 }, height: { max: 720 } },
    full_hd: { width: { max: 1980 }, height: { max: 1080 } },
};
export const PC_PROPRIETARY_CONSTRAINTS = {};
export const WEBCAM_SIMULCAST_ENCODINGS = [
    { rid: 'r0', scalabilityMode: 'L1T3' },
    { rid: 'r1', scalabilityMode: 'L1T3' },
    { rid: 'r2', scalabilityMode: 'L1T3' },
];
export const WEBCAM_SIMULCAST_ENCODINGS_VGA = [
    { rid: 'r0', scalabilityMode: 'L1T2' },
    { rid: 'r1', scalabilityMode: 'L1T2' },
];
export const SCREENSHARE_SIMULCAST_ENCODINGS = [
    {
        active: false,
        adaptivePtime: false,
        networkPriority: 'low',
        priority: 'low',
    },
    {
        active: false,
        adaptivePtime: false,
        networkPriority: 'low',
        priority: 'low',
    },
    {
        active: true,
        adaptivePtime: false,
        networkPriority: 'low',
        priority: 'low',
        maxBitrate: 2_500_000,
    },
];
export const SCREENSHARE_SIMULCAST_ENCODINGS_RTP = [
    {
        active: false,
        priority: 'low',
    },
    {
        active: false,
        priority: 'low',
    },
    {
        active: true,
        priority: 'low',
    },
];
export const WEBCAM_SIMULCAST_ENCODINGS_HIVE_CLIENT = {
    '320': [
        {
            rid: 'q',
            maxBitrate: 250_000,
            maxFramerate: 30.0,
            scalabilityMode: 'L1T3',
        },
    ],
    '640': [
        {
            rid: 'q',
            scaleResolutionDownBy: 2.0,
            maxBitrate: 250_000,
            maxFramerate: 30.0,
            scalabilityMode: 'L1T3',
        },
        {
            rid: 'h',
            maxBitrate: 750_000,
            maxFramerate: 30.0,
            scalabilityMode: 'L1T3',
        },
    ],
    '1280': [
        {
            rid: 'q',
            scaleResolutionDownBy: 4.0,
            maxBitrate: 250_000,
            maxFramerate: 30.0,
            scalabilityMode: 'L1T3',
        },
        {
            rid: 'h',
            scaleResolutionDownBy: 2.0,
            maxBitrate: 750_000,
            maxFramerate: 30.0,
            scalabilityMode: 'L1T3',
        },
        {
            rid: 'f',
            maxBitrate: 1_500_000,
            maxFramerate: 30.0,
            scalabilityMode: 'L1T3',
        },
    ],
};
export const WEBCAM_KSVC_ENCODINGS = [
    { scalabilityMode: 'L3T3_KEY' },
];
export const SCREEN_SHARING_SIMULCAST_ENCODINGS = [
    { dtx: true, maxBitrate: 1500000 },
    { dtx: true, maxBitrate: 6000000 },
];
export const SCREEN_SHARING_SVC_ENCODINGS = [
    { scalabilityMode: 'L3T3', dtx: true },
];
export const HD_QUALITY_PEERS_LIMIT = 1;
export var PRODUCERS_TYPE;
(function (PRODUCERS_TYPE) {
    PRODUCERS_TYPE["WEBCAM"] = "webcam";
    PRODUCERS_TYPE["MIC"] = "mic";
    PRODUCERS_TYPE["SCREENSHARE_VIDEO"] = "screenshare_video";
    PRODUCERS_TYPE["SCREENSHARE_AUDIO"] = "screenshare_audio";
})(PRODUCERS_TYPE || (PRODUCERS_TYPE = {}));
