// @dytesdk/ui-kit/dist/collection/components/dyte-text-message-view/dyte-text-message-view.js
import { h } from '@stencil/core';
import { hasOnlyEmojis } from '../../utils/string';
/**
 * A component which renders a text message from chat.
 */
export class DyteTextMessageView {
  constructor() {
    this.text = undefined;
    this.isMarkdown = false;
  }
  render() {
    return (h("p", { class: { text: true, emoji: hasOnlyEmojis(this.text) } }, this.isMarkdown ? h("dyte-markdown-view", { text: this.text }) : this.text));
  }
  static get is() { return "dyte-text-message-view"; }
  static get originalStyleUrls() {
    return {
      "$": ["dyte-text-message-view.css"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dyte-text-message-view.css"]
    };
  }
  static get properties() {
    return {
      "text": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": true,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Text message"
        },
        "attribute": "text",
        "reflect": false
      },
      "isMarkdown": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Renders text as markdown (default = true)"
        },
        "attribute": "is-markdown",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
}
