// @dytesdk/react-ui-kit/dist/es/components/stencil-generated/index.js
/* eslint-disable */
/* tslint:disable */
/* auto-generated react proxies */
import { createReactComponent } from './react-component-lib';
import { defineCustomElements } from '@dytesdk/ui-kit/loader';
defineCustomElements();
export const DyteAi = /*@__PURE__*/ createReactComponent('dyte-ai');
export const DyteAiChat = /*@__PURE__*/ createReactComponent('dyte-ai-chat');
export const DyteAiHome = /*@__PURE__*/ createReactComponent('dyte-ai-home');
export const DyteAiToggle = /*@__PURE__*/ createReactComponent('dyte-ai-toggle');
export const DyteAiTranscriptions = /*@__PURE__*/ createReactComponent('dyte-ai-transcriptions');
export const DyteAudioGrid = /*@__PURE__*/ createReactComponent('dyte-audio-grid');
export const DyteAudioTile = /*@__PURE__*/ createReactComponent('dyte-audio-tile');
export const DyteAudioVisualizer = /*@__PURE__*/ createReactComponent('dyte-audio-visualizer');
export const DyteAvatar = /*@__PURE__*/ createReactComponent('dyte-avatar');
export const DyteBreakoutRoomManager = /*@__PURE__*/ createReactComponent('dyte-breakout-room-manager');
export const DyteBreakoutRoomParticipants = /*@__PURE__*/ createReactComponent('dyte-breakout-room-participants');
export const DyteBreakoutRoomsManager = /*@__PURE__*/ createReactComponent('dyte-breakout-rooms-manager');
export const DyteBreakoutRoomsToggle = /*@__PURE__*/ createReactComponent('dyte-breakout-rooms-toggle');
export const DyteBroadcastMessageModal = /*@__PURE__*/ createReactComponent('dyte-broadcast-message-modal');
export const DyteButton = /*@__PURE__*/ createReactComponent('dyte-button');
export const DyteCameraToggle = /*@__PURE__*/ createReactComponent('dyte-camera-toggle');
export const DyteCaptionToggle = /*@__PURE__*/ createReactComponent('dyte-caption-toggle');
export const DyteChannelCreator = /*@__PURE__*/ createReactComponent('dyte-channel-creator');
export const DyteChannelDetails = /*@__PURE__*/ createReactComponent('dyte-channel-details');
export const DyteChannelHeader = /*@__PURE__*/ createReactComponent('dyte-channel-header');
export const DyteChannelSelectorUi = /*@__PURE__*/ createReactComponent('dyte-channel-selector-ui');
export const DyteChannelSelectorView = /*@__PURE__*/ createReactComponent('dyte-channel-selector-view');
export const DyteChat = /*@__PURE__*/ createReactComponent('dyte-chat');
export const DyteChatComposerUi = /*@__PURE__*/ createReactComponent('dyte-chat-composer-ui');
export const DyteChatComposerView = /*@__PURE__*/ createReactComponent('dyte-chat-composer-view');
export const DyteChatMessage = /*@__PURE__*/ createReactComponent('dyte-chat-message');
export const DyteChatMessagesUi = /*@__PURE__*/ createReactComponent('dyte-chat-messages-ui');
export const DyteChatMessagesUiPaginated = /*@__PURE__*/ createReactComponent('dyte-chat-messages-ui-paginated');
export const DyteChatSearchResults = /*@__PURE__*/ createReactComponent('dyte-chat-search-results');
export const DyteChatSelectorUi = /*@__PURE__*/ createReactComponent('dyte-chat-selector-ui');
export const DyteChatToggle = /*@__PURE__*/ createReactComponent('dyte-chat-toggle');
export const DyteClock = /*@__PURE__*/ createReactComponent('dyte-clock');
export const DyteConfirmationModal = /*@__PURE__*/ createReactComponent('dyte-confirmation-modal');
export const DyteControlbar = /*@__PURE__*/ createReactComponent('dyte-controlbar');
export const DyteControlbarButton = /*@__PURE__*/ createReactComponent('dyte-controlbar-button');
export const DyteCounter = /*@__PURE__*/ createReactComponent('dyte-counter');
export const DyteDebugger = /*@__PURE__*/ createReactComponent('dyte-debugger');
export const DyteDebuggerAudio = /*@__PURE__*/ createReactComponent('dyte-debugger-audio');
export const DyteDebuggerIssuesUi = /*@__PURE__*/ createReactComponent('dyte-debugger-issues-ui');
export const DyteDebuggerNetwork = /*@__PURE__*/ createReactComponent('dyte-debugger-network');
export const DyteDebuggerPermissionsUi = /*@__PURE__*/ createReactComponent('dyte-debugger-permissions-ui');
export const DyteDebuggerScreenshare = /*@__PURE__*/ createReactComponent('dyte-debugger-screenshare');
export const DyteDebuggerToggle = /*@__PURE__*/ createReactComponent('dyte-debugger-toggle');
export const DyteDebuggerVideo = /*@__PURE__*/ createReactComponent('dyte-debugger-video');
export const DyteDialog = /*@__PURE__*/ createReactComponent('dyte-dialog');
export const DyteDialogManager = /*@__PURE__*/ createReactComponent('dyte-dialog-manager');
export const DyteDraftAttachmentView = /*@__PURE__*/ createReactComponent('dyte-draft-attachment-view');
export const DyteEmojiPicker = /*@__PURE__*/ createReactComponent('dyte-emoji-picker');
export const DyteEmojiPickerButton = /*@__PURE__*/ createReactComponent('dyte-emoji-picker-button');
export const DyteEndedScreen = /*@__PURE__*/ createReactComponent('dyte-ended-screen');
export const DyteFileDropzone = /*@__PURE__*/ createReactComponent('dyte-file-dropzone');
export const DyteFileMessage = /*@__PURE__*/ createReactComponent('dyte-file-message');
export const DyteFileMessageView = /*@__PURE__*/ createReactComponent('dyte-file-message-view');
export const DyteFilePickerButton = /*@__PURE__*/ createReactComponent('dyte-file-picker-button');
export const DyteFullscreenToggle = /*@__PURE__*/ createReactComponent('dyte-fullscreen-toggle');
export const DyteGrid = /*@__PURE__*/ createReactComponent('dyte-grid');
export const DyteGridPagination = /*@__PURE__*/ createReactComponent('dyte-grid-pagination');
export const DyteHeader = /*@__PURE__*/ createReactComponent('dyte-header');
export const DyteIcon = /*@__PURE__*/ createReactComponent('dyte-icon');
export const DyteIdleScreen = /*@__PURE__*/ createReactComponent('dyte-idle-screen');
export const DyteImageMessage = /*@__PURE__*/ createReactComponent('dyte-image-message');
export const DyteImageMessageView = /*@__PURE__*/ createReactComponent('dyte-image-message-view');
export const DyteImageViewer = /*@__PURE__*/ createReactComponent('dyte-image-viewer');
export const DyteInformationTooltip = /*@__PURE__*/ createReactComponent('dyte-information-tooltip');
export const DyteJoinStage = /*@__PURE__*/ createReactComponent('dyte-join-stage');
export const DyteLeaveButton = /*@__PURE__*/ createReactComponent('dyte-leave-button');
export const DyteLeaveMeeting = /*@__PURE__*/ createReactComponent('dyte-leave-meeting');
export const DyteLivestreamIndicator = /*@__PURE__*/ createReactComponent('dyte-livestream-indicator');
export const DyteLivestreamPlayer = /*@__PURE__*/ createReactComponent('dyte-livestream-player');
export const DyteLivestreamToggle = /*@__PURE__*/ createReactComponent('dyte-livestream-toggle');
export const DyteLogo = /*@__PURE__*/ createReactComponent('dyte-logo');
export const DyteMarkdownView = /*@__PURE__*/ createReactComponent('dyte-markdown-view');
export const DyteMeeting = /*@__PURE__*/ createReactComponent('dyte-meeting');
export const DyteMeetingTitle = /*@__PURE__*/ createReactComponent('dyte-meeting-title');
export const DyteMenu = /*@__PURE__*/ createReactComponent('dyte-menu');
export const DyteMenuItem = /*@__PURE__*/ createReactComponent('dyte-menu-item');
export const DyteMenuList = /*@__PURE__*/ createReactComponent('dyte-menu-list');
export const DyteMessageListView = /*@__PURE__*/ createReactComponent('dyte-message-list-view');
export const DyteMessageView = /*@__PURE__*/ createReactComponent('dyte-message-view');
export const DyteMicToggle = /*@__PURE__*/ createReactComponent('dyte-mic-toggle');
export const DyteMixedGrid = /*@__PURE__*/ createReactComponent('dyte-mixed-grid');
export const DyteMoreToggle = /*@__PURE__*/ createReactComponent('dyte-more-toggle');
export const DyteMuteAllButton = /*@__PURE__*/ createReactComponent('dyte-mute-all-button');
export const DyteMuteAllConfirmation = /*@__PURE__*/ createReactComponent('dyte-mute-all-confirmation');
export const DyteNameTag = /*@__PURE__*/ createReactComponent('dyte-name-tag');
export const DyteNetworkIndicator = /*@__PURE__*/ createReactComponent('dyte-network-indicator');
export const DyteNotification = /*@__PURE__*/ createReactComponent('dyte-notification');
export const DyteNotifications = /*@__PURE__*/ createReactComponent('dyte-notifications');
export const DyteOverlayModal = /*@__PURE__*/ createReactComponent('dyte-overlay-modal');
export const DytePaginatedList = /*@__PURE__*/ createReactComponent('dyte-paginated-list');
export const DyteParticipant = /*@__PURE__*/ createReactComponent('dyte-participant');
export const DyteParticipantCount = /*@__PURE__*/ createReactComponent('dyte-participant-count');
export const DyteParticipantSetup = /*@__PURE__*/ createReactComponent('dyte-participant-setup');
export const DyteParticipantTile = /*@__PURE__*/ createReactComponent('dyte-participant-tile');
export const DyteParticipants = /*@__PURE__*/ createReactComponent('dyte-participants');
export const DyteParticipantsAudio = /*@__PURE__*/ createReactComponent('dyte-participants-audio');
export const DyteParticipantsStageList = /*@__PURE__*/ createReactComponent('dyte-participants-stage-list');
export const DyteParticipantsStageQueue = /*@__PURE__*/ createReactComponent('dyte-participants-stage-queue');
export const DyteParticipantsToggle = /*@__PURE__*/ createReactComponent('dyte-participants-toggle');
export const DyteParticipantsViewerList = /*@__PURE__*/ createReactComponent('dyte-participants-viewer-list');
export const DyteParticipantsWaitingList = /*@__PURE__*/ createReactComponent('dyte-participants-waiting-list');
export const DytePermissionsMessage = /*@__PURE__*/ createReactComponent('dyte-permissions-message');
export const DytePipToggle = /*@__PURE__*/ createReactComponent('dyte-pip-toggle');
export const DytePluginMain = /*@__PURE__*/ createReactComponent('dyte-plugin-main');
export const DytePlugins = /*@__PURE__*/ createReactComponent('dyte-plugins');
export const DytePluginsToggle = /*@__PURE__*/ createReactComponent('dyte-plugins-toggle');
export const DytePoll = /*@__PURE__*/ createReactComponent('dyte-poll');
export const DytePollForm = /*@__PURE__*/ createReactComponent('dyte-poll-form');
export const DytePolls = /*@__PURE__*/ createReactComponent('dyte-polls');
export const DytePollsToggle = /*@__PURE__*/ createReactComponent('dyte-polls-toggle');
export const DyteRecordingIndicator = /*@__PURE__*/ createReactComponent('dyte-recording-indicator');
export const DyteRecordingToggle = /*@__PURE__*/ createReactComponent('dyte-recording-toggle');
export const DyteRemoteAccessManager = /*@__PURE__*/ createReactComponent('dyte-remote-access-manager');
export const DyteScreenShareToggle = /*@__PURE__*/ createReactComponent('dyte-screen-share-toggle');
export const DyteScreenshareView = /*@__PURE__*/ createReactComponent('dyte-screenshare-view');
export const DyteSettings = /*@__PURE__*/ createReactComponent('dyte-settings');
export const DyteSettingsAudio = /*@__PURE__*/ createReactComponent('dyte-settings-audio');
export const DyteSettingsToggle = /*@__PURE__*/ createReactComponent('dyte-settings-toggle');
export const DyteSettingsVideo = /*@__PURE__*/ createReactComponent('dyte-settings-video');
export const DyteSetupScreen = /*@__PURE__*/ createReactComponent('dyte-setup-screen');
export const DyteSidebar = /*@__PURE__*/ createReactComponent('dyte-sidebar');
export const DyteSidebarUi = /*@__PURE__*/ createReactComponent('dyte-sidebar-ui');
export const DyteSimpleGrid = /*@__PURE__*/ createReactComponent('dyte-simple-grid');
export const DyteSpinner = /*@__PURE__*/ createReactComponent('dyte-spinner');
export const DyteSpotlightGrid = /*@__PURE__*/ createReactComponent('dyte-spotlight-grid');
export const DyteSpotlightIndicator = /*@__PURE__*/ createReactComponent('dyte-spotlight-indicator');
export const DyteStage = /*@__PURE__*/ createReactComponent('dyte-stage');
export const DyteStageToggle = /*@__PURE__*/ createReactComponent('dyte-stage-toggle');
export const DyteSwitch = /*@__PURE__*/ createReactComponent('dyte-switch');
export const DyteTabBar = /*@__PURE__*/ createReactComponent('dyte-tab-bar');
export const DyteTextComposerView = /*@__PURE__*/ createReactComponent('dyte-text-composer-view');
export const DyteTextField = /*@__PURE__*/ createReactComponent('dyte-text-field');
export const DyteTextMessage = /*@__PURE__*/ createReactComponent('dyte-text-message');
export const DyteTextMessageView = /*@__PURE__*/ createReactComponent('dyte-text-message-view');
export const DyteTooltip = /*@__PURE__*/ createReactComponent('dyte-tooltip');
export const DyteTranscript = /*@__PURE__*/ createReactComponent('dyte-transcript');
export const DyteTranscripts = /*@__PURE__*/ createReactComponent('dyte-transcripts');
export const DyteUiProvider = /*@__PURE__*/ createReactComponent('dyte-ui-provider');
export const DyteViewerCount = /*@__PURE__*/ createReactComponent('dyte-viewer-count');
export const DyteWaitingScreen = /*@__PURE__*/ createReactComponent('dyte-waiting-screen');
export const DyteWebinarStageToggle = /*@__PURE__*/ createReactComponent('dyte-webinar-stage-toggle');
