// @dyteinternals/utils/dist/constant/defaultPreset.js
import { DyteViewTypes, RecorderType, WaitingRoomTypesV1 } from '../types';
export const permissions = {
    view_type: DyteViewTypes.groupCall,
    accept_waiting_requests: false,
    accept_present_requests: false,
    request_produce: false,
    can_allow_participant_audio: false,
    can_allow_participant_screensharing: false,
    can_allow_participant_video: false,
    can_spotlight: false,
    request_kick_participant: false,
    kick_participant: false,
    pin_participant: false,
    can_edit_display_name: false,
    can_record: false,
    can_livestream: false,
    can_present: true,
    waiting_room_type: WaitingRoomTypesV1.skipOnAccept,
    recorder_type: RecorderType.none,
    plugins: {
        can_close: true,
        can_start: true,
    },
    polls: {
        can_create: true,
        can_vote: true,
        can_view: true,
    },
    produce: {
        video: {
            allow: true,
            quality: 'vga',
            frame_rate: 24,
        },
        audio: true,
        screenshare: {
            allow: true,
            quality: 'hd',
            frame_rate: 5,
        },
    },
    chat: {
        public: {
            can_send: true,
            text: true,
            files: true,
        },
        private: {
            can_send: false,
            can_receive: false,
            text: false,
            files: false,
        },
    },
    connected_meetings: {
        can_alter_connected_meetings: false,
        can_switch_connected_meetings: false,
        can_switch_to_parent_meeting: false,
    },
    reactions: false,
    hidden_participant: false,
    is_recorder: false,
    show_participant_list: true,
    can_change_participant_role: false,
    can_change_theme: false,
    max_screenshare_count: 1,
};
export const theme = {
    setup_screen: {
        is_enabled: true,
    },
    alone_here: {
        is_enabled: true,
    },
    waiting_room: {
        is_enabled: false,
        enable_preview: true,
    },
    control_bar: {
        is_enabled: true,
        elements: {
            plugins: true,
            screenshare: true,
            invite: true,
            participants: true,
            chat: true,
            reactions: false,
            polls: true,
            fullscreen: true,
            layout: true,
        },
    },
    header: {
        is_enabled: true,
        elements: {
            timer: true,
            title: true,
            participant_count: true,
            change_layout: true,
        },
    },
    pip_mode: true,
    auto_tune: true,
    grid: {
        multi: {
            maxVideoCount: 6,
            videoFit: 'cover',
        },
        single: {
            maxVideoCount: 6,
            videoFit: 'cover',
        },
        defaultView: 'multi',
    },
    controls: {
        pip_toggle: false,
    },
};
