// @dytesdk/web-core/types/src/utils/flags.js
export var FlagsmithIdentifierPrefix;
(function (FlagsmithIdentifierPrefix) {
    FlagsmithIdentifierPrefix["PARTICIPANT"] = "PARTICIPANT";
    FlagsmithIdentifierPrefix["PEER"] = "PEER";
    FlagsmithIdentifierPrefix["CLIENT"] = "CLIENT";
})(FlagsmithIdentifierPrefix || (FlagsmithIdentifierPrefix = {}));
export const FlagsmithFeatureFlags = {
    INTERNAL_CALL_STATS: 'internal_call_stats',
    SIMULCAST: 'simulcast',
    CHAT_SOCKET_SERVER: 'chat_socket_server',
    POLL_SOCKET_SERVER: 'poll_socket_server',
    PLUGIN_SOCKET_SERVER: 'plugin_socket_server',
    NR_OTEL_WEB: 'nr_otel_web',
    CONNECTED_MEETINGS: 'connected_meetings',
    FEAT_SPOTLIGHT: 'feat_spotlight',
    ICE_RESTART_ON_FAILED_STATE: 'ice_restart_on_failed_state',
    ICE_RESTART_ON_DISCONNECTED_STATE: 'ice_restart_on_disconnected_state',
    ENABLE_ICE_STATE_LOGGING: 'enable_ice_state_logging',
    REMOVE_OPERATIONAL_MIC: 'web_core_remove_operational_mic',
    SUPPRESS_PEER_MUTE_UNMUTE_EMITS: 'web_core_suppress_peer_mute_unmute_emits',
    SKIP_OTEL_TRACES: 'skip_otel_traces',
    NEW_LOCAL_MEDIA_HANDLER: 'new_local_media_handler',
    CALLSTATS_INGESTION_LAYER: 'callstats_ingestion_layer',
    USE_USERIDS_IN_CHAT: 'use_userids_in_chat',
    CUSTOM_PING_PONG: 'custom_ping_pong',
    ENABLE_HIVE_SIMULCAST: 'enable_hive_simulcast',
    ENABLE_HIVE_TRANSPORT_RECONNECTION_ON_ICE_FAILED: 'enable_hive_transport_reconnection_on_ice_failed',
    BYPASS_LOG_EXCLUSION_LIST: 'bypass_log_exclusion_list',
    LOG_LEVEL: 'log_level',
    V1_PLUGINS: 'v1_plugins',
    SCREENSHARE_DTX: 'screenshare_dtx',
    SCREENSHARE_PRIORITY: 'screenshare_priority',
    SCREENSHARE_MIN_BITRATE: 'screenshare_minbitrate',
    SCREENSHARE_MAX_DIMENSIONS: 'screenshare_maxdimensions',
    SCREENSHARE_SIMULCAST: 'screenshare_simulcast',
    DISABLE_WEBCAM_LAYERS_ON_SCREENSHARE: 'disable_webcam_layers_on_screenshare',
    SCREENSHARE_FORCE_GOOG_CONFERENCE: 'screenshare_force_goog_conference',
    LIVESTREAM: 'feat_livestream',
    FETCH_RETRY: 'fetch_retry',
    DEVICE_REMOVE_EXP: 'device_remove_exp',
    DISABLE_WEBCAM_SIMULCAST: 'webcore_disable_webcam_simulcast',
    OVERRIDE_WEBCAM_SIMULCAST: 'override_webcam_simulcast',
    SOCKET_POLLING: 'socket_polling',
    NS_NETWORK_SCORE: 'ns_network_score',
    FEAT_PAGINATED_CHAT: 'feat_paginated_chat',
    VAL_MIN_FRAMERATE: 'val_min_framerate',
    EXTENSION_HACK: 'extension_hack',
    VAL_MAX_FRAMERATE_EXT: 'val_max_framerate_ext',
    VAL_MIN_FRAMERATE_EXT: 'val_min_framerate_ext',
    SCREEENSHARE_ERR_HACK: 'screenshare_err_hack',
};
