// @dytesdk/web-core/types/src/utils/mediaPolyfill.js
import { browserSpecs } from '../browser/BrowserCapabilities';
export default function setupMediaPolyfill() {
    if (browserSpecs.isElectron()) {
        if (window.dyteElectronGetDisplayMediaSource) {
            navigator.mediaDevices.getDisplayMedia = async (constraints) => {
                const source = await window.dyteElectronGetDisplayMediaSource({
                    types: ['window', 'screen'],
                });
                let video = false;
                if (constraints.video) {
                    const videoConstrainst = constraints.video;
                    video = {
                        mandatory: {
                            chromeMediaSource: 'desktop',
                            chromeMediaSourceId: source.id,
                            maxWidth: videoConstrainst.width.max,
                            maxHeight: videoConstrainst.height.max,
                        },
                    };
                }
                const electronConstraints = {
                    audio: false,
                    video,
                };
                const stream = await navigator.mediaDevices
                    .getUserMedia(electronConstraints);
                return stream;
            };
        }
    }
}
