// @dytesdk/web-core/types/__mocks__/window/MediaStreamTrack.js
import { vi } from 'vitest';
let readyState = 'live';
const MediaStreamTrackMock = vi.fn(() => {
    const eventListeners = {};
    function addEventListener(event, cb) {
        eventListeners[event] = cb;
    }
    return {
        kind: vi.fn(),
        id: vi.fn(),
        label: vi.fn(),
        enabled: true,
        muted: vi.fn(),
        onmute: vi.fn(),
        onended: vi.fn(),
        contentHint: vi.fn(),
        applyConstraints: vi.fn(),
        clone: vi.fn(),
        getCapabilities: vi.fn(),
        getConstraints: vi.fn(),
        oncapturehandlechange: vi.fn(),
        getSettings: vi.fn(),
        addEventListener,
        removeEventListener: vi.fn(),
        stop: vi.fn(() => {
            readyState = 'ended';
        }),
        readyState,
        customResetReadyState: vi.fn(() => {
            readyState = 'live';
        }),
        emit: async (event) => {
            await eventListeners[event]?.();
        },
    };
}).mockName('MediaStreamTrack');
export default MediaStreamTrackMock;
