// @dytesdk/web-core/types/__mocks__/window/Navigator.js
import { vi } from 'vitest';
import MediaDevicesMock from './MediaDevices';
const NavigatorMock = vi.fn(() => ({
    mediaDevices: MediaDevicesMock,
    vendorSub: vi.fn(),
    productSub: vi.fn(),
    vendor: vi.fn(),
    maxTouchPoints: vi.fn(),
    scheduling: vi.fn(),
    userActivation: vi.fn(),
    doNotTrack: vi.fn(),
    geolocation: vi.fn(),
    connection: vi.fn(),
    plugins: vi.fn(),
    mimeTypes: vi.fn(),
    pdfViewerEnabled: vi.fn(),
    webkitTemporaryStorage: vi.fn(),
    webkitPersistentStorage: vi.fn(),
    hardwareConcurrency: vi.fn(),
    cookieEnabled: vi.fn(),
    appCodeName: vi.fn(),
    appName: vi.fn(),
    appVersion: vi.fn(),
    platform: vi.fn(),
    product: vi.fn(),
    userAgent: vi.fn(),
    language: vi.fn(),
    languages: vi.fn(),
    onLine: vi.fn(),
    webdriver: vi.fn(),
    getGamepads: vi.fn(),
    javaEnabled: vi.fn(),
    sendBeacon: vi.fn(),
    vibrate: vi.fn(),
    constructor: vi.fn(),
    clipboard: vi.fn(),
    credentials: vi.fn(),
    keyboard: vi.fn(),
    managed: vi.fn(),
    storage: vi.fn(),
    serviceWorker: vi.fn(),
    virtualKeyboard: vi.fn(),
    wakeLock: vi.fn(),
    deviceMemory: vi.fn(),
    ink: vi.fn(),
    hid: vi.fn(),
    locks: vi.fn(),
    mediaCapabilities: vi.fn(),
    mediaSession: vi.fn(),
    permissions: vi.fn(),
    presentation: vi.fn(),
    serial: vi.fn(),
    usb: vi.fn(),
    windowControlsOverlay: vi.fn(),
    xr: vi.fn(),
    userAgentData: vi.fn(),
    clearAppBadge: vi.fn(),
    getBattery: vi.fn(),
    getUserMedia: vi.fn(),
    requestMIDIAccess: vi.fn(),
    requestMediaKeySystemAccess: vi.fn(),
    setAppBadge: vi.fn(),
    webkitGetUserMedia: vi.fn(),
    getInstalledRelatedApps: vi.fn(),
    registerProtocolHandler: vi.fn(),
    unregisterProtocolHandler: vi.fn(),
}));
export default NavigatorMock;
