// @dytesdk/web-core/types/src/store/SessionEvents.js
export var SessionEvents;
(function (SessionEvents) {
    SessionEvents["NEW_PRODUCER"] = "NEW_PRODUCER";
    SessionEvents["ROOM_NODE_CONNECTED"] = "ROOM_NODE_CONNECTED";
    SessionEvents["ROOM_NODE_DISCONNECTED"] = "ROOM_NODE_DISCONNECTED";
    SessionEvents["ROOM_NODE_CONNECTION_ERROR"] = "ROOM_NODE_CONNECTION_ERROR";
    SessionEvents["SOCKET_SERVICE_CONNECTED"] = "SOCKET_SERVICE_CONNECTED";
    SessionEvents["SOCKET_SERVICE_DISCONNECTED"] = "SOCKET_SERVICE_DISCONNECTED";
    SessionEvents["SOCKET_SERVICE_RECONNECTING"] = "SOCKET_SERVICE_RECONNECTING";
    SessionEvents["SOCKET_SERVICE_RECONNECTION_ATTEMPT"] = "SOCKET_SERVICE_RECONNECTION_ATTEMPT";
    SessionEvents["SOCKET_SERVICE_RECONNECT_FAILURE"] = "SOCKET_SERVICE_RECONNECT_FAILURE";
    SessionEvents["SOCKET_SERVICE_RECONNECTED"] = "SOCKET_SERVICE_RECONNECTED";
    SessionEvents["SOCKET_SERVICE_FAILED"] = "SOCKET_SERVICE_FAILED";
    SessionEvents["ROOM_JOINED"] = "ROOM_JOINED";
    SessionEvents["SOCKET_SERVICE_ROOM_JOINED"] = "SOCKET_SERVICE_ROOM_JOINED";
    SessionEvents["SELF_ROOM_JOINED"] = "SELF_ROOM_JOINED";
    SessionEvents["PRODUCER_SCORE_UPDATE"] = "PRODUCER_SCORE_UPDATE";
    SessionEvents["CONSUMER_SCORE_UPDATE"] = "CONSUMER_SCORE_UPDATE";
    SessionEvents["PRODUCER_STATUS_UPDATE"] = "PRODUCER_STATUS_UPDATE";
    SessionEvents["CONSUMER_STATUS_UPDATE"] = "CONSUMER_STATUS_UPDATE";
    SessionEvents["LOW_PRODUCER_SCORE"] = "LOW_PRODUCER_SCORE";
    SessionEvents["LOW_CONSUMER_SCORE"] = "LOW_CONSUMER_SCORE";
    SessionEvents["MEDIA_PERMISSION_ERROR"] = "MEDIA_PERMISSION_ERROR";
    SessionEvents["MEDIA_PERMISSION_UPDATE"] = "MEDIA_PERMISSION_UPDATE";
    SessionEvents["WAITLISTED"] = "WAIT_LISTED";
    SessionEvents["WAITLIST_ACCEPTED"] = "websocket/waitlist-accepted";
    SessionEvents["WAITLIST_REJECTED"] = "websocket/waitlist-rejected";
    SessionEvents["WAITLIST_PEER_ADDED"] = "websocket/waitlist-peer-added";
    SessionEvents["WAITLIST_PEER_REMOVED"] = "websocket/waitlist-peer-closed";
    SessionEvents["MESSAGE"] = "websocket/message";
    SessionEvents["ROOM_MESSAGE"] = "websocket/room-message";
    SessionEvents["DISABLE_AUDIO"] = "websocket/disable-audio";
    SessionEvents["DISABLE_VIDEO"] = "websocket/disable-video";
    SessionEvents["PEER_JOINED"] = "websocket/peer-joined";
    SessionEvents["PEER_CLOSED"] = "websocket/peer-closed";
    SessionEvents["PEER_MUTED"] = "websocket/peer-muted";
    SessionEvents["PEER_UNMUTED"] = "websocket/peer-unmuted";
    SessionEvents["PEER_PINNED"] = "websocket/pin-peer";
    SessionEvents["PEER_UNPINNED"] = "websocker/unpin-peer";
    SessionEvents["MUTE_ALL"] = "websocket/mute-all";
    SessionEvents["MUTE_ALL_VIDEO"] = "websocket/mute-all-video";
    SessionEvents["CONSUMER_CLOSED"] = "websocket/consumer-closed";
    SessionEvents["CONSUMER_PAUSED"] = "websocket/consumer-paused";
    SessionEvents["CONSUMER_RESUMED"] = "websocket/consumer-resumed";
    SessionEvents["PRODUCER_CLOSED"] = "websocket/producer-closed";
    SessionEvents["NEW_CONSUMER"] = "websocket/new-consumer";
    SessionEvents["CONNECTED_MEETING_STATE"] = "websocket/connectedMeetingState";
    SessionEvents["MOVE_TO_CONNECTED_MEETING"] = "websocket/transferMeeting";
    SessionEvents["GET_CHAT_MESSAGES"] = "websocket/get-chat-messages";
    SessionEvents["NEW_CHAT_MESSAGE"] = "websocket/new-chat-message";
    SessionEvents["PIN_CHAT_MESSAGE"] = "websocket/pin-chat-message";
    SessionEvents["UNPIN_CHAT_MESSAGE"] = "websocket/unpin-chat-message";
    SessionEvents["GET_POLLS"] = "websocket/get-polls";
    SessionEvents["UPDATE_POLL"] = "websocket/update-poll";
    SessionEvents["SELECTED_PEERS"] = "websocket/selected-peers";
    SessionEvents["GET_PAGE"] = "websocket/get-page";
    SessionEvents["ACTIVE_SPEAKER"] = "websocket/active-speaker";
    SessionEvents["NO_ACTIVE_SPEAKER"] = "websocket/no-active-speaker";
    SessionEvents["PRODUCER_SCORE"] = "websocket/producer-score";
    SessionEvents["CONSUMER_SCORE"] = "websocket/consumer-score";
    SessionEvents["ENABLE_PLUGIN"] = "websocket/enable-plugin";
    SessionEvents["DISABLE_PLUGIN"] = "websocket/disable-plugin";
    SessionEvents["PLUGIN_EVENT"] = "websocket/plugin-event";
    SessionEvents["PLUGIN_DATA"] = "websocket/plugin-data";
    SessionEvents["PLUGIN_INTERNAL_DATA"] = "websocket/plugin-internal-data";
    SessionEvents["KICKED"] = "websocket/kicked";
    SessionEvents["RECORDING_STARTED"] = "websocket/recording-started";
    SessionEvents["RECORDING_STOPPED"] = "websocket/recording-stopped";
    SessionEvents["ICE_CONNECTED"] = "ICE_CONNECTED";
    SessionEvents["ICE_DISCONNECTED"] = "ICE_DISCONNECTED";
    SessionEvents["ICE_FAILED"] = "ICE_FAILED";
    SessionEvents["ICE_RECONNECTING"] = "ICE_RECONNECTING";
    SessionEvents["PRODUCER_TOGGLE"] = "hive/producer-toggle";
    SessionEvents["CONSUMER_TOGGLE"] = "hive/consumer-toggle";
    SessionEvents["SELECTED_PEERS_DIFF"] = "hive/selected-peers-diff";
    SessionEvents["REFRESH_GRID"] = "hive/refresh-grid";
    SessionEvents["RESET_PRODUCER_STATE"] = "hive/reset-producer-state";
    SessionEvents["HIVE_RECORDING_STARTED"] = "hive/recording-started";
    SessionEvents["HIVE_RECORDING_STOPPED"] = "hive/recording-stopped";
    SessionEvents["HIVE_TRANSPORT_STATE_CHANGED"] = "hive/transport-state-changed";
    SessionEvents["ROOM_STATE"] = "sockethub/room-state";
    SessionEvents["PEER_DISPLAY_NAME_CHANGED"] = "hive/display-name-changed";
    SessionEvents["ASSERT_SPOTLIGHTER"] = "websocket/assert-spotlighter";
    SessionEvents["PEER_REQUESTED_TO_JOIN_STAGE"] = "websocket/request-to-join-stage-peer-added";
    SessionEvents["PEER_REJECTED_TO_JOIN_STAGE"] = "websocket/peer-rejected-to-join-stage";
    SessionEvents["PEER_ADDED_TO_STAGE"] = "websocket/peer-added-to-stage";
    SessionEvents["PEER_REMOVED_FROM_STAGE"] = "websocket/peer-removed-from-stage";
    SessionEvents["PEER_STARTED_PRESENTING"] = "websocket/peer-started-presenting";
    SessionEvents["PEER_STOPPED_PRESENTING"] = "websocket/peer-stopped-presenting";
    SessionEvents["REQUEST_TO_JOIN_STAGE_ACCEPTED"] = "websocket/request-to-join-stage-accepted";
    SessionEvents["REQUEST_TO_JOIN_STAGE_REJECTED"] = "websocket/request-to-join-stage-rejected";
    SessionEvents["STARTED_PRESENTING"] = "websocket/started-presenting";
    SessionEvents["STOPPED_PRESENTING"] = "websocket/stopped-presenting";
    SessionEvents["REMOVED_FROM_STAGE"] = "websocket/removed-from-stage";
    SessionEvents["JOIN_MEDIA_ROOM"] = "JOIN_MEDIA_ROOM";
    SessionEvents["LEAVE_MEDIA_ROOM"] = "LEAVE_MEDIA_ROOM";
})(SessionEvents || (SessionEvents = {}));
export const temp = {};
