// @dytesdk/ui-kit/dist/collection/components/dyte-channel-selector-view/dyte-channel-selector-view.js
import { Host, h } from '@stencil/core';
import { defaultIconPack, useLanguage } from '../../exports';
import { debounce } from 'lodash-es';
export class DyteChannelSelectorView {
  constructor() {
    this.calculateListHeight = debounce(() => {
      if (this.viewAs === 'list' && this.$listEl) {
        let height = 0;
        const slotEl = this.$el.shadowRoot.querySelector('slot[name="header"]');
        if (slotEl) {
          slotEl.assignedElements().forEach((e) => (height += e.offsetHeight));
        }
        if (this.$searchEl) {
          height += this.$searchEl.offsetHeight;
        }
        this.$listEl.style.height = `${window.innerHeight - height - 16}px`;
      }
    }, 60);
    this.getFilteredChannels = () => {
      if (this.searchQuery.trim() === '') {
        return this.channels;
      }
      return this.channels.filter((channel) => {
        return channel['name'].toLowerCase().includes(this.searchQuery.toLowerCase());
      });
    };
    this.toggleDropdown = () => {
      this.showDropdown = !this.showDropdown;
    };
    this.getChannelById = (id) => {
      return this.channels.find((channel) => channel.id === id);
    };
    this.getTotalUnreads = () => {
      return this.channels.reduce((acc, curr) => {
        return acc + curr.unreadCount;
      }, 0);
    };
    this.onChannelClickHandler = (channel) => {
      this.channelChange.emit(channel);
      if (this.viewAs === 'dropdown') {
        this.showDropdown = false;
      }
    };
    this.channels = undefined;
    this.selectedChannelId = undefined;
    this.disableSearch = false;
    this.hideAvatar = false;
    this.iconPack = defaultIconPack;
    this.t = useLanguage();
    this.viewAs = 'list';
    this.searchQuery = '';
    this.showDropdown = false;
  }
  connectedCallback() {
    this.resizeObserver = new ResizeObserver(this.calculateListHeight);
  }
  componentDidLoad() {
    this.resizeObserver.observe(this.$el);
    this.calculateListHeight();
  }
  disconnectedCallback() {
    var _a;
    (_a = this.resizeObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
    this.calculateListHeight.cancel();
  }
  getTimeLabel(messageDate) {
    const today = new Date();
    const yesterday = new Date(today);
    yesterday.setDate(today.getDate() - 1);
    const firstDayOfWeek = new Date(today);
    firstDayOfWeek.setDate(today.getDate() - today.getDay() - 1);
    if (messageDate.toDateString() === today.toDateString()) {
      return messageDate.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
    }
    else if (messageDate.toDateString() === yesterday.toDateString()) {
      return this.t('date.yesteday');
    }
    else if (messageDate > firstDayOfWeek) {
      const weekdays = [
        'date.sunday',
        'date.monday',
        'date.tuesday',
        'date.wednesday',
        'date.thursday',
        'date.friday',
        'date.saturday',
      ];
      return this.t(weekdays[messageDate.getDay()]);
    }
    else {
      return Intl.DateTimeFormat([], {
        day: '2-digit',
        month: '2-digit',
        year: '2-digit',
      }).format(messageDate);
    }
  }
  render() {
    const filteredChannels = this.getFilteredChannels();
    const shouldShowDropdown = this.viewAs === 'list' || (this.viewAs === 'dropdown' && this.showDropdown);
    return (h(Host, null, this.viewAs === 'list' && h("slot", { name: "header" }), shouldShowDropdown && (h("div", { class: {
        dropdown: this.viewAs === 'dropdown',
        scrollbar: this.viewAs === 'dropdown',
        list: this.viewAs === 'list',
      } }, !this.disableSearch && (h("div", { class: "search-container", ref: (el) => (this.$searchEl = el) }, h("input", { type: "text", placeholder: this.t('chat.search_conversations'), value: this.searchQuery, onInput: (e) => (this.searchQuery = e.target.value) }), h("dyte-icon", { icon: this.iconPack.search }))), h("div", { class: "channels-container scrollbar", ref: (el) => (this.$listEl = el) }, filteredChannels.map((channel) => {
      return (h("button", { class: { channel: true, active: this.selectedChannelId === channel.id }, onClick: () => this.onChannelClickHandler(channel) }, !this.hideAvatar && (h("div", null, channel.icon ? (h("dyte-icon", { class: "avatar-icon", icon: this.iconPack[channel.icon] })) : (h("dyte-avatar", { participant: {
          name: channel.name,
          picture: channel.avatarUrl,
        } })))), h("div", { class: "channel-data" }, h("div", { class: "col" }, h("div", { class: "name" }, channel.name), channel.latestMessage && (h("div", { class: {
          'last-message': true,
          'no-message': !channel.latestMessage,
        } }, h("dyte-text-message-view", { isMarkdown: true, text: channel.latestMessage })))), h("div", { class: "col channel-meta" }, channel.latestMessageTime && (h("time", { class: "time" }, this.getTimeLabel(channel.latestMessageTime))), channel.unreadCount > 0 && (h("div", { class: "unread-count" }, channel.unreadCount))))));
    })))), this.viewAs === 'dropdown' && (h("button", { class: "dropdown-trigger", onClick: this.toggleDropdown }, h("span", null, this.selectedChannelId &&
      `${this.t('to')} ${this.getChannelById(this.selectedChannelId).name}`, this.getTotalUnreads() > 0 && (h("div", { class: "unread-count" }, this.getTotalUnreads()))), h("dyte-icon", { icon: this.showDropdown ? this.iconPack.chevron_up : this.iconPack.chevron_down })))));
  }
  static get is() { return "dyte-channel-selector-view"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["dyte-channel-selector-view.css"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dyte-channel-selector-view.css"]
    };
  }
  static get properties() {
    return {
      "channels": {
        "type": "unknown",
        "mutable": false,
        "complexType": {
          "original": "{\n    id: string;\n    name: string;\n    avatarUrl?: string;\n    icon?: keyof IconPack;\n    latestMessage?: string;\n    latestMessageTime?: Date;\n    unreadCount?: number;\n  }[]",
          "resolved": "{ id: string; name: string; avatarUrl?: string; icon?: \"pin\" | \"disconnected\" | \"dismiss\" | \"settings\" | \"image\" | \"pip_on\" | \"pip_off\" | \"viewers\" | \"mic_off\" | \"mic_on\" | \"video_off\" | \"video_on\" | \"participants\" | \"breakout_rooms\" | \"search\" | \"chat\" | \"add\" | \"delete\" | \"warning\" | \"people\" | \"people_checked\" | \"poll\" | \"rocket\" | \"call_end\" | \"share\" | \"share_screen_start\" | \"share_screen_stop\" | \"share_screen_person\" | \"clock\" | \"send\" | \"more_vertical\" | \"chevron_down\" | \"chevron_up\" | \"chevron_left\" | \"chevron_right\" | \"wifi\" | \"speaker\" | \"speaker_off\" | \"download\" | \"full_screen_maximize\" | \"full_screen_minimize\" | \"copy\" | \"attach\" | \"emoji_multiple\" | \"image_off\" | \"wand\" | \"recording\" | \"subtract\" | \"stop_recording\" | \"pin_off\" | \"spinner\" | \"shuffle\" | \"edit\" | \"back\" | \"save\" | \"web\" | \"checkmark\" | \"spotlight\" | \"join_stage\" | \"leave_stage\" | \"signal_1\" | \"signal_2\" | \"signal_3\" | \"signal_4\" | \"signal_5\" | \"start_livestream\" | \"stop_livestream\" | \"debug\" | \"info\" | \"devices\" | \"horizontal_dots\" | \"ai_sparkle\" | \"meeting_ai\" | \"create_channel\" | \"create_channel_illustration\" | \"captionsOn\" | \"captionsOff\"; latestMessage?: string; latestMessageTime?: Date; unreadCount?: number; }[]",
          "references": {
            "IconPack": {
              "location": "import",
              "path": "../../exports"
            },
            "Date": {
              "location": "global"
            }
          }
        },
        "required": true,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Channels"
        }
      },
      "selectedChannelId": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Selected channel id"
        },
        "attribute": "selected-channel-id",
        "reflect": false
      },
      "disableSearch": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Disables search bar (default = false)"
        },
        "attribute": "disable-search",
        "reflect": false,
        "defaultValue": "false"
      },
      "hideAvatar": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Hides avatar (default = false)"
        },
        "attribute": "hide-avatar",
        "reflect": false,
        "defaultValue": "false"
      },
      "iconPack": {
        "type": "unknown",
        "mutable": false,
        "complexType": {
          "original": "IconPack",
          "resolved": "{ people: string; people_checked: string; chat: string; poll: string; participants: string; rocket: string; call_end: string; share: string; mic_on: string; mic_off: string; video_on: string; video_off: string; share_screen_start: string; share_screen_stop: string; share_screen_person: string; clock: string; dismiss: string; send: string; search: string; more_vertical: string; chevron_down: string; chevron_up: string; chevron_left: string; chevron_right: string; settings: string; wifi: string; speaker: string; speaker_off: string; download: string; full_screen_maximize: string; full_screen_minimize: string; copy: string; attach: string; image: string; emoji_multiple: string; image_off: string; disconnected: string; wand: string; recording: string; subtract: string; stop_recording: string; warning: string; pin: string; pin_off: string; spinner: string; breakout_rooms: string; add: string; shuffle: string; edit: string; delete: string; back: string; save: string; web: string; checkmark: string; spotlight: string; join_stage: string; leave_stage: string; pip_off: string; pip_on: string; signal_1: string; signal_2: string; signal_3: string; signal_4: string; signal_5: string; start_livestream: string; stop_livestream: string; viewers: string; debug: string; info: string; devices: string; horizontal_dots: string; ai_sparkle: string; meeting_ai: string; create_channel: string; create_channel_illustration: string; captionsOn: string; captionsOff: string; }",
          "references": {
            "IconPack": {
              "location": "import",
              "path": "../../exports"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Icon Pack"
        },
        "defaultValue": "defaultIconPack"
      },
      "t": {
        "type": "unknown",
        "mutable": false,
        "complexType": {
          "original": "DyteI18n",
          "resolved": "(key: \"pin\" | \"unpin\" | \"kick\" | \"ended\" | \"disconnected\" | (string & {}) | \"about_call\" | \"screen\" | \"camera\" | \"leave\" | \"dismiss\" | \"logo\" | \"page\" | \"more\" | \"page.prev\" | \"page.next\" | \"layout\" | \"layout.auto\" | \"settings\" | \"file\" | \"image\" | \"connection\" | \"leave_confirmation\" | \"cancel\" | \"yes\" | \"you\" | \"to\" | \"mute\" | \"pinned\" | \"pip_on\" | \"pip_off\" | \"viewers\" | \"join\" | \"joined\" | \"create\" | \"close\" | \"ask\" | \"type\" | \"activate\" | \"mic_off\" | \"disable_mic\" | \"mic_on\" | \"enable_mic\" | \"audio\" | \"test\" | \"mute_all\" | \"mute_all.description\" | \"mute_all.header\" | \"mute_all.allow_unmute\" | \"video_off\" | \"disable_video\" | \"video_on\" | \"enable_video\" | \"video\" | \"offline\" | \"offline.description\" | \"disconnected.description\" | \"participants\" | \"participants.errors.empty_results\" | \"participants.empty_list\" | \"participants.turn_off_video\" | \"polls\" | \"polls.by\" | \"polls.question\" | \"polls.question.placeholder\" | \"polls.answers\" | \"polls.option\" | \"polls.option.placeholder\" | \"polls.results.anon\" | \"polls.results.hide\" | \"polls.create\" | \"polls.cancel\" | \"polls.errors.question_required\" | \"polls.errors.empty_option\" | \"screenshare\" | \"screenshare.min_preview\" | \"screenshare.max_preview\" | \"screenshare.shared\" | \"screenshare.start\" | \"screenshare.stop\" | \"screenshare.error.unknown\" | \"screenshare.error.max_count\" | \"plugins\" | \"perm_denied\" | \"perm_denied.audio\" | \"perm_denied.video\" | \"perm_denied.screenshare\" | \"perm_denied.audio.chrome.message\" | \"perm_denied.video.chrome.message\" | \"perm_denied.screenshare.chrome.message\" | \"perm_denied.audio.safari.message\" | \"perm_denied.video.safari.message\" | \"perm_denied.screenshare.safari.message\" | \"perm_denied.audio.edge.message\" | \"perm_denied.video.edge.message\" | \"perm_denied.screenshare.edge.message\" | \"perm_denied.audio.firefox.message\" | \"perm_denied.video.firefox.message\" | \"perm_denied.screenshare.firefox.message\" | \"perm_denied.audio.others.message\" | \"perm_denied.video.others.message\" | \"perm_denied.screenshare.others.message\" | \"perm_sys_denied\" | \"perm_sys_denied.audio\" | \"perm_sys_denied.video\" | \"perm_sys_denied.screenshare\" | \"perm_sys_denied.audio.macos.message\" | \"perm_sys_denied.video.macos.message\" | \"perm_sys_denied.screenshare.macos.message\" | \"perm_sys_denied.audio.ios.message\" | \"perm_sys_denied.video.ios.message\" | \"perm_sys_denied.screenshare.ios.message\" | \"perm_sys_denied.audio.windows.message\" | \"perm_sys_denied.video.windows.message\" | \"perm_sys_denied.screenshare.windows.message\" | \"perm_sys_denied.audio.android.message\" | \"perm_sys_denied.video.android.message\" | \"perm_sys_denied.screenshare.android.message\" | \"perm_sys_denied.audio.others.message\" | \"perm_sys_denied.video.others.message\" | \"perm_sys_denied.screenshare.others.message\" | \"perm_could_not_start\" | \"perm_could_not_start.audio\" | \"perm_could_not_start.video\" | \"perm_could_not_start.screenshare\" | \"perm_could_not_start.audio.message\" | \"perm_could_not_start.video.message\" | \"perm_could_not_start.screenshare.message\" | \"full_screen\" | \"full_screen.exit\" | \"waitlist.header_title\" | \"waitlist.body_text\" | \"waitlist.deny_request\" | \"waitlist.accept_request\" | \"waitlist.accept_all\" | \"stage_request.header_title\" | \"stage_request.deny_request\" | \"stage_request.accept_request\" | \"stage_request.accept_all\" | \"stage_request.deny_all\" | \"stage_request.approval_pending\" | \"stage_request.denied\" | \"stage_request.request\" | \"stage_request.requested\" | \"stage_request.cancel_request\" | \"stage_request.leave_stage\" | \"stage_request.request_tip\" | \"stage_request.leave_tip\" | \"stage_request.pending_tip\" | \"stage_request.denied_tip\" | \"stage.empty_host\" | \"stage.empty_host_summary\" | \"stage.empty_viewer\" | \"stage.remove_from_stage\" | \"stage.invited_notification\" | \"stage.add_to_stage\" | \"stage.join_title\" | \"stage.join_summary\" | \"stage.join_cancel\" | \"stage.join_confirm\" | \"setup_screen.join_in_as\" | \"setup_screen.joining_as\" | \"setup_screen.your_name\" | \"stage.reconnecting\" | \"recording.label\" | \"recording.indicator\" | \"recording.started\" | \"recording.stopped\" | \"recording.paused\" | \"recording.error.start\" | \"recording.error.stop\" | \"recording.error.resume\" | \"recording.start\" | \"recording.stop\" | \"recording.resume\" | \"recording.starting\" | \"recording.stopping\" | \"recording.loading\" | \"recording.idle\" | \"audio_playback\" | \"audio_playback.title\" | \"audio_playback.description\" | \"breakout_rooms\" | \"breakout_rooms.room_config_header\" | \"breakout_rooms.join_breakout_header\" | \"breakout_rooms.empty\" | \"breakout_rooms.delete\" | \"breakout_rooms.switch\" | \"breakout_rooms.main_room\" | \"breakout_rooms.shuffle_participants\" | \"breakout_rooms.deselect\" | \"breakout_rooms.selected\" | \"breakout_rooms.num_of_rooms\" | \"breakout_rooms.approx\" | \"breakout_rooms.participants_per_room\" | \"breakout_rooms.division_text\" | \"breakout_rooms.start_breakout\" | \"breakout_rooms.close_breakout\" | \"breakout_rooms.update_breakout\" | \"breakout_rooms.discard_changes\" | \"breakout_rooms.room\" | \"breakout_rooms.rooms\" | \"breakout_rooms.room_name\" | \"breakout_rooms.edit_room_name\" | \"breakout_rooms.save_room_name\" | \"breakout_rooms.add_room\" | \"breakout_rooms.add_room_brief\" | \"breakout_rooms.select_all\" | \"breakout_rooms.unassign_all\" | \"breakout_rooms.assign\" | \"breakout_rooms.assign_participants\" | \"breakout_rooms.none_assigned\" | \"breakout_rooms.drag_drop_participants\" | \"breakout_rooms.click_drop_participants\" | \"breakout_rooms.status.assign_multiple\" | \"breakout_rooms.status.select_room\" | \"breakout_rooms.ephemeral_status.participants_assigned\" | \"breakout_rooms.ephemeral_status.participants_assigned_randomly\" | \"breakout_rooms.ephemeral_status.changes_discarded\" | \"breakout_rooms.confirm_modal.start_breakout.header\" | \"breakout_rooms.confirm_modal.start_breakout.content\" | \"breakout_rooms.confirm_modal.start_breakout.cancelText\" | \"breakout_rooms.confirm_modal.start_breakout.ctaText\" | \"breakout_rooms.confirm_modal.close_breakout.header\" | \"breakout_rooms.confirm_modal.close_breakout.content\" | \"breakout_rooms.confirm_modal.close_breakout.ctaText\" | \"breakout_rooms.move_reason.started_msg\" | \"breakout_rooms.move_reason.started_desc\" | \"breakout_rooms.move_reason.closed_msg\" | \"breakout_rooms.move_reason.closed_desc\" | \"breakout_rooms.move_reason.switch_room\" | \"breakout_rooms.move_reason.switch_main_room\" | \"breakout_rooms.all_assigned\" | \"breakout_rooms.empty_main_room\" | \"breakout_rooms.leave_confirmation\" | \"breakout_rooms.leave_confirmation.main_room_btn\" | \"ai\" | \"ai.meeting_ai\" | \"ai.home\" | \"ai.transcriptions\" | \"ai.personal\" | \"ai.caption_view\" | \"ai.chat.tooltip\" | \"ai.chat.summerise\" | \"ai.chat.agenda\" | \"search\" | \"search.could_not_find\" | \"search.empty\" | \"end\" | \"end.all\" | \"ended.rejected\" | \"ended.left\" | \"ended.kicked\" | \"ended.disconnected\" | \"ended.network\" | \"network\" | \"network.reconnecting\" | \"network.delay\" | \"network.delay_extended\" | \"network.restored\" | \"network.lost\" | \"network.lost_extended\" | \"network.disconnected\" | \"network.leaving\" | \"livestream\" | \"livestream.indicator\" | \"livestream.skip\" | \"livestream.idle\" | \"livestream.starting\" | \"livestream.stopping\" | \"livestream.error.not_supported\" | \"livestream.error.not_found\" | \"livestream.error.unknown\" | \"livestream.error.sync\" | \"livestream.error.start\" | \"livestream.error.stop\" | \"livestream.go_live\" | \"livestream.end_live\" | \"livestream.error\" | \"cta.help\" | \"cta.continue\" | \"cta.reload\" | \"cta.confirmation\" | \"remote_access.empty\" | \"remote_access.requests\" | \"remote_access.allow\" | \"remote_access.grant\" | \"remote_access.indicator\" | \"chat\" | \"chat.new\" | \"chat.new_channel\" | \"chat.channel_name\" | \"chat.member_name\" | \"chat.add_members\" | \"chat.delete_msg\" | \"chat.edit_msg\" | \"chat.send_msg\" | \"chat.send_attachment\" | \"chat.send_img\" | \"chat.send_file\" | \"chat.send_emoji\" | \"chat.update_msg\" | \"chat.channel_members\" | \"chat.img.loading\" | \"chat.error.img_not_found\" | \"chat.error.empty_results\" | \"chat.img.shared_by\" | \"chat.reply\" | \"chat.message_placeholder\" | \"chat.click_to_send\" | \"chat.search_msgs\" | \"chat.search_conversations\" | \"chat.start_conversation\" | \"chat.empty_search\" | \"chat.empty_channel\" | \"chat.cancel_upload\" | \"chat.view_chats\" | \"chat.everyone\" | \"chat.pinned_msgs\" | \"chat.toggle_pinned_msgs\" | \"date.today\" | \"date.yesteday\" | \"date.sunday\" | \"date.monday\" | \"date.tuesday\" | \"date.wednesday\" | \"date.thursday\" | \"date.friday\" | \"date.saturday\" | \"list.empty\" | \"grid.listening\" | \"transcript.off\" | \"transcript.on\" | \"settings.notification_sound\" | \"settings.microphone_input\" | \"settings.speaker_output\" | \"settings.mirror_video\" | \"settings.camera_off\") => string",
          "references": {
            "DyteI18n": {
              "location": "import",
              "path": "../../exports"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Language"
        },
        "defaultValue": "useLanguage()"
      },
      "viewAs": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "'dropdown' | 'list'",
          "resolved": "\"dropdown\" | \"list\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Render as dropdown or list (default = list)"
        },
        "attribute": "view-as",
        "reflect": false,
        "defaultValue": "'list'"
      }
    };
  }
  static get states() {
    return {
      "searchQuery": {},
      "showDropdown": {}
    };
  }
  static get events() {
    return [{
        "method": "channelChange",
        "name": "channelChange",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [],
          "text": "Event emitted when selected channel changes"
        },
        "complexType": {
          "original": "{\n    id: string;\n    name: string;\n    avatarUrl?: string;\n    icon?: keyof IconPack;\n    latestMessage?: string;\n    latestMessageTime?: Date;\n    unreadCount?: number;\n  }",
          "resolved": "{ id: string; name: string; avatarUrl?: string; icon?: \"pin\" | \"disconnected\" | \"dismiss\" | \"settings\" | \"image\" | \"pip_on\" | \"pip_off\" | \"viewers\" | \"mic_off\" | \"mic_on\" | \"video_off\" | \"video_on\" | \"participants\" | \"breakout_rooms\" | \"search\" | \"chat\" | \"add\" | \"delete\" | \"warning\" | \"people\" | \"people_checked\" | \"poll\" | \"rocket\" | \"call_end\" | \"share\" | \"share_screen_start\" | \"share_screen_stop\" | \"share_screen_person\" | \"clock\" | \"send\" | \"more_vertical\" | \"chevron_down\" | \"chevron_up\" | \"chevron_left\" | \"chevron_right\" | \"wifi\" | \"speaker\" | \"speaker_off\" | \"download\" | \"full_screen_maximize\" | \"full_screen_minimize\" | \"copy\" | \"attach\" | \"emoji_multiple\" | \"image_off\" | \"wand\" | \"recording\" | \"subtract\" | \"stop_recording\" | \"pin_off\" | \"spinner\" | \"shuffle\" | \"edit\" | \"back\" | \"save\" | \"web\" | \"checkmark\" | \"spotlight\" | \"join_stage\" | \"leave_stage\" | \"signal_1\" | \"signal_2\" | \"signal_3\" | \"signal_4\" | \"signal_5\" | \"start_livestream\" | \"stop_livestream\" | \"debug\" | \"info\" | \"devices\" | \"horizontal_dots\" | \"ai_sparkle\" | \"meeting_ai\" | \"create_channel\" | \"create_channel_illustration\" | \"captionsOn\" | \"captionsOff\"; latestMessage?: string; latestMessageTime?: Date; unreadCount?: number; }",
          "references": {
            "IconPack": {
              "location": "import",
              "path": "../../exports"
            },
            "Date": {
              "location": "global"
            }
          }
        }
      }];
  }
  static get elementRef() { return "$el"; }
}
