// @dyteinternals/utils/dist/preset/defaultPresetV2.js
import cloneDeep from 'lodash-es/cloneDeep';
import { RecorderType } from '../types';
import { MediaProductionPermissionType, ViewType, WaitingRoomTypes } from '../types/HybridPreset';
import { getDefaultDesignTokens } from './defaultDesignTokens';
const defaultPresetV2 = {
    permissions: {
        // webinar
        can_accept_production_requests: false,
        can_edit_display_name: true,
        accept_waiting_requests: false,
        disable_participant_audio: false,
        disable_participant_screensharing: false,
        disable_participant_video: false,
        can_spotlight: false,
        kick_participant: false,
        pin_participant: false,
        can_record: false,
        can_livestream: false,
        waiting_room_type: WaitingRoomTypes.Skip,
        plugins: {
            can_close: true,
            can_start: true,
            can_edit_config: false,
            config: {},
        },
        polls: {
            can_create: true,
            can_vote: true,
            can_view: true,
        },
        media: {
            video: {
                can_produce: MediaProductionPermissionType.Allowed,
            },
            audio: {
                can_produce: MediaProductionPermissionType.Allowed,
            },
            screenshare: {
                can_produce: MediaProductionPermissionType.Allowed,
            },
        },
        chat: {
            public: {
                can_send: true,
                text: true,
                files: true,
            },
            private: {
                can_send: false,
                can_receive: false,
                text: false,
                files: false,
            },
        },
        hidden_participant: false,
        is_recorder: false,
        recorder_type: RecorderType.none,
        show_participant_list: true,
        transcription_enabled: false,
        can_change_participant_permissions: false,
        connected_meetings: {
            can_alter_connected_meetings: false,
            can_switch_connected_meetings: false,
            can_switch_to_parent_meeting: false,
        },
        stage_enabled: false,
        stage_access: undefined,
        accept_stage_requests: false,
    },
    ui: {
        design_tokens: getDefaultDesignTokens(),
        config_diff: {}
    },
    config: {
        view_type: ViewType.GroupCall,
        media: {
            audio: {
                enable_stereo: false,
                enable_high_bitrate: false,
            },
            video: {
                quality: 'vga',
                frame_rate: 24,
            },
            screenshare: {
                quality: 'hd',
                frame_rate: 5,
            },
        },
        max_video_streams: {
            mobile: 6,
            desktop: 6,
        },
        max_screenshare_count: 1,
    },
    version: '2.0.0',
};
export function getDefaultPresetV2() {
    return cloneDeep(defaultPresetV2);
}
