// @dyteinternals/utils/dist/preset/defaultPresetV1.js
import cloneDeep from 'lodash-es/cloneDeep';
import { RecorderType, WaitingRoomTypesV1 } from '../types';
const defaultPresetV1 = {
    permissions: {
        view_type: null,
        accept_waiting_requests: false,
        accept_present_requests: false,
        request_produce: false,
        can_allow_participant_audio: false,
        can_allow_participant_screensharing: false,
        can_allow_participant_video: false,
        can_spotlight: false,
        kick_participant: false,
        pin_participant: false,
        can_edit_display_name: false,
        can_record: false,
        can_livestream: false,
        recorder_type: RecorderType.none,
        waiting_room_type: WaitingRoomTypesV1.skip,
        max_screenshare_count: 1,
        plugins: {
            can_close: true,
            can_start: true,
            can_edit_acl: false,
            config: {},
        },
        polls: {
            can_create: true,
            can_vote: true,
            can_view: true,
        },
        produce: {
            video: {
                allow: true,
                quality: 'vga',
                frame_rate: 24,
            },
            audio: true,
            screenshare: {
                allow: true,
                quality: 'hd',
                frame_rate: 5,
            },
        },
        chat: {
            public: {
                can_send: true,
                text: true,
                files: true,
            },
            private: {
                can_send: false,
                can_receive: false,
                text: false,
                files: false,
            },
        },
        reactions: false,
        hidden_participant: false,
        is_recorder: false,
        show_participant_list: true,
        can_change_participant_role: false,
        can_change_theme: false,
        can_present: false,
        connected_meetings: {
            can_alter_connected_meetings: false,
            can_switch_connected_meetings: false,
            can_switch_to_parent_meeting: false,
        },
    },
    theme: {
        setup_screen: {
            is_enabled: false,
        },
        alone_here: {
            is_enabled: false,
        },
        waiting_room: {
            is_enabled: false,
            enable_preview: true,
        },
        control_bar: {
            is_enabled: true,
            elements: {
                plugins: true,
                screenshare: true,
                invite: false,
                participants: true,
                chat: true,
                reactions: false,
                polls: true,
                fullscreen: true,
                layout: false,
            },
        },
        header: {
            is_enabled: true,
            elements: {
                timer: true,
                title: true,
                participant_count: true,
                change_layout: true,
            },
        },
        pip_mode: true,
        auto_tune: true,
        colors: {
            primary: '#2160FD',
            secondary: '#1A1A1A',
            text: '#EEEEEE',
            background: '#1A1A1A',
            textPrimary: '#EEEEEE',
            videoBackground: '#1A1A1A',
        },
        dimensions: {
            mode: 'fillParent',
        },
        grid: {
            multi: {
                maxVideoCount: 6,
                videoFit: 'cover',
            },
            single: {
                maxVideoCount: 6,
                videoFit: 'cover',
            },
            defaultView: 'MULTI',
        },
        controls: {
            pip_toggle: false,
        },
        plugins: [],
    },
    version: '0.5.0',
};
export function getDefaultPresetV1() {
    return cloneDeep(defaultPresetV1);
}
