// @dyteinternals/utils/dist/preset/defaultHybridPreset.js
import cloneDeep from 'lodash-es/cloneDeep';
import { RecorderType } from '../types';
import { WaitingRoomTypes, MediaProductionPermissionType, MediaConsumptionPermissionType, ViewType } from '../types/HybridPreset';
import { getDefaultDesignTokens } from './defaultDesignTokens';
const defaultHybridPreset = {
    permissions: {
        // webinar
        can_accept_production_requests: false,
        can_edit_display_name: true,
        accept_waiting_requests: false,
        disable_participant_audio: false,
        disable_participant_screensharing: false,
        disable_participant_video: false,
        can_spotlight: false,
        kick_participant: false,
        pin_participant: false,
        can_record: false,
        can_livestream: false,
        waiting_room_type: WaitingRoomTypes.Skip,
        plugins: {
            can_close: true,
            can_start: true,
            can_edit_config: false,
            config: {},
        },
        polls: {
            can_create: true,
            can_vote: true,
            can_view: true,
        },
        media: {
            video: {
                can_produce: MediaProductionPermissionType.Allowed,
                can_consume: MediaConsumptionPermissionType.Allowed,
            },
            audio: {
                can_produce: MediaProductionPermissionType.Allowed,
            },
            screenshare: {
                can_produce: MediaProductionPermissionType.Allowed,
                can_consume: MediaConsumptionPermissionType.Allowed,
            },
        },
        chat: {
            public: {
                can_send: true,
                text: true,
                files: true,
            },
            private: {
                can_send: false,
                can_receive: false,
                text: false,
                files: false,
            },
            channel: {
                // Make sure these are the same in default to ensure consistency in developer portal
                can_create: 'ALL',
                can_delete: 'ALL',
                can_update: 'ALL',
                // end same
                can_read_all: false,
            },
            message: {
                // Make sure these are the same in default to ensure consistency in developer portal
                can_delete: 'ALL',
                can_edit: 'ALL',
                // end same
                delete_cutoff_time_seconds: 0,
                edit_cutoff_time_seconds: 0,
            }
        },
        hidden_participant: false,
        is_recorder: false,
        recorder_type: RecorderType.none,
        show_participant_list: true,
        transcription_enabled: false,
        can_change_participant_permissions: false,
        connected_meetings: {
            can_alter_connected_meetings: false,
            can_switch_connected_meetings: false,
            can_switch_to_parent_meeting: false,
        },
        stage_enabled: false,
        stage_access: undefined,
        accept_stage_requests: false,
    },
    ui: {
        oldTheme: {
            setup_screen: {
                is_enabled: false,
            },
            alone_here: {
                is_enabled: false,
            },
            waiting_room: {
                is_enabled: false,
                enable_preview: true,
            },
            control_bar: {
                is_enabled: true,
                elements: {
                    plugins: true,
                    screenshare: true,
                    invite: false,
                    participants: true,
                    chat: true,
                    reactions: false,
                    polls: true,
                    fullscreen: true,
                    layout: true,
                },
            },
            header: {
                is_enabled: true,
                elements: {
                    timer: true,
                    title: true,
                    participant_count: true,
                    change_layout: true,
                },
            },
            pip_mode: true,
            auto_tune: true,
            colors: {
                primary: '#2160FD',
                secondary: '#1A1A1A',
                text: '#EEEEEE',
                background: '#1A1A1A',
                textPrimary: '#EEEEEE',
                videoBackground: '#1A1A1A',
            },
            dimensions: {
                mode: 'fillParent',
            },
            grid: {
                multi: {
                    maxVideoCount: 6,
                    videoFit: 'cover',
                },
                single: {
                    maxVideoCount: 6,
                    videoFit: 'cover',
                },
                defaultView: 'MULTI',
            },
            controls: {
                pip_toggle: false,
            },
            plugins: [],
        },
        design_tokens: getDefaultDesignTokens(),
        config_diff: {}
    },
    config: {
        view_type: ViewType.GroupCall,
        media: {
            audio: {
                enable_stereo: false,
                enable_high_bitrate: false,
            },
            video: {
                quality: 'vga',
                frame_rate: 24,
            },
            screenshare: {
                quality: 'hd',
                frame_rate: 5,
            },
        },
        max_video_streams: {
            mobile: 6,
            desktop: 6,
        },
        max_screenshare_count: 1,
    },
    version: 'hybrid',
};
export function getDefaultHybridPreset() {
    return cloneDeep(defaultHybridPreset);
}
