// @dytesdk/ui-kit/dist/collection/components/dyte-debugger-screenshare/dyte-debugger-screenshare.js
import { Host, h } from '@stencil/core';
import { defaultIconPack } from '../../exports';
import { useLanguage } from '../../lib/lang';
import storeState from '../../lib/store';
export class DyteDebuggerAudio {
  constructor() {
    this.lowVolCount = 0;
    this.silenceCount = 0;
    this.frozenFramesCount = 0;
    this.blackFramesCount = 0;
    this.meeting = undefined;
    this.states = undefined;
    this.size = undefined;
    this.iconPack = defaultIconPack;
    this.t = useLanguage();
    this.activeTab = 'Report';
    this.stage = 0;
    this.mediaPermission = 'NOT_REQUESTED';
    this.screenshareEnabled = false;
    this.audioTrackMetadata = {};
    this.videoTrackMetadata = {};
    this.screenshareTest = false;
    this.audioTrackStats = {
      lowVolume: false,
      clip: false,
      silence: false,
      volume: 0,
      timestamp: 0,
    };
    this.videoTrackStats = {
      isBlackFrame: false,
      isFrozenFrame: false,
      timestamp: 0,
      videoScore: 0,
    };
  }
  connectedCallback() {
    this.meetingChanged(this.meeting);
    this.updateProgress();
  }
  disconnectedCallback() {
    var _a, _b;
    const { self, troubleshoot } = this.meeting;
    this.screenshareTest && this.toggleScreenshareTest();
    self === null || self === void 0 ? void 0 : self.removeListener('mediaPermissionUpdate', this.mediaPermissionUpdateListener);
    self.removeListener('screenShareUpdate', this.screenShareUpdateListener);
    (_a = troubleshoot === null || troubleshoot === void 0 ? void 0 : troubleshoot.screenshare) === null || _a === void 0 ? void 0 : _a.removeListener('audioTrackStats', this.audioTrackUpdateListener);
    (_b = troubleshoot === null || troubleshoot === void 0 ? void 0 : troubleshoot.screenshare) === null || _b === void 0 ? void 0 : _b.removeListener('videoTrackStats', this.videoTrackUpdateListener);
  }
  async meetingChanged(meeting) {
    var _a, _b, _c;
    if (!meeting)
      return;
    const { self, troubleshoot } = meeting;
    self.addListener('mediaPermissionUpdate', this.mediaPermissionUpdateListener.bind(this));
    self.addListener('screenShareUpdate', this.screenShareUpdateListener.bind(this));
    (_a = troubleshoot === null || troubleshoot === void 0 ? void 0 : troubleshoot.screenshare) === null || _a === void 0 ? void 0 : _a.on('audioTrackStats', this.audioTrackUpdateListener.bind(this));
    (_b = troubleshoot === null || troubleshoot === void 0 ? void 0 : troubleshoot.screenshare) === null || _b === void 0 ? void 0 : _b.on('videoTrackStats', this.videoTrackUpdateListener.bind(this));
    const permission = (_c = self === null || self === void 0 ? void 0 : self.mediaPermissions) === null || _c === void 0 ? void 0 : _c.screenshare;
    this.mediaPermissionUpdateListener({ kind: 'screenshare', message: permission });
    const screenShareEnabled = self === null || self === void 0 ? void 0 : self.screenShareEnabled;
    this.screenShareUpdateListener({ screenShareEnabled });
  }
  mediaPermissionUpdateListener({ kind, message }) {
    if (kind !== 'screenshare')
      return;
    this.mediaPermission = message;
    if (this.mediaPermission !== 'ACCEPTED') {
      this.stage = 0;
      this.updateProgress();
    }
  }
  screenShareUpdateListener({ screenShareEnabled }) {
    this.screenshareEnabled = screenShareEnabled;
    if (!this.screenshareEnabled && this.screenshareTest)
      this.toggleScreenshareTest();
  }
  audioTrackUpdateListener(stats) {
    this.audioTrackStats = Object.assign(Object.assign({}, stats), { volume: this.normalizeVolume(stats === null || stats === void 0 ? void 0 : stats.volume) });
  }
  videoTrackUpdateListener(stats) {
    this.videoTrackStats = Object.assign(Object.assign({}, stats), { videoScore: this.normalizeVideoScore(stats === null || stats === void 0 ? void 0 : stats.videoScore) });
  }
  async toggleScreenshareTest() {
    var _a, _b, _c;
    this.screenshareTest = !this.screenshareTest;
    const troubleshooter = (_a = this.meeting) === null || _a === void 0 ? void 0 : _a.troubleshoot.screenshare;
    if (this.screenshareTest && this.screenshareEnabled) {
      this.audioTrackMetadata = (_b = troubleshooter === null || troubleshooter === void 0 ? void 0 : troubleshooter.getAudioTrackMetaData()) !== null && _b !== void 0 ? _b : {};
      this.videoTrackMetadata = (_c = troubleshooter === null || troubleshooter === void 0 ? void 0 : troubleshooter.getVideoTrackMetaData()) !== null && _c !== void 0 ? _c : {};
      await troubleshooter.startAudioTrackAnalysis();
      await troubleshooter.startVideoTrackAnalysis();
    }
    else {
      troubleshooter === null || troubleshooter === void 0 ? void 0 : troubleshooter.stopAudioTrackAnalysis();
      troubleshooter === null || troubleshooter === void 0 ? void 0 : troubleshooter.stopVideoTrackAnalysis();
    }
  }
  setActiveTab(state) {
    this.activeTab = state;
    this.updateProgress();
  }
  changeStage(stage) {
    if (this.mediaPermission !== 'ACCEPTED')
      return;
    stage = Math.max(0, stage);
    stage = Math.min(1, stage);
    this.stage = stage;
    this.updateProgress();
    if (this.screenshareTest)
      this.toggleScreenshareTest();
  }
  updateProgress() {
    if (this.activeTab === 'Troubleshoot') {
      this.testProgress.emit(((this.stage + 1) * 100) / 2);
    }
    else
      this.testProgress.emit(0);
  }
  getDeviceInformationStatus() {
    let message = 'All systems are functional.';
    let icon = 'checkmark';
    let style = 'success';
    if (this.audioTrackMetadata) {
      const { channelCount: count, channelInterpretation: mode, channelType: type, outputs, } = this.audioTrackMetadata;
      if (mode === 'discrete' && outputs > count) {
        message = 'Audio quality might be degraded.';
        icon = 'warning';
        style = 'warning';
      }
      else if ((type === 'STEREO' && outputs === 1) ||
        (type === 'QUAD' && outputs === 1) ||
        (type === 'QUAD' && outputs === 2)) {
        message = 'Audio quality might be degraded.';
        icon = 'warning';
        style = 'warning';
      }
      if ((type === 'MONO' && count < 1) ||
        (type === 'STEREO' && count < 2) ||
        (type === 'QUAD' && count < 4)) {
        message = 'Looks like your device is not functioning properly.';
        icon = 'warning';
        style = 'error';
      }
    }
    return (h("div", { class: `row status-text ${style}` }, h("dyte-icon", { size: "sm", icon: this.iconPack[icon] }), this.t(message)));
  }
  getAudioAnalysisStatus() {
    var _a;
    let message = 'All systems are functional.';
    let icon = 'checkmark';
    let style = 'success';
    if (!((_a = this.audioTrackMetadata) === null || _a === void 0 ? void 0 : _a.label))
      return (h("div", { class: `row status-text warning` }, h("dyte-icon", { size: "sm", icon: this.iconPack.warning }), this.t('Not Available')));
    if (this.detectLowVolume()) {
      message = 'Low volume detected.';
      icon = 'warning';
      style = 'warning';
    }
    if (this.audioTrackStats.clip) {
      message = 'Your audio may be noisy.';
      icon = 'warning';
      style = 'error';
    }
    if (this.detectSilence()) {
      message = 'Silence detected.';
      icon = 'warning';
      style = 'error';
    }
    return (h("div", { class: `row status-text ${style}` }, h("dyte-icon", { size: "sm", icon: this.iconPack[icon] }), this.t(message)));
  }
  getVideoAnalysisStatus() {
    var _a, _b;
    let message = 'All systems are functional.';
    let icon = 'checkmark';
    let style = 'success';
    if ((_a = this.videoTrackStats) === null || _a === void 0 ? void 0 : _a.isBlackFrame) {
      this.blackFramesCount++;
    }
    else {
      this.blackFramesCount = 0;
    }
    if ((_b = this.videoTrackStats) === null || _b === void 0 ? void 0 : _b.isFrozenFrame) {
      this.frozenFramesCount++;
    }
    else {
      this.frozenFramesCount = 0;
    }
    if (this.frozenFramesCount > 4) {
      message = 'Looks like the video is frozen.';
      icon = 'warning';
      style = 'error';
    }
    if (this.blackFramesCount > 4) {
      message = 'Video is not being processed by the camera.';
      icon = 'warning';
      style = 'error';
    }
    return (h("div", { class: `row status-text ${style}` }, h("dyte-icon", { size: "sm", icon: this.iconPack[icon] }), this.t(message)));
  }
  formatLabel(camelCaseLabel) {
    const result = camelCaseLabel.replace(/([A-Z])/g, ' $1');
    return result.charAt(0).toUpperCase() + result.slice(1);
  }
  nullToStr(val) {
    if (!val)
      return 'Not Available';
    return val;
  }
  normalizeVolume(val = 0) {
    let max = 0.3;
    const min = 0.001;
    if (val > max)
      max = val;
    return Math.max(5, Math.round((val * 100) / (max - min)));
  }
  normalizeVideoScore(val = 0) {
    let max = 1;
    const min = 0;
    if (val > max)
      max = val;
    return Math.max(5, Math.round((val * 10) / (max - min)));
  }
  detectLowVolume() {
    var _a;
    if ((_a = this.audioTrackStats) === null || _a === void 0 ? void 0 : _a.lowVolume) {
      this.lowVolCount++;
    }
    else {
      this.lowVolCount = 0;
    }
    return this.lowVolCount > 10;
  }
  detectSilence() {
    var _a;
    if ((_a = this.audioTrackStats) === null || _a === void 0 ? void 0 : _a.silence) {
      this.silenceCount++;
    }
    else {
      this.silenceCount = 0;
    }
    return this.silenceCount > 5;
  }
  render() {
    var _a, _b;
    const defaults = {
      meeting: this.meeting,
      states: this.states || storeState,
      iconPack: this.iconPack,
      t: this.t,
      size: this.size,
    };
    return (h(Host, null, h("div", { id: "header" }, ['Report', 'Troubleshoot'].map((section) => (h("dyte-button", { key: section, variant: "ghost", class: { active: this.activeTab === section }, onClick: () => this.setActiveTab(section), iconPack: this.iconPack, t: this.t }, this.t(section))))), this.activeTab === 'Troubleshoot' && (h("div", { class: "tab-body" }, h("div", null, h("div", { class: "stage-indicator" }, "Step ", this.stage + 1, "/2"), this.stage === 0 && (h("dyte-debugger-permissions-ui", Object.assign({}, defaults, { mediaType: "screenshare" }))), this.stage === 1 &&
      (this.screenshareEnabled ? (h("div", { class: "col" }, h("div", { class: "text" }, "Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots."), h("dyte-button", { onClick: () => this.toggleScreenshareTest() }, this.screenshareTest ? 'Stop' : 'Start', " Screenshare Tests"), this.screenshareTest && (h("div", null, h("div", { class: "sub-title" }, "Device Information", h("dyte-information-tooltip", null, h("div", { slot: "tootlip-text", class: "tooltip-text" }, Object.keys(this.audioTrackMetadata).map((key) => {
        if (key === 'label')
          return;
        return (h("div", null, "Audio ", this.formatLabel(key), ":", ' ', h("span", null, this.audioTrackMetadata[key])));
      }), Object.keys(this.videoTrackMetadata).map((key) => {
        if (key === 'label')
          return;
        return (h("div", null, "Video ", this.formatLabel(key), ":", ' ', h("span", null, JSON.stringify(this.videoTrackMetadata[key]))));
      })))), h("div", { class: "row align-start" }, h("div", { class: "label" }, h("span", null, "Audio Label"), ":"), h("div", { class: "text" }, this.nullToStr((_a = this.audioTrackMetadata) === null || _a === void 0 ? void 0 : _a.label))), h("div", { class: "row align-start" }, h("div", { class: "label" }, h("span", null, "Video Label"), ":"), h("div", { class: "text" }, this.nullToStr((_b = this.videoTrackMetadata) === null || _b === void 0 ? void 0 : _b.label))), this.getDeviceInformationStatus(), h("div", { class: "sub-title" }, "Device Analysis"), h("div", { class: "row " }, h("div", { class: "label" }, h("span", null, "Audio Analysis"), " :"), this.getAudioAnalysisStatus()), h("div", { class: "row" }, h("div", { class: "label" }, h("span", null, "Video Analysis"), " :"), this.getVideoAnalysisStatus()))))) : (h("div", { class: "col" }, h("div", { class: "info" }, h("dyte-icon", { size: "sm", icon: this.iconPack.warning }), "You are not sharing your screen."), h("div", { class: "text" }, "In order for us to test your screenshare, you need to be sharing your screen. You can enable screenshare from the control bar."))))), h("div", { class: "stage-manager" }, this.stage < 1 && (h("div", { class: { 'stage-button': true, disabled: this.mediaPermission !== 'ACCEPTED' }, onClick: () => this.changeStage(this.stage + 1) }, "Next", h("dyte-icon", { size: "sm", icon: this.iconPack.chevron_right }))), this.stage > 0 && (h("div", { class: { 'stage-button': true, disabled: this.mediaPermission !== 'ACCEPTED' }, onClick: () => this.changeStage(this.stage - 1) }, h("dyte-icon", { size: "sm", icon: this.iconPack.chevron_left }), "Prev"))))), this.activeTab === 'Report' && (h("div", { class: "tab-body" }, h("dyte-debugger-issues-ui", Object.assign({}, defaults, { mediaType: "screenshare" }))))));
  }
  static get is() { return "dyte-debugger-screenshare"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["dyte-debugger-screenshare.css"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dyte-debugger-screenshare.css"]
    };
  }
  static get properties() {
    return {
      "meeting": {
        "type": "unknown",
        "mutable": false,
        "complexType": {
          "original": "Meeting",
          "resolved": "DyteClient",
          "references": {
            "Meeting": {
              "location": "import",
              "path": "../../types/dyte-client"
            }
          }
        },
        "required": true,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Meeting object"
        }
      },
      "states": {
        "type": "unknown",
        "mutable": false,
        "complexType": {
          "original": "States",
          "resolved": "States",
          "references": {
            "States": {
              "location": "import",
              "path": "../../exports"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "States object"
        }
      },
      "size": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "Size",
          "resolved": "\"lg\" | \"md\" | \"sm\" | \"xl\"",
          "references": {
            "Size": {
              "location": "import",
              "path": "../../exports"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Size"
        },
        "attribute": "size",
        "reflect": true
      },
      "iconPack": {
        "type": "unknown",
        "mutable": false,
        "complexType": {
          "original": "IconPack",
          "resolved": "{ people: string; people_checked: string; chat: string; poll: string; participants: string; rocket: string; call_end: string; share: string; mic_on: string; mic_off: string; video_on: string; video_off: string; share_screen_start: string; share_screen_stop: string; share_screen_person: string; clock: string; dismiss: string; send: string; search: string; more_vertical: string; chevron_down: string; chevron_up: string; chevron_left: string; chevron_right: string; settings: string; wifi: string; speaker: string; speaker_off: string; download: string; full_screen_maximize: string; full_screen_minimize: string; copy: string; attach: string; image: string; emoji_multiple: string; image_off: string; disconnected: string; wand: string; recording: string; subtract: string; stop_recording: string; warning: string; pin: string; pin_off: string; spinner: string; breakout_rooms: string; add: string; shuffle: string; edit: string; delete: string; back: string; save: string; web: string; checkmark: string; spotlight: string; join_stage: string; leave_stage: string; pip_off: string; pip_on: string; signal_1: string; signal_2: string; signal_3: string; signal_4: string; signal_5: string; start_livestream: string; stop_livestream: string; viewers: string; debug: string; info: string; devices: string; horizontal_dots: string; ai_sparkle: string; meeting_ai: string; create_channel: string; create_channel_illustration: string; captionsOn: string; captionsOff: string; }",
          "references": {
            "IconPack": {
              "location": "import",
              "path": "../../exports"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Icon pack"
        },
        "defaultValue": "defaultIconPack"
      },
      "t": {
        "type": "unknown",
        "mutable": false,
        "complexType": {
          "original": "DyteI18n",
          "resolved": "(key: \"pin\" | \"unpin\" | \"kick\" | \"ended\" | \"disconnected\" | (string & {}) | \"about_call\" | \"screen\" | \"camera\" | \"leave\" | \"dismiss\" | \"logo\" | \"page\" | \"more\" | \"page.prev\" | \"page.next\" | \"layout\" | \"layout.auto\" | \"settings\" | \"file\" | \"image\" | \"connection\" | \"leave_confirmation\" | \"cancel\" | \"yes\" | \"you\" | \"to\" | \"mute\" | \"pinned\" | \"pip_on\" | \"pip_off\" | \"viewers\" | \"join\" | \"joined\" | \"create\" | \"close\" | \"ask\" | \"type\" | \"activate\" | \"mic_off\" | \"disable_mic\" | \"mic_on\" | \"enable_mic\" | \"audio\" | \"test\" | \"mute_all\" | \"mute_all.description\" | \"mute_all.header\" | \"mute_all.allow_unmute\" | \"video_off\" | \"disable_video\" | \"video_on\" | \"enable_video\" | \"video\" | \"offline\" | \"offline.description\" | \"disconnected.description\" | \"participants\" | \"participants.errors.empty_results\" | \"participants.empty_list\" | \"participants.turn_off_video\" | \"polls\" | \"polls.by\" | \"polls.question\" | \"polls.question.placeholder\" | \"polls.answers\" | \"polls.option\" | \"polls.option.placeholder\" | \"polls.results.anon\" | \"polls.results.hide\" | \"polls.create\" | \"polls.cancel\" | \"polls.errors.question_required\" | \"polls.errors.empty_option\" | \"screenshare\" | \"screenshare.min_preview\" | \"screenshare.max_preview\" | \"screenshare.shared\" | \"screenshare.start\" | \"screenshare.stop\" | \"screenshare.error.unknown\" | \"screenshare.error.max_count\" | \"plugins\" | \"perm_denied\" | \"perm_denied.audio\" | \"perm_denied.video\" | \"perm_denied.screenshare\" | \"perm_denied.audio.chrome.message\" | \"perm_denied.video.chrome.message\" | \"perm_denied.screenshare.chrome.message\" | \"perm_denied.audio.safari.message\" | \"perm_denied.video.safari.message\" | \"perm_denied.screenshare.safari.message\" | \"perm_denied.audio.edge.message\" | \"perm_denied.video.edge.message\" | \"perm_denied.screenshare.edge.message\" | \"perm_denied.audio.firefox.message\" | \"perm_denied.video.firefox.message\" | \"perm_denied.screenshare.firefox.message\" | \"perm_denied.audio.others.message\" | \"perm_denied.video.others.message\" | \"perm_denied.screenshare.others.message\" | \"perm_sys_denied\" | \"perm_sys_denied.audio\" | \"perm_sys_denied.video\" | \"perm_sys_denied.screenshare\" | \"perm_sys_denied.audio.macos.message\" | \"perm_sys_denied.video.macos.message\" | \"perm_sys_denied.screenshare.macos.message\" | \"perm_sys_denied.audio.ios.message\" | \"perm_sys_denied.video.ios.message\" | \"perm_sys_denied.screenshare.ios.message\" | \"perm_sys_denied.audio.windows.message\" | \"perm_sys_denied.video.windows.message\" | \"perm_sys_denied.screenshare.windows.message\" | \"perm_sys_denied.audio.android.message\" | \"perm_sys_denied.video.android.message\" | \"perm_sys_denied.screenshare.android.message\" | \"perm_sys_denied.audio.others.message\" | \"perm_sys_denied.video.others.message\" | \"perm_sys_denied.screenshare.others.message\" | \"perm_could_not_start\" | \"perm_could_not_start.audio\" | \"perm_could_not_start.video\" | \"perm_could_not_start.screenshare\" | \"perm_could_not_start.audio.message\" | \"perm_could_not_start.video.message\" | \"perm_could_not_start.screenshare.message\" | \"full_screen\" | \"full_screen.exit\" | \"waitlist.header_title\" | \"waitlist.body_text\" | \"waitlist.deny_request\" | \"waitlist.accept_request\" | \"waitlist.accept_all\" | \"stage_request.header_title\" | \"stage_request.deny_request\" | \"stage_request.accept_request\" | \"stage_request.accept_all\" | \"stage_request.deny_all\" | \"stage_request.approval_pending\" | \"stage_request.denied\" | \"stage_request.request\" | \"stage_request.requested\" | \"stage_request.cancel_request\" | \"stage_request.leave_stage\" | \"stage_request.request_tip\" | \"stage_request.leave_tip\" | \"stage_request.pending_tip\" | \"stage_request.denied_tip\" | \"stage.empty_host\" | \"stage.empty_host_summary\" | \"stage.empty_viewer\" | \"stage.remove_from_stage\" | \"stage.invited_notification\" | \"stage.add_to_stage\" | \"stage.join_title\" | \"stage.join_summary\" | \"stage.join_cancel\" | \"stage.join_confirm\" | \"setup_screen.join_in_as\" | \"setup_screen.joining_as\" | \"setup_screen.your_name\" | \"stage.reconnecting\" | \"recording.label\" | \"recording.indicator\" | \"recording.started\" | \"recording.stopped\" | \"recording.paused\" | \"recording.error.start\" | \"recording.error.stop\" | \"recording.error.resume\" | \"recording.start\" | \"recording.stop\" | \"recording.resume\" | \"recording.starting\" | \"recording.stopping\" | \"recording.loading\" | \"recording.idle\" | \"audio_playback\" | \"audio_playback.title\" | \"audio_playback.description\" | \"breakout_rooms\" | \"breakout_rooms.room_config_header\" | \"breakout_rooms.join_breakout_header\" | \"breakout_rooms.empty\" | \"breakout_rooms.delete\" | \"breakout_rooms.switch\" | \"breakout_rooms.main_room\" | \"breakout_rooms.shuffle_participants\" | \"breakout_rooms.deselect\" | \"breakout_rooms.selected\" | \"breakout_rooms.num_of_rooms\" | \"breakout_rooms.approx\" | \"breakout_rooms.participants_per_room\" | \"breakout_rooms.division_text\" | \"breakout_rooms.start_breakout\" | \"breakout_rooms.close_breakout\" | \"breakout_rooms.update_breakout\" | \"breakout_rooms.discard_changes\" | \"breakout_rooms.room\" | \"breakout_rooms.rooms\" | \"breakout_rooms.room_name\" | \"breakout_rooms.edit_room_name\" | \"breakout_rooms.save_room_name\" | \"breakout_rooms.add_room\" | \"breakout_rooms.add_room_brief\" | \"breakout_rooms.select_all\" | \"breakout_rooms.unassign_all\" | \"breakout_rooms.assign\" | \"breakout_rooms.assign_participants\" | \"breakout_rooms.none_assigned\" | \"breakout_rooms.drag_drop_participants\" | \"breakout_rooms.click_drop_participants\" | \"breakout_rooms.status.assign_multiple\" | \"breakout_rooms.status.select_room\" | \"breakout_rooms.ephemeral_status.participants_assigned\" | \"breakout_rooms.ephemeral_status.participants_assigned_randomly\" | \"breakout_rooms.ephemeral_status.changes_discarded\" | \"breakout_rooms.confirm_modal.start_breakout.header\" | \"breakout_rooms.confirm_modal.start_breakout.content\" | \"breakout_rooms.confirm_modal.start_breakout.cancelText\" | \"breakout_rooms.confirm_modal.start_breakout.ctaText\" | \"breakout_rooms.confirm_modal.close_breakout.header\" | \"breakout_rooms.confirm_modal.close_breakout.content\" | \"breakout_rooms.confirm_modal.close_breakout.ctaText\" | \"breakout_rooms.move_reason.started_msg\" | \"breakout_rooms.move_reason.started_desc\" | \"breakout_rooms.move_reason.closed_msg\" | \"breakout_rooms.move_reason.closed_desc\" | \"breakout_rooms.move_reason.switch_room\" | \"breakout_rooms.move_reason.switch_main_room\" | \"breakout_rooms.all_assigned\" | \"breakout_rooms.empty_main_room\" | \"breakout_rooms.leave_confirmation\" | \"breakout_rooms.leave_confirmation.main_room_btn\" | \"ai\" | \"ai.meeting_ai\" | \"ai.home\" | \"ai.transcriptions\" | \"ai.personal\" | \"ai.caption_view\" | \"ai.chat.tooltip\" | \"ai.chat.summerise\" | \"ai.chat.agenda\" | \"search\" | \"search.could_not_find\" | \"search.empty\" | \"end\" | \"end.all\" | \"ended.rejected\" | \"ended.left\" | \"ended.kicked\" | \"ended.disconnected\" | \"ended.network\" | \"network\" | \"network.reconnecting\" | \"network.delay\" | \"network.delay_extended\" | \"network.restored\" | \"network.lost\" | \"network.lost_extended\" | \"network.disconnected\" | \"network.leaving\" | \"livestream\" | \"livestream.indicator\" | \"livestream.skip\" | \"livestream.idle\" | \"livestream.starting\" | \"livestream.stopping\" | \"livestream.error.not_supported\" | \"livestream.error.not_found\" | \"livestream.error.unknown\" | \"livestream.error.sync\" | \"livestream.error.start\" | \"livestream.error.stop\" | \"livestream.go_live\" | \"livestream.end_live\" | \"livestream.error\" | \"cta.help\" | \"cta.continue\" | \"cta.reload\" | \"cta.confirmation\" | \"remote_access.empty\" | \"remote_access.requests\" | \"remote_access.allow\" | \"remote_access.grant\" | \"remote_access.indicator\" | \"chat\" | \"chat.new\" | \"chat.new_channel\" | \"chat.channel_name\" | \"chat.member_name\" | \"chat.add_members\" | \"chat.delete_msg\" | \"chat.edit_msg\" | \"chat.send_msg\" | \"chat.send_attachment\" | \"chat.send_img\" | \"chat.send_file\" | \"chat.send_emoji\" | \"chat.update_msg\" | \"chat.channel_members\" | \"chat.img.loading\" | \"chat.error.img_not_found\" | \"chat.error.empty_results\" | \"chat.img.shared_by\" | \"chat.reply\" | \"chat.message_placeholder\" | \"chat.click_to_send\" | \"chat.search_msgs\" | \"chat.search_conversations\" | \"chat.start_conversation\" | \"chat.empty_search\" | \"chat.empty_channel\" | \"chat.cancel_upload\" | \"chat.view_chats\" | \"chat.everyone\" | \"chat.pinned_msgs\" | \"chat.toggle_pinned_msgs\" | \"date.today\" | \"date.yesteday\" | \"date.sunday\" | \"date.monday\" | \"date.tuesday\" | \"date.wednesday\" | \"date.thursday\" | \"date.friday\" | \"date.saturday\" | \"list.empty\" | \"grid.listening\" | \"transcript.off\" | \"transcript.on\" | \"settings.notification_sound\" | \"settings.microphone_input\" | \"settings.speaker_output\" | \"settings.mirror_video\" | \"settings.camera_off\") => string",
          "references": {
            "DyteI18n": {
              "location": "import",
              "path": "../../exports"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Language"
        },
        "defaultValue": "useLanguage()"
      }
    };
  }
  static get states() {
    return {
      "activeTab": {},
      "stage": {},
      "mediaPermission": {},
      "screenshareEnabled": {},
      "audioTrackMetadata": {},
      "videoTrackMetadata": {},
      "screenshareTest": {},
      "audioTrackStats": {},
      "videoTrackStats": {}
    };
  }
  static get events() {
    return [{
        "method": "testProgress",
        "name": "testProgress",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [],
          "text": "Test Progress Event"
        },
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        }
      }];
  }
  static get watchers() {
    return [{
        "propName": "meeting",
        "methodName": "meetingChanged"
      }];
  }
}
