// @dytesdk/docs-ai-react/dist/index.esm.js
import React, { useRef, useState, useEffect, Fragment } from 'react';
import ReactMarkdown from 'react-markdown';
import remarkGfm from 'remark-gfm';
import { Transition, Dialog } from '@headlessui/react';

function styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') { return; }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = "/*! tailwindcss v3.3.2 | MIT License | https://tailwindcss.com*/*,:after,:before{border:0 solid #e5e7eb;box-sizing:border-box}:after,:before{--tw-content:\"\"}html{-webkit-text-size-adjust:100%;font-feature-settings:normal;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,Noto Sans,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol,Noto Color Emoji;font-variation-settings:normal;line-height:1.5;-moz-tab-size:4;-o-tab-size:4;tab-size:4}body{line-height:inherit;margin:0}hr{border-top-width:1px;color:inherit;height:0}abbr:where([title]){-webkit-text-decoration:underline dotted;text-decoration:underline dotted}h1,h2,h3,h4,h5,h6{font-size:inherit;font-weight:inherit}a{color:inherit;text-decoration:inherit}b,strong{font-weight:bolder}code,kbd,pre,samp{font-family:ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,Liberation Mono,Courier New,monospace;font-size:1em}small{font-size:80%}sub,sup{font-size:75%;line-height:0;position:relative;vertical-align:baseline}sub{bottom:-.25em}sup{top:-.5em}table{border-collapse:collapse;border-color:inherit;text-indent:0}button,input,optgroup,select,textarea{color:inherit;font-family:inherit;font-size:100%;font-weight:inherit;line-height:inherit;margin:0;padding:0}button,select{text-transform:none}[type=button],[type=reset],[type=submit],button{-webkit-appearance:button;background-color:transparent;background-image:none}:-moz-focusring{outline:auto}:-moz-ui-invalid{box-shadow:none}progress{vertical-align:baseline}::-webkit-inner-spin-button,::-webkit-outer-spin-button{height:auto}[type=search]{-webkit-appearance:textfield;outline-offset:-2px}::-webkit-search-decoration{-webkit-appearance:none}::-webkit-file-upload-button{-webkit-appearance:button;font:inherit}summary{display:list-item}blockquote,dd,dl,figure,h1,h2,h3,h4,h5,h6,hr,p,pre{margin:0}fieldset{margin:0}fieldset,legend{padding:0}menu,ol,ul{list-style:none;margin:0;padding:0}textarea{resize:vertical}input::-moz-placeholder,textarea::-moz-placeholder{color:#9ca3af;opacity:1}input::placeholder,textarea::placeholder{color:#9ca3af;opacity:1}[role=button],button{cursor:pointer}:disabled{cursor:default}audio,canvas,embed,iframe,img,object,svg,video{display:block;vertical-align:middle}img,video{height:auto;max-width:100%}[hidden]{display:none}.markdown-container p{display:block;margin:1em 0}.markdown-container a{color:(internal value);cursor:pointer;text-decoration:underline}.markdown-container ol{display:block;list-style-type:decimal;margin:1em 0;padding-left:40px}.markdown-container ul{display:block;list-style-type:disc;margin:1em 0 1 em;padding-left:40px}.markdown-container ol li{display:list-item}.markdown-container pre{--tw-bg-opacity:1;background-color:rgb(225 226 227/var(--tw-bg-opacity));border-radius:8px;display:block;font-family:monospace;margin:1em 0;overflow:auto;padding:10px;white-space:pre}:is([data-theme=dark] .markdown-container pre){--tw-bg-opacity:1;background-color:rgb(18 18 18/var(--tw-bg-opacity))}.markdown-container table{border-collapse:separate;border-color:gray;border-spacing:2px;display:table}.markdown-container tbody{border-color:inherit;display:table-row-group;vertical-align:middle}.markdown-container thead{border-color:inherit;display:table-header-group;vertical-align:middle}.markdown-container td,.markdown-container th{display:table-cell;vertical-align:inherit}.markdown-container th{font-weight:700;text-align:center}.markdown-container tr{border-color:inherit;display:table-row;vertical-align:inherit}*,:after,:before{--tw-border-spacing-x:0;--tw-border-spacing-y:0;--tw-translate-x:0;--tw-translate-y:0;--tw-rotate:0;--tw-skew-x:0;--tw-skew-y:0;--tw-scale-x:1;--tw-scale-y:1;--tw-pan-x: ;--tw-pan-y: ;--tw-pinch-zoom: ;--tw-scroll-snap-strictness:proximity;--tw-gradient-from-position: ;--tw-gradient-via-position: ;--tw-gradient-to-position: ;--tw-ordinal: ;--tw-slashed-zero: ;--tw-numeric-figure: ;--tw-numeric-spacing: ;--tw-numeric-fraction: ;--tw-ring-inset: ;--tw-ring-offset-width:0px;--tw-ring-offset-color:#fff;--tw-ring-color:rgba(59,130,246,.5);--tw-ring-offset-shadow:0 0 #0000;--tw-ring-shadow:0 0 #0000;--tw-shadow:0 0 #0000;--tw-shadow-colored:0 0 #0000;--tw-blur: ;--tw-brightness: ;--tw-contrast: ;--tw-grayscale: ;--tw-hue-rotate: ;--tw-invert: ;--tw-saturate: ;--tw-sepia: ;--tw-drop-shadow: ;--tw-backdrop-blur: ;--tw-backdrop-brightness: ;--tw-backdrop-contrast: ;--tw-backdrop-grayscale: ;--tw-backdrop-hue-rotate: ;--tw-backdrop-invert: ;--tw-backdrop-opacity: ;--tw-backdrop-saturate: ;--tw-backdrop-sepia: }::backdrop{--tw-border-spacing-x:0;--tw-border-spacing-y:0;--tw-translate-x:0;--tw-translate-y:0;--tw-rotate:0;--tw-skew-x:0;--tw-skew-y:0;--tw-scale-x:1;--tw-scale-y:1;--tw-pan-x: ;--tw-pan-y: ;--tw-pinch-zoom: ;--tw-scroll-snap-strictness:proximity;--tw-gradient-from-position: ;--tw-gradient-via-position: ;--tw-gradient-to-position: ;--tw-ordinal: ;--tw-slashed-zero: ;--tw-numeric-figure: ;--tw-numeric-spacing: ;--tw-numeric-fraction: ;--tw-ring-inset: ;--tw-ring-offset-width:0px;--tw-ring-offset-color:#fff;--tw-ring-color:rgba(59,130,246,.5);--tw-ring-offset-shadow:0 0 #0000;--tw-ring-shadow:0 0 #0000;--tw-shadow:0 0 #0000;--tw-shadow-colored:0 0 #0000;--tw-blur: ;--tw-brightness: ;--tw-contrast: ;--tw-grayscale: ;--tw-hue-rotate: ;--tw-invert: ;--tw-saturate: ;--tw-sepia: ;--tw-drop-shadow: ;--tw-backdrop-blur: ;--tw-backdrop-brightness: ;--tw-backdrop-contrast: ;--tw-backdrop-grayscale: ;--tw-backdrop-hue-rotate: ;--tw-backdrop-invert: ;--tw-backdrop-opacity: ;--tw-backdrop-saturate: ;--tw-backdrop-sepia: }.fixed{position:fixed}.relative{position:relative}.inset-0{inset:0}.z-10{z-index:10}.mx-auto{margin-left:auto;margin-right:auto}.my-4{margin-bottom:1rem;margin-top:1rem}.mr-2{margin-right:.5rem}.mt-2{margin-top:.5rem}.mt-4{margin-top:1rem}.inline{display:inline}.flex{display:flex}.h-4{height:1rem}.h-6{height:1.5rem}.h-9{height:2.25rem}.h-full{height:100%}.max-h-\\[520px\\]{max-height:520px}.w-4{width:1rem}.w-48{width:12rem}.w-6{width:1.5rem}.w-9{width:2.25rem}.w-full{width:100%}.max-w-2xl{max-width:42rem}.flex-1{flex:1 1 0%}.scale-100{--tw-scale-x:1;--tw-scale-y:1}.scale-100,.scale-95{transform:translate(var(--tw-translate-x),var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y))}.scale-95{--tw-scale-x:.95;--tw-scale-y:.95}.transform{transform:translate(var(--tw-translate-x),var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y))}.cursor-text{cursor:text}.flex-col{flex-direction:column}.items-center{align-items:center}.justify-center{justify-content:center}.justify-between{justify-content:space-between}.overflow-auto{overflow:auto}.overflow-hidden{overflow:hidden}.overflow-y-auto{overflow-y:auto}.rounded-2xl{border-radius:1rem}.rounded-lg{border-radius:.5rem}.rounded-md{border-radius:.375rem}.border{border-width:1px}.border-\\[\\#f1f2f3\\]{--tw-border-opacity:1;border-color:rgb(241 242 243/var(--tw-border-opacity))}.border-transparent{border-color:transparent}.bg-\\[\\#f1f2f3\\]{--tw-bg-opacity:1;background-color:rgb(241 242 243/var(--tw-bg-opacity))}.bg-black{--tw-bg-opacity:1;background-color:rgb(0 0 0/var(--tw-bg-opacity))}.bg-slate-200{--tw-bg-opacity:1;background-color:rgb(226 232 240/var(--tw-bg-opacity))}.bg-white{--tw-bg-opacity:1;background-color:rgb(255 255 255/var(--tw-bg-opacity))}.bg-opacity-10{--tw-bg-opacity:0.1}.bg-opacity-25{--tw-bg-opacity:0.25}.p-4{padding:1rem}.px-2{padding-left:.5rem;padding-right:.5rem}.px-4{padding-left:1rem;padding-right:1rem}.py-1{padding-bottom:.25rem;padding-top:.25rem}.py-2{padding-bottom:.5rem;padding-top:.5rem}.py-3{padding-bottom:.75rem;padding-top:.75rem}.text-left{text-align:left}.text-center{text-align:center}.text-right{text-align:right}.align-middle{vertical-align:middle}.text-3xl{font-size:1.875rem;line-height:2.25rem}.text-sm{font-size:.875rem;line-height:1.25rem}.text-black{--tw-text-opacity:1;color:rgb(0 0 0/var(--tw-text-opacity))}.text-blue-500{--tw-text-opacity:1;color:rgb(59 130 246/var(--tw-text-opacity))}.text-gray-600{--tw-text-opacity:1;color:rgb(75 85 99/var(--tw-text-opacity))}.text-neutral-500{--tw-text-opacity:1;color:rgb(115 115 115/var(--tw-text-opacity))}.text-orange-600{--tw-text-opacity:1;color:rgb(234 88 12/var(--tw-text-opacity))}.opacity-0{opacity:0}.opacity-100{opacity:1}.shadow-lg{--tw-shadow:0 10px 15px -3px rgba(0,0,0,.1),0 4px 6px -4px rgba(0,0,0,.1);--tw-shadow-colored:0 10px 15px -3px var(--tw-shadow-color),0 4px 6px -4px var(--tw-shadow-color)}.shadow-lg,.shadow-xl{box-shadow:var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow)}.shadow-xl{--tw-shadow:0 20px 25px -5px rgba(0,0,0,.1),0 8px 10px -6px rgba(0,0,0,.1);--tw-shadow-colored:0 20px 25px -5px var(--tw-shadow-color),0 8px 10px -6px var(--tw-shadow-color)}.transition-all{transition-duration:.15s;transition-property:all;transition-timing-function:cubic-bezier(.4,0,.2,1)}.duration-200{transition-duration:.2s}.duration-300{transition-duration:.3s}.ease-in{transition-timing-function:cubic-bezier(.4,0,1,1)}.ease-out{transition-timing-function:cubic-bezier(0,0,.2,1)}.hover\\:border:hover{border-width:1px}.focus\\:border-blue-600:focus,.hover\\:border-blue-600:hover{--tw-border-opacity:1;border-color:rgb(37 99 235/var(--tw-border-opacity))}.focus\\:outline-none:focus{outline:2px solid transparent;outline-offset:2px}:is([data-theme=dark] .dark\\:border-\\[\\#131313\\]){--tw-border-opacity:1;border-color:rgb(19 19 19/var(--tw-border-opacity))}:is([data-theme=dark] .dark\\:bg-black){--tw-bg-opacity:1;background-color:rgb(0 0 0/var(--tw-bg-opacity))}:is([data-theme=dark] .dark\\:bg-neutral-800){--tw-bg-opacity:1;background-color:rgb(38 38 38/var(--tw-bg-opacity))}:is([data-theme=dark] .dark\\:bg-neutral-900){--tw-bg-opacity:1;background-color:rgb(23 23 23/var(--tw-bg-opacity))}:is([data-theme=dark] .dark\\:bg-neutral-950){--tw-bg-opacity:1;background-color:rgb(10 10 10/var(--tw-bg-opacity))}:is([data-theme=dark] .dark\\:text-neutral-400){--tw-text-opacity:1;color:rgb(163 163 163/var(--tw-text-opacity))}:is([data-theme=dark] .dark\\:text-white){--tw-text-opacity:1;color:rgb(255 255 255/var(--tw-text-opacity))}:is([data-theme=dark] .dark\\:text-zinc-400){--tw-text-opacity:1;color:rgb(161 161 170/var(--tw-text-opacity))}:is([data-theme=dark] .dark\\:focus\\:border-blue-600:focus){--tw-border-opacity:1;border-color:rgb(37 99 235/var(--tw-border-opacity))}";
styleInject(css_248z,{"insertAt":"top"});

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

function __spreadArray(to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
}

var DyteIcon = function () { return (React.createElement("svg", { width: "24", height: "24", viewBox: "0 0 24 24", fill: "none", xmlns: "http://www.w3.org/2000/svg" },
    React.createElement("path", { d: "M14 14H22V22H14V14Z", fill: "#2160FD" }),
    React.createElement("path", { d: "M18 2.0328C17.6695 2.01105 17.336 2 17 2C8.71573 2 2 8.71573 2 17C2 17.3358 2.01103 17.669 2.03275 17.9992H12V11.9992H18V2.0328Z", fill: "#2160FD" }))); };

var WarningIcon = function () { return (React.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", fill: "none", viewBox: "0 0 24 24", strokeWidth: 1.5, stroke: "currentColor", className: "w-4 h-4 inline" },
    React.createElement("path", { strokeLinecap: "round", strokeLinejoin: "round", d: "M12 9v3.75m-9.303 3.376c-.866 1.5.217 3.374 1.948 3.374h14.71c1.73 0 2.813-1.874 1.948-3.374L13.949 3.378c-.866-1.5-3.032-1.5-3.898 0L2.697 16.126zM12 15.75h.007v.008H12v-.008z" }))); };

/**
 * Implementation of ChatBot component
 *
 * @author Harsh Tomar
 */
var defaultInitialMessage = {
    from: "AI",
    message: "Hi, How can I help you today?",
};
function ChatBot(props) {
    var _this = this;
    var initialMessage = props.initialMessage;
    var messagesCache = sessionStorage.getItem("@docsai/messages");
    var messageContainer = useRef(null);
    var _a = useState(""), query = _a[0], setQuery = _a[1];
    var _b = useState(false), loading = _b[0], setLoading = _b[1];
    var _c = useState(messagesCache
        ? JSON.parse(messagesCache)
        : [initialMessage || defaultInitialMessage]), messages = _c[0], setMessages = _c[1];
    useEffect(function () {
        var _a, _b;
        // Scroll to Bottom on every message
        (_a = messageContainer.current) === null || _a === void 0 ? void 0 : _a.scroll({
            top: (_b = messageContainer.current) === null || _b === void 0 ? void 0 : _b.scrollHeight,
            behavior: "smooth",
        });
    }, [messages]);
    useEffect(function () {
        console.log("setting");
        console.log({ messages: messages });
        sessionStorage.setItem("@docsai/messages", JSON.stringify(messages));
    }, [messages]);
    var askQuestion = function (query) { return __awaiter(_this, void 0, void 0, function () {
        var resObj, res_1, err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 3, 4, 5]);
                    if (loading) {
                        console.warn("Still thinking...");
                        return [2 /*return*/];
                    }
                    setLoading(true);
                    setMessages(function (messages) { return __spreadArray(__spreadArray([], messages, true), [
                        { from: "user", message: query },
                    ], false); });
                    setMessages(function (messages) { return __spreadArray(__spreadArray([], messages, true), [
                        { from: "AI", message: "Thinking..." },
                    ], false); });
                    return [4 /*yield*/, fetch("https://docs-ai.dyte.io/query", {
                            method: "POST",
                            headers: {
                                "Content-Type": "application/json",
                            },
                            body: JSON.stringify({ query: query }),
                        })];
                case 1:
                    resObj = _a.sent();
                    return [4 /*yield*/, resObj.json()];
                case 2:
                    res_1 = _a.sent();
                    setMessages(function (messages) { return __spreadArray(__spreadArray([], messages.slice(0, -1), true), [
                        { from: "AI", message: res_1.response.text },
                    ], false); });
                    return [3 /*break*/, 5];
                case 3:
                    err_1 = _a.sent();
                    console.error(err_1);
                    return [3 /*break*/, 5];
                case 4:
                    setLoading(false);
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); };
    var _handleSubmit = function (e) { return __awaiter(_this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    e.preventDefault();
                    setQuery("");
                    return [4 /*yield*/, askQuestion(query)];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    }); };
    return (React.createElement("div", { className: "h-full flex flex-col bg-white dark:bg-neutral-950 overflow-auto" },
        React.createElement("div", { className: "max-h-[520px] flex-1 flex flex-col overflow-auto " },
            React.createElement("div", { className: "overflow-auto flex-1 text-black dark:text-white", ref: messageContainer }, messages.map(function (message, index) { return (React.createElement("div", { className: "p-4 flex justify-center ".concat(message.from === "AI"
                    ? "bg-[#f1f2f3] dark:bg-neutral-800"
                    : "bg-white dark:bg-neutral-900"), key: index },
                React.createElement("div", { className: "w-full flex overflow-auto" },
                    React.createElement("div", null,
                        React.createElement("div", { className: "w-9 h-9 flex justify-center items-center rounded-md bg-black bg-opacity-10 dark:bg-black mt-4 mr-2" }, message.from === "AI" ? (React.createElement(DyteIcon, null)) : ("U"))),
                    React.createElement("div", { className: "markdown-container overflow-auto py-1" },
                        React.createElement(ReactMarkdown, { linkTarget: "_blank", remarkPlugins: [remarkGfm] }, message.message))))); }))),
        React.createElement("div", { className: "p-4 " },
            React.createElement("div", { className: " mx-auto" },
                React.createElement("form", { onSubmit: _handleSubmit },
                    React.createElement("input", { type: "text", className: "px-4 py-3 w-full text-black dark:text-white dark:bg-neutral-800 rounded-lg shadow-lg border border-[#f1f2f3] dark:border-[#131313] focus:outline-none focus:border-blue-600 dark:focus:border-blue-600 ", placeholder: "Ask me anything about Dyte", value: query, onChange: function (e) { return setQuery(e.target.value); }, autoFocus: true })),
                React.createElement("div", { className: "flex justify-between" },
                    React.createElement("p", { className: "mt-2 text-left text-sm text-orange-600" },
                        React.createElement(WarningIcon, null),
                        " Still in beta, some responses may be incorrect"),
                    React.createElement("p", { className: "mt-2 text-right text-sm text-neutral-500 dark:text-neutral-400" },
                        "Powered by",
                        " ",
                        React.createElement("a", { className: "text-blue-500", href: "https://dyte.io/blog" }, "Dyte AI")))))));
}

function Modal(props) {
    var _a = useState(props.open), isOpen = _a[0], setIsOpen = _a[1];
    function closeModal() {
        setIsOpen(false);
        props.onClose();
    }
    useEffect(function () {
        setIsOpen(props.open);
    }, [props.open]);
    return (React.createElement(Transition, { show: isOpen, as: Fragment },
        React.createElement(Dialog, { as: "div", className: "relative z-10", onClose: closeModal },
            React.createElement(Transition.Child, { as: Fragment, enter: "ease-out duration-300", enterFrom: "opacity-0", enterTo: "opacity-100", leave: "ease-in duration-200", leaveFrom: "opacity-100", leaveTo: "opacity-0" },
                React.createElement("div", { className: "fixed inset-0 bg-black bg-opacity-25" })),
            React.createElement("div", { className: "fixed inset-0 overflow-y-auto" },
                React.createElement("div", { className: "flex items-center justify-center p-4 text-center" },
                    React.createElement(Transition.Child, { as: Fragment, enter: "ease-out duration-300", enterFrom: "opacity-0 scale-95", enterTo: "opacity-100 scale-100", leave: "ease-in duration-200", leaveFrom: "opacity-100 scale-100", leaveTo: "opacity-0 scale-95" },
                        React.createElement(Dialog.Panel, { className: "w-full max-w-2xl transform rounded-2xl text-left overflow-hidden align-middle shadow-xl transition-all" }, props.children)))))));
}

var SearchIcon = function () { return (React.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", fill: "none", viewBox: "0 0 24 24", strokeWidth: 1.5, stroke: "currentColor", className: "w-6 h-6" },
    React.createElement("path", { strokeLinecap: "round", strokeLinejoin: "round", d: "M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z" }))); };
function DocsAI() {
    var _a = useState(false), isOpen = _a[0], setIsOpen = _a[1];
    function openModal() {
        setIsOpen(true);
    }
    return (React.createElement(React.Fragment, null,
        React.createElement("div", { className: "flex justify-center py-2" },
            React.createElement("button", { onClick: openModal, className: "w-48 py-2 px-2 flex bg-slate-200 dark:bg-black text-gray-600 dark:text-zinc-400 rounded-md shadow-xl border border-transparent hover:border hover:border-blue-600 cursor-text" },
                React.createElement("span", { className: "inline" },
                    React.createElement(SearchIcon, null)),
                "\u00A0\u00A0Ask me Anything")),
        React.createElement(Modal, { open: isOpen, onClose: function () { return setIsOpen(false); } },
            React.createElement(ChatBot, null))));
}

export { ChatBot, DocsAI };
