// @dytesdk/web-core/types/src/utils/opentelemetry-helpers/loggableMediaState.js
export default function getLoggableCurrentMediaStates(participant) {
    if (!participant) {
        return {};
    }
    return {
        media: {
            audio: {
                enabled: participant.audioEnabled,
                trackId: participant.audioTrack?.id,
                permission: 'mediaPermissions' in participant ? participant.mediaPermissions?.audio : null,
            },
            video: {
                enabled: participant.videoEnabled,
                trackId: participant.videoTrack?.id,
                permission: 'mediaPermissions' in participant ? participant.mediaPermissions?.video : null,
            },
            screenshare: {
                enabled: participant.screenShareEnabled,
                permission: 'mediaPermissions' in participant ? participant.mediaPermissions?.screenshare : null,
                audio: {
                    enabled: participant.screenShareTracks?.audio?.enabled,
                    trackId: participant.screenShareTracks?.audio?.id,
                },
                video: {
                    enabled: participant.screenShareTracks?.video?.enabled,
                    trackId: participant.screenShareTracks?.video?.id,
                },
            },
        },
    };
}
